// @creatio/base/esm2022/lib/public/esq/data-model/filters/base-filter.mjs
import { classToPlain } from 'class-transformer';
/**
 * @public
 * @abstract
 * @description The base class of the filter.
 * @extends {ɵSerializable}
 */
export class ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [this._instanceOfKey];
    }
    /**
     * @public
     * @description Method for checking if an instance is of the current class type.
     * @param instance - The instance to check.
     * @returns `true` if the instance is of the current class type, `false` otherwise.
     */
    static [Symbol.hasInstance](instance) {
        const instanceOfKeys = instance?.constructor?.['_instanceOfKeys'] ?? [];
        return instanceOfKeys.includes(this._instanceOfKey);
    }
    constructor(filterType) {
        /**
         * @description Indicates that the filter is active and will be taken into account when building the query.
         */
        this.isEnabled = true;
        /**
         * @description Drop the time for the date / time type parameters.
         */
        this.trimDateTimeParameterToDate = false;
        this.filterType = filterType;
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        throw new Error(`fromJson not supported on filter type ${this.constructor.name}`);
    }
    serializeItem(item) {
        return typeof item.toJson === 'function'
            ? item.toJson()
            : classToPlain(item, {
                exposeUnsetFields: false,
                excludePrefixes: ['_'],
            });
    }
    /**
     * Converts instance to json object.
     * @public
     */
    toJson() {
        return classToPlain(this, {
            exposeUnsetFields: false,
            excludePrefixes: ['_'],
        });
    }
}
