// @creatio/base/esm2022/lib/public/esq/data-model/filters/between-filter.mjs
import { ɵSingleFilter } from './single-filter';
import { ɵComparisonType, ɵFilterType } from '../enums';
/**
 * @public
 * @description Filter class of the "Between" type.
 * @extends {ɵSingleFilter}
 */
export class ɵBetweenFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BetweenFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(leftExpression, rightLessExpression, rightGreaterExpression) {
        super(ɵFilterType.Between, ɵComparisonType.Between, leftExpression);
        this.rightLessExpression = rightLessExpression?.clone();
        this.rightGreaterExpression = rightGreaterExpression?.clone();
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const filter = new ɵBetweenFilter(expressionParser.fromJson(dto['leftExpression'], filterParser), expressionParser.fromJson(dto['rightLessExpression'], filterParser), expressionParser.fromJson(dto['rightGreaterExpression'], filterParser));
        return filter;
    }
    clone() {
        const filter = new ɵBetweenFilter(this.leftExpression, this.rightLessExpression, this.rightGreaterExpression);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
    toJson() {
        return {
            ...super.toJson(),
            rightLessExpression: this.serializeItem(this.rightLessExpression),
            rightGreaterExpression: this.serializeItem(this.rightGreaterExpression),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            ...super._getUsedColumns(),
            ...this.rightLessExpression._getUsedColumns(),
            ...this.rightGreaterExpression._getUsedColumns(),
        ];
    }
}
