// @creatio/base/esm2022/lib/public/esq/data-model/filters/filter-group.mjs
import { __decorate, __metadata } from "tslib";
import { Exclude, Expose } from 'class-transformer';
import { ɵBaseFilter } from './base-filter';
import { ɵBetweenFilter } from './between-filter';
import { ɵCompareFilter } from './compare-filter';
import { ɵExistsFilter } from './exists-filter';
import { ɵInFilter } from './in-filter';
import { ɵIsNullFilter } from './isnull-filter';
import { ɵgenerateGuid } from '../../../../public/types';
import { ɵValidationUtilities } from '../../../exceptions';
import { ɵComparisonType } from '../enums/comparison-type.enum';
import { ɵFilterType } from '../enums/filter-type.enum';
import { ɵLogicalOperatorType } from '../enums/logical-operator-type.enum';
import { ɵQueryMacrosType } from '../enums/query-macros-type.enum';
import { ɵColumnExpression } from '../expressions/columns/column-expression';
import { ɵMacrosFunctionExpression } from '../expressions/functions/macros-function-expression';
import { ɵParameterExpression } from '../expressions/parameters/parameter-expression';
/**
 * @public
 * @description Group of filters. Groups of filters can be nested into each other,
 * i.e. The collection itself can be an element of another collection.
 * @extends {ɵBaseFilter}
 */
export class ɵFilterGroup extends ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_FilterGroup'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseFilter._instanceOfKeys, this._instanceOfKey];
    }
    /**
     * @public
     * @description Return collecion of deserialized filters.
     */
    get items() {
        const result = {};
        this.filters.forEach((filter, key) => {
            result[key] = this.serializeItem(filter);
        });
        return result;
    }
    constructor(logicalOperation = ɵLogicalOperatorType.And, rootSchemaName) {
        super(ɵFilterType.FilterGroup);
        /**
         * @public
         * @description Collection of child filters.
         */
        this.filters = new Map();
        this.logicalOperation = logicalOperation;
        this.rootSchemaName = rootSchemaName;
    }
    /**
     * @public
     * @description Parses filter group from json.
     */
    static fromJson(dto, filterParser) {
        const logicalOperation = dto['logicalOperation'];
        const group = new ɵFilterGroup(logicalOperation || ɵLogicalOperatorType.And, dto['rootSchemaName']);
        Object.entries(dto['items']).forEach(([key, filterDTO]) => {
            group.add(filterParser.fromJson(filterDTO), key);
        });
        return group;
    }
    /** {@inheritDoc ɵSerializable.toJson} */
    toJson() {
        const result = {};
        result['filterType'] = this.filterType;
        result['isEnabled'] = this.isEnabled;
        result['logicalOperation'] = this.logicalOperation;
        result['trimDateTimeParameterToDate'] = this.trimDateTimeParameterToDate;
        result['items'] = this.items;
        if (this.rootSchemaName) {
            result['rootSchemaName'] = this.rootSchemaName;
        }
        return result;
    }
    _createParameterValueExpressions(values) {
        return values.map((value) => new ɵParameterExpression({
            value: value,
        }));
    }
    /**
     * @public
     * @description Returns filter item by index.
     * @param index Item index.
     * @returns Filter.
     */
    getItemByIndex(index) {
        return this.items[Object.keys(this.items)[index]];
    }
    /**
     * @public
     * @description Creates and adds compare filter by value.
     * @param comparisonType Comparison type.
     * @param columnPath Column path.
     * @param parameterValue Parameter value.
     * @param filterKey Filter key.
     */
    addSchemaColumnFilterWithParameter(comparisonType, columnPath, parameterValue, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const rightExpression = new ɵParameterExpression({ value: parameterValue });
        const filter = new ɵCompareFilter(comparisonType, leftExpression, rightExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds in filter by values.
     * @param comparisonType Comparison type.
     * @param columnPath Column path.
     * @param parameterValues Parameter values.
     * @param filterKey Filter key.
     */
    addSchemaColumnInFilterWithParameters(comparisonType, columnPath, parameterValues, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const expressions = this._createParameterValueExpressions(parameterValues);
        const filter = new ɵInFilter(comparisonType, leftExpression, expressions);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds in filter for primary column.
     * @param primaryColumnValues Primary column values.
     * @param filterKey Filter key.
     */
    addPrimarySchemaColumnInFilter(primaryColumnValues, filterKey) {
        const leftExpression = new ɵMacrosFunctionExpression({ macrosType: ɵQueryMacrosType.PrimaryColumn });
        const expressions = this._createParameterValueExpressions(primaryColumnValues);
        const filter = new ɵInFilter(ɵComparisonType.Equal, leftExpression, expressions);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds null filter.
     * @param columnPath Path of column.
     * @param filterKey Filter key.
     */
    addSchemaColumnIsNullFilter(columnPath, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵIsNullFilter(ɵComparisonType.Is_null, leftExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds not null filter.
     * @param columnPath Column path.
     * @param filterKey Filter key.
     */
    addSchemaColumnIsNotNullFilter(columnPath, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵIsNullFilter(ɵComparisonType.Is_not_null, leftExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds filter between less value and greater value.
     * @param columnPath Column path.
     * @param lessParamValue Less parameter value.
     * @param greaterParamValue Greater parameter value.
     * @param filterKey Filter key.
     */
    addSchemaColumnBetweenFilterWithParameters(columnPath, lessParamValue, greaterParamValue, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const rightLessExpression = new ɵParameterExpression({ value: lessParamValue });
        const rightGreaterExpression = new ɵParameterExpression({ value: greaterParamValue });
        const filter = new ɵBetweenFilter(leftExpression, rightLessExpression, rightGreaterExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds exists filter.
     * @param columnPath Column path.
     * @param subFilters Filters for exists expression.
     * @param filterKey Filter key.
     */
    addExistsFilter(columnPath, subFilters, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵExistsFilter(leftExpression, subFilters);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds not exists filter.
     * @param columnPath Column path.
     * @param subFilters Filters for exist expression.
     * @param filterKey Filter key.
     */
    addNotExistsFilter(columnPath, subFilters, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵExistsFilter(leftExpression, subFilters, ɵComparisonType.Not_exists);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Adds filter.
     * @param filter Instance of filter.
     * @param filterKey Filter key.
     */
    add(filter, filterKey) {
        ɵValidationUtilities.checkArgumentEmpty('filter', filter);
        const key = filterKey || ɵgenerateGuid();
        this.filters.set(key, filter);
    }
    /**
     * @public
     * @description Clones instance.
     */
    clone() {
        const filterGroup = new ɵFilterGroup(this.logicalOperation, this.rootSchemaName);
        this.filters.forEach((filter, filterKey) => {
            filterGroup.filters.set(filterKey, filter.clone());
        });
        filterGroup.isEnabled = this.isEnabled;
        filterGroup.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filterGroup;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        const result = [];
        this.filters.forEach((filter) => {
            let filterUsedColumns = filter._getUsedColumns ? filter._getUsedColumns() : [];
            filterUsedColumns = filterUsedColumns.map((usedColumn) => ({
                path: usedColumn.path,
                schemaName: usedColumn.schemaName ?? this.rootSchemaName,
            }));
            result.push(...filterUsedColumns);
        });
        return result;
    }
}
__decorate([
    Exclude(),
    __metadata("design:type", Object)
], ɵFilterGroup.prototype, "filters", void 0);
__decorate([
    Expose(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], ɵFilterGroup.prototype, "items", null);
