// @creatio-devkit/common/esm2022/lib/internal/services/interface-designer-item-registry/interface-designer-item-registry.mjs
import { ɵBaseRegistry } from '../../../internal/services/base-registry';
import { checkCrtTypeNamePascalCase, checkOnlyOneRegistration, checkTypeNameFormat, } from '../../../internal/services/element-registration-checker';
/**
 * @internal
 */
export const interfaceDesignerItemRegistryKey = 'InterfaceDesignerItemRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.InterfaceDesignerItemRegistry
 */
export class InterfaceDesignerItemRegistry extends ɵBaseRegistry {
    static { this.registryKey = interfaceDesignerItemRegistryKey; }
    static validate(config) {
        checkTypeNameFormat(config.type, 'InterfaceDesignerItemTypeName');
        checkCrtTypeNamePascalCase(config.type);
        checkOnlyOneRegistration(config.type, config, this.interfaceDesignerItemRegistrationInfos);
    }
    static register(config) {
        this.interfaceDesignerItemRegistrationInfos.set(config.type, config);
    }
    static get interfaceDesignerItemRegistrationInfos() {
        return this.store.typeInfos;
    }
    static get(type) {
        return this.interfaceDesignerItemRegistrationInfos.get(type);
    }
}
