// @creatio/base/esm2022/lib/public/esq/data-model/enums/admin-unit-role-sources.enum.mjs
/**
 * @public
 * @description SysAdminUnitInRole source type.
 */
export var eAdminUnitRoleSources;
(function (eAdminUnitRoleSources) {
    /**
     * @description Empty value.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["None"] = 0] = "None";
    /**
     * @description Self.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["Self"] = 1] = "Self";
    /**
     * @description Explicit entry into role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["ExplicitEntry"] = 2] = "ExplicitEntry";
    /**
     * @description Delegated role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["Delegated"] = 4] = "Delegated";
    /**
     * @description Gets functional role from organizational role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["FuncRoleFromOrgRole"] = 8] = "FuncRoleFromOrgRole";
    /**
     * @description Gets role up hierarchy from role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["UpHierarchy"] = 16] = "UpHierarchy";
    /**
     * @description Gets role as a manager.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["AsManager"] = 32] = "AsManager";
    /**
     * @description Gets all roles.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["All"] = 63] = "All";
})(eAdminUnitRoleSources || (eAdminUnitRoleSources = {}));
