// @creatio/base/esm2022/lib/public/esq/data-model/enums/comparison-type.enum.mjs
/**
 * @public
 * @description The type of comparison operation in the filter.
 */
/* eslint-disable @typescript-eslint/naming-convention */
export var ɵComparisonType;
(function (ɵComparisonType) {
    /**
     * @public
     * @description Value is between two specified values.
     */
    ɵComparisonType[ɵComparisonType["Between"] = 0] = "Between";
    /**
     * @public
     * @description Value contains a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Contain"] = 11] = "Contain";
    /**
     * @public
     * @description Value ends with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["End_with"] = 13] = "End_with";
    /**
     * @public
     * @description Value is equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Equal"] = 3] = "Equal";
    /**
     * @public
     * @description Value exists (is not null or undefined).
     */
    ɵComparisonType[ɵComparisonType["Exists"] = 15] = "Exists";
    /**
     * @public
     * @description Value is greater than a specified value.
     */
    ɵComparisonType[ɵComparisonType["Greater"] = 7] = "Greater";
    /**
     * @public
     * @description Value is greater than or equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Greater_or_equal"] = 8] = "Greater_or_equal";
    /**
     * @public
     * @description Value is not null.
     */
    ɵComparisonType[ɵComparisonType["Is_not_null"] = 2] = "Is_not_null";
    /**
     * @public
     * @description Value is null.
     */
    ɵComparisonType[ɵComparisonType["Is_null"] = 1] = "Is_null";
    /**
     * @public
     * @description Value is less than a specified value.
     */
    ɵComparisonType[ɵComparisonType["Less"] = 5] = "Less";
    /**
     * @public
     * @description Value is less than or equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Less_or_equal"] = 6] = "Less_or_equal";
    /**
     * @public
     * @description Value does not contain a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_contain"] = 12] = "Not_contain";
    /**
     * @public
     * @description Value does not end with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_end_with"] = 14] = "Not_end_with";
    /**
     * @public
     * @description Value is not equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Not_equal"] = 4] = "Not_equal";
    /**
     * @public
     * @description Value does not exist (is null or undefined).
     */
    ɵComparisonType[ɵComparisonType["Not_exists"] = 16] = "Not_exists";
    /**
     * @public
     * @description Value does not start with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_start_with"] = 10] = "Not_start_with";
    /**
     * @public
     * @description Value starts with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Start_with"] = 9] = "Start_with";
})(ɵComparisonType || (ɵComparisonType = {}));
