// @creatio/base/esm2022/lib/public/esq/data-model/enums/filter-type.enum.mjs
/**
 * @public
 * @description The type of filter.
 */
export var ɵFilterType;
(function (ɵFilterType) {
    /**
     * @public
     * @description No filter applied.
     */
    ɵFilterType[ɵFilterType["None"] = 0] = "None";
    /**
     * @public
     * @description Comparison filter (e.g., equal to, greater than).
     */
    ɵFilterType[ɵFilterType["Compare"] = 1] = "Compare";
    /**
     * @public
     * @description Checks if a value is null.
     */
    ɵFilterType[ɵFilterType["IsNull"] = 2] = "IsNull";
    /**
     * @public
     * @description Checks if a value is between two specified values.
     */
    ɵFilterType[ɵFilterType["Between"] = 3] = "Between";
    /**
     * @public
     * @description Checks if a value is in a specified set of values.
     */
    ɵFilterType[ɵFilterType["In"] = 4] = "In";
    /**
     * @public
     * @description Checks if a value exists.
     */
    ɵFilterType[ɵFilterType["Exists"] = 5] = "Exists";
    /**
     * @public
     * @description A group of filters combined with logical operators (AND, OR).
     */
    ɵFilterType[ɵFilterType["FilterGroup"] = 6] = "FilterGroup";
})(ɵFilterType || (ɵFilterType = {}));
