// @creatio/base/esm2022/lib/public/esq/data-model/enums/function-type.enum.mjs
/**
 * @public
 * @description The type of function in the filter.
 */
export var ɵFunctionType;
(function (ɵFunctionType) {
    /**
     * @description Function type not defined.
     * */
    ɵFunctionType[ɵFunctionType["None"] = 0] = "None";
    /**
     * @description Substitution by macro.
     * */
    ɵFunctionType[ɵFunctionType["Macros"] = 1] = "Macros";
    /**
     * @description Aggregating function.
     * */
    ɵFunctionType[ɵFunctionType["Aggregation"] = 2] = "Aggregation";
    /**
     * @description Date fragment.
     * */
    ɵFunctionType[ɵFunctionType["DatePart"] = 3] = "DatePart";
    /**
     * @description The size of the value in bytes, used for binary data.
     * */
    ɵFunctionType[ɵFunctionType["Length"] = 4] = "Length";
    /**
     * @description Window query function
     * */
    ɵFunctionType[ɵFunctionType["Window"] = 5] = "Window";
    /**
     * @description Date add.
     * */
    ɵFunctionType[ɵFunctionType["DateAdd"] = 6] = "DateAdd";
    /**
     * @description Date diff.
     * */
    ɵFunctionType[ɵFunctionType["DateDiff"] = 7] = "DateDiff";
})(ɵFunctionType || (ɵFunctionType = {}));
