// @creatio/base/esm2022/lib/public/esq/data-model/enums/query-macros-type.enum.mjs
/**
 * @public
 * @description The type of macros.
 */
export var ɵQueryMacrosType;
(function (ɵQueryMacrosType) {
    /**
     * @description Macro type not defined.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["None"] = 0] = "None";
    /**
     * @description Current user.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUser"] = 1] = "CurrentUser";
    /**
     * @description Current user contact.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUserContact"] = 2] = "CurrentUserContact";
    /**
     * @description Yesterday.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Yesterday"] = 3] = "Yesterday";
    /**
     * @description Today.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Today"] = 4] = "Today";
    /**
     * @description Tomorrow.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Tomorrow"] = 5] = "Tomorrow";
    /**
     * @description Previous week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousWeek"] = 6] = "PreviousWeek";
    /**
     * @description This week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentWeek"] = 7] = "CurrentWeek";
    /**
     * @description Next week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextWeek"] = 8] = "NextWeek";
    /**
     * @description Previous month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousMonth"] = 9] = "PreviousMonth";
    /**
     * @description Current month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentMonth"] = 10] = "CurrentMonth";
    /**
     * @description Next month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextMonth"] = 11] = "NextMonth";
    /**
     * @description Previous quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousQuarter"] = 12] = "PreviousQuarter";
    /**
     * @description Current quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentQuarter"] = 13] = "CurrentQuarter";
    /**
     * @description Next quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextQuarter"] = 14] = "NextQuarter";
    /**
     * @description Previous half of the year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHalfYear"] = 15] = "PreviousHalfYear";
    /**
     * @description Current half-year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHalfYear"] = 16] = "CurrentHalfYear";
    /**
     * @description Next half-year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHalfYear"] = 17] = "NextHalfYear";
    /**
     * @description Last year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousYear"] = 18] = "PreviousYear";
    /**
     * @description This year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentYear"] = 19] = "CurrentYear";
    /**
     * @description Previous hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHour"] = 20] = "PreviousHour";
    /**
     * @description Current hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHour"] = 21] = "CurrentHour";
    /**
     * @description Next hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHour"] = 22] = "NextHour";
    /**
     * @description Next year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextYear"] = 23] = "NextYear";
    /**
     * @description Next N Days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDays"] = 24] = "NextNDays";
    /**
     * @description Previous N days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDays"] = 25] = "PreviousNDays";
    /**
     * @description Next N hours.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNHours"] = 26] = "NextNHours";
    /**
     * @description Previous N hours.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNHours"] = 27] = "PreviousNHours";
    /**
     * @description Primary column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColumn"] = 34] = "PrimaryColumn";
    /**
     * @description Primary displayed column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryDisplayColumn"] = 35] = "PrimaryDisplayColumn";
    /**
     * @description Primary image display column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryImageColumn"] = 36] = "PrimaryImageColumn";
    /**
     * @description Anniversary today.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearToday"] = 37] = "DayOfYearToday";
    /**
     * @description Anniversary on the date computed as today plus days offset.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearTodayPlusDaysOffset"] = 38] = "DayOfYearTodayPlusDaysOffset";
    /**
     * @description Anniversary on the next several days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDaysOfYear"] = 39] = "NextNDaysOfYear";
    /**
     * @description Anniversary on the previous several days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDaysOfYear"] = 40] = "PreviousNDaysOfYear";
    /**
     * @description Primary color display column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColorColumn"] = 41] = "PrimaryColorColumn";
})(ɵQueryMacrosType || (ɵQueryMacrosType = {}));
