// @creatio-devkit/common/esm2022/lib/internal/services/global-converters-registry/converter-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkCrtTypeNamePascalCase, checkTypeNameFormat } from '../element-registration-checker';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ConverterRegistry
 */
export class ConverterRegistry extends ɵBaseRegistry {
    static { this.registryKey = 'ConverterRegistry'; }
    static has(type) {
        return this.store.typeInfos.has(type);
    }
    static find(type) {
        return this.store.typeInfos.get(type);
    }
    static get(type) {
        const registryItem = this.find(type);
        if (!registryItem) {
            throw new Error(`'${type}' converter doesn't exist`);
        }
        return registryItem;
    }
    static register(item) {
        checkTypeNameFormat(item.type, 'ConverterTypeName');
        checkCrtTypeNamePascalCase(item.type);
        // TODO RND-21723
        const existingItem = this.find(item.type);
        this.store.typeInfos.set(item.type, {
            ...(existingItem ?? {}),
            ...item,
        });
    }
    static clear() {
        this.store.typeInfos.clear();
    }
}
