// @creatio/base/esm2022/lib/public/services/model/data-source-parameter.mjs
/**
 * @public
 * @description Enum for model parameter types.
 */
export var ɵModelParameterType;
(function (ɵModelParameterType) {
    /**
     * @public
     * @description Parameter represents primary column value.
     */
    ɵModelParameterType["PrimaryColumnValue"] = "primaryColumnValue";
    /**
     * @public
     * @description Parameter represents column value.
     */
    ɵModelParameterType["ColumnValue"] = "columnValue";
    /**
     * @public
     * @description Parameter represents filter value.
     */
    ɵModelParameterType["Filter"] = "filter";
    /**
     * @public
     * @description Parameter represents primary display value filter.
     */
    ɵModelParameterType["PrimaryDisplayValueFilter"] = "primaryDisplayValueFilter";
    /**
     * @public
     * @description Parameter is empty.
     */
    ɵModelParameterType["Empty"] = "empty";
    /**
     * @public
     * @description Parameter is never used.
     */
    ɵModelParameterType["Never"] = "never";
})(ɵModelParameterType || (ɵModelParameterType = {}));
