// @creatio-devkit/common/esm2022/lib/public/functions/bootstrap-crt-module.mjs
import { bootstrappedCrtModules } from './bootstrapped-crt-modules';
import { HandlerSourceType } from '../../internal/enums';
import { LazyModuleRef } from '../../internal/models/lazy-module-ref';
import { BootstrapNotifier, ConverterRegistry, DesignTimeViewElementRegistry, InterfaceDesignerItemRegistry, LazyModulesManager, RemoteEntriesRegistry, RequestHandlerRegistryService, ValidatorRegistry, ViewElementRegistry, } from '../../internal/services';
import { applyLocalizeMetadataFn, getConverterMetadata, getDesignTimeViewElementMetadata, getInterfaceDesignerItemMetadata, getModuleMetadata, getRequestHandlerMetadata, getValidatorMetadata, getViewElementMetadata, hasInterfaceDesignerItemMetadata, instantiate, } from '../../internal/utils';
/**
 * @internal
 */
function checkRemoteName(remoteName) {
    if (!RemoteEntriesRegistry.items.has(remoteName)) {
        throw new Error(`Remote entry with name '${remoteName}' does not exist`);
    }
}
/**
 * @internal
 */
function resistantForEach(items, callback) {
    items.forEach((item) => {
        try {
            callback(item);
        }
        catch (error) {
            console.error(error);
        }
    });
}
/**
 * @internal
 */
function registerViewElements(types) {
    resistantForEach(types, (type) => {
        const registrationConfig = getViewElementMetadata(type);
        ViewElementRegistry.validate(registrationConfig);
        ViewElementRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyViewElements(configs, lazyModuleRef) {
    resistantForEach(configs, (config) => {
        const registryItem = { ...config, lazy: true, lazyModuleRef };
        ViewElementRegistry.validate(registryItem);
        ViewElementRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerInterfaceDesignerItems(types, localizeFn) {
    resistantForEach(types, (type) => {
        if (!hasInterfaceDesignerItemMetadata(type)) {
            return;
        }
        let config = getInterfaceDesignerItemMetadata(type);
        config = applyLocalizeMetadataFn(config, localizeFn);
        const viewElementMetadata = getViewElementMetadata(type);
        config.type = viewElementMetadata.type;
        InterfaceDesignerItemRegistry.validate(config);
        InterfaceDesignerItemRegistry.register(config);
    });
}
/**
 * @internal
 */
function registerDesignTimeViewElements(types) {
    resistantForEach(types, (type) => {
        const registrationConfig = getDesignTimeViewElementMetadata(type);
        DesignTimeViewElementRegistry.validate(registrationConfig);
        DesignTimeViewElementRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerValidators(types, options) {
    resistantForEach(types, (type) => {
        const metadata = getValidatorMetadata(type);
        const registrationConfig = {
            ...metadata,
            instantiate: function (injectionContext) {
                const resolveDependency = options?.resolveDependency
                    ? (token) => options.resolveDependency(token, injectionContext)
                    : null;
                return instantiate(type, resolveDependency);
            },
        };
        ValidatorRegistry.validate(registrationConfig);
        ValidatorRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyValidators(configs, lazyModuleRef) {
    resistantForEach(configs, (config) => {
        const registryItem = { ...config, lazy: true, lazyModuleRef };
        ValidatorRegistry.validate(registryItem);
        ValidatorRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerConverters(types, options) {
    resistantForEach(types, (type) => {
        const metadata = getConverterMetadata(type);
        const registrationConfig = {
            ...metadata,
            instantiate: function (injectionContext) {
                const resolveDependency = options?.resolveDependency
                    ? (token) => options.resolveDependency(token, injectionContext)
                    : null;
                return instantiate(type, resolveDependency);
            },
        };
        ConverterRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyConverters(descriptors, lazyModuleRef) {
    resistantForEach(descriptors, (descriptor) => {
        const registryItem = { ...descriptor, lazy: true, lazyModuleRef };
        ConverterRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerRequestHandlers(types, options, remoteName) {
    if (options.moduleType === 'remote' && !remoteName) {
        console.error("To use request handlers, 'remoteName' parameter should be specified.");
        return;
    }
    resistantForEach(types, (type) => {
        const metadata = getRequestHandlerMetadata(type);
        let registrationConfig;
        const createHandler = (injectionContext) => {
            const resolveDependency = options?.resolveDependency
                ? (token) => options.resolveDependency(token, injectionContext)
                : null;
            return instantiate(type, resolveDependency);
        };
        if (options.moduleType === 'host') {
            registrationConfig = {
                ...metadata,
                source: { type: HandlerSourceType.Host },
                createHandler,
            };
        }
        else {
            const remote = RemoteEntriesRegistry.items.get(remoteName);
            registrationConfig = {
                ...metadata,
                source: {
                    type: HandlerSourceType.Remote,
                    packageName: remote.packageName,
                },
                createHandler,
            };
        }
        RequestHandlerRegistryService.register(registrationConfig, type);
    });
}
/**
 * @internal
 */
function registerLazyRequestHandlers(descriptors, options, remoteName, lazyModuleRef) {
    resistantForEach(descriptors, (descriptor) => {
        const registrationConfig = {
            ...descriptor,
            lazy: true,
            lazyModuleRef,
        };
        if (options.moduleType === 'host') {
            registrationConfig.source = { type: HandlerSourceType.Host };
        }
        else {
            const remote = RemoteEntriesRegistry.items.get(remoteName);
            registrationConfig.source = {
                type: HandlerSourceType.Remote,
                packageName: remote.packageName,
            };
        }
        RequestHandlerRegistryService.register(registrationConfig);
    });
}
/**
 * @internal
 */
function bootstrapLazyCrtModule(name, descriptor, options, lazyModuleRef) {
    try {
        if (descriptor.viewElements) {
            registerLazyViewElements(descriptor.viewElements, lazyModuleRef);
        }
        if (descriptor.requestHandlers) {
            registerLazyRequestHandlers(descriptor.requestHandlers, options, name, lazyModuleRef);
        }
        if (descriptor.validators) {
            registerLazyValidators(descriptor.validators, lazyModuleRef);
        }
        if (descriptor.converters) {
            registerLazyConverters(descriptor.converters, lazyModuleRef);
        }
        descriptor.includes?.forEach((m) => {
            bootstrapLazyCrtModule(name, m, options, lazyModuleRef);
        });
    }
    catch (error) {
        console.error(error);
    }
}
/**
 * @internal
 */
function bootstrapDefaultCrtModule(name, type, options, root) {
    if (bootstrappedCrtModules.has(type)) {
        return;
    }
    const source = options.moduleType === 'host' ? 'host' : name;
    bootstrappedCrtModules.set(type, { source, root, options });
    try {
        const moduleDefinition = getModuleMetadata(type);
        if (moduleDefinition.viewElements) {
            registerViewElements(moduleDefinition.viewElements);
            registerInterfaceDesignerItems(moduleDefinition.viewElements, options?.localizeMetadata);
        }
        if (moduleDefinition.designTimeViewElements) {
            registerDesignTimeViewElements(moduleDefinition.designTimeViewElements);
        }
        if (moduleDefinition.requestHandlers) {
            registerRequestHandlers(moduleDefinition.requestHandlers, options, name);
        }
        if (moduleDefinition.validators) {
            registerValidators(moduleDefinition.validators, options);
        }
        if (moduleDefinition.converters) {
            registerConverters(moduleDefinition.converters, options);
        }
        moduleDefinition.includes?.forEach((includedModuleMetadata) => {
            if (typeof includedModuleMetadata === 'function') {
                bootstrapDefaultCrtModule(name, includedModuleMetadata, options, false);
                bootstrapPredefinedModules(name, includedModuleMetadata, options);
            }
            else if (typeof includedModuleMetadata === 'object') {
                if (options.moduleType === 'remote' && !name) {
                    console.error("To use lazy loading, 'remoteName' parameter should be specified.");
                    return;
                }
                const ref = new LazyModuleRef(includedModuleMetadata, (x) => bootstrapDefaultCrtModule(name, x, options, false), source);
                bootstrapLazyCrtModule(name, includedModuleMetadata, options, ref);
                LazyModulesManager.instance.add(ref);
            }
        });
    }
    catch (error) {
        console.error(error);
    }
}
/**
 * @internal
 */
function bootstrapPredefinedModules(name, type, options) {
    const modulePrefix = 'module:';
    const metadataKeys = Reflect.getOwnMetadataKeys(type);
    for (const key of metadataKeys) {
        if (typeof key === 'string' && key.startsWith(modulePrefix)) {
            const moduleType = key.substring(modulePrefix.length);
            const predefinedBootstrapper = Reflect.getOwnMetadata('moduleBootstrapper:' + moduleType, type);
            try {
                predefinedBootstrapper(name, type, options);
            }
            catch (error) {
                console.error(error);
            }
        }
    }
}
/**
 * @internal
 */
export function bootstrapCrtModule(...args) {
    const name = typeof args[0] === 'string' ? args.shift() : undefined;
    const type = args[0];
    const options = args[1] ?? {};
    options.moduleType ??= 'remote';
    if (name) {
        if (options.moduleType === 'remote') {
            checkRemoteName(name);
        }
        else {
            throw new Error('Host module cannot have name');
        }
    }
    bootstrapDefaultCrtModule(name, type, options, true);
    bootstrapPredefinedModules(name, type, options);
    if (options.moduleType === 'remote') {
        BootstrapNotifier.instance.notify(type);
    }
}
