// @creatio/base/esm2022/lib/public/esq/data-model/expressions/functions/macros-function-expression.mjs
import { ɵFunctionExpression } from './function-expression';
import { ɵFunctionType } from '../../enums';
/**
 * @public
 * @description A class that describes expression with macros function.
 * @extends {ɵFunctionExpression}
 */
export class ɵMacrosFunctionExpression extends ɵFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_MacrosFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Function type. Default 'Macros'.
         */
        this.functionType = ɵFunctionType.Macros;
        this.macrosType = config?.macrosType;
        this.functionArgument = config?.functionArgument;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = filterParser && expressionParser && dto['functionArgument']
            ? expressionParser.fromJson(dto['functionArgument'], filterParser)
            : undefined;
        return new ɵMacrosFunctionExpression({ functionArgument, macrosType: dto['macrosType'] });
    }
    /**
     * @description method to create cloned instance.
     */
    clone() {
        return new ɵMacrosFunctionExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.functionArgument?._getUsedColumns() ?? [];
    }
}
