// @creatio/base/esm2022/lib/public/esq/data-model/columns/base-query-column.mjs
import { ɵValidationUtilities } from '../../../../public/exceptions';
import { ɵOrderDirection } from '../enums';
/**
 * @public
 * @abstract
 * @classdesc Class contains info about column for query.
 */
export class ɵBaseQueryColumn {
    constructor(expression) {
        /**
         * @description Order direction.
         */
        this.orderDirection = ɵOrderDirection.None;
        /**
         * @description Column order position.
         */
        this.orderPosition = -1;
        /**
         * @description If true then column data will be present in the result set.
         */
        this.isVisible = true;
        this.expression = expression;
    }
    /**
     * Adds ordering options.
     * @param orderDirection Direction.
     * @param orderPosition Position.
     * @returns Self.
     */
    withOrdering(orderDirection, orderPosition) {
        ɵValidationUtilities.checkArgumentEmpty('orderDirection', orderDirection);
        ɵValidationUtilities.checkArgumentEmpty('orderPosition', orderPosition);
        ɵValidationUtilities.checkArgumentOutOfRange('orderPosition', orderPosition, { begin: -1 });
        this.orderDirection = orderDirection;
        this.orderPosition = orderPosition;
        return this;
    }
    /**
     * Adds options.
     * @param options Options.
     * @returns Self.
     */
    withOptions(options) {
        ɵValidationUtilities.checkArgumentEmpty('options', options);
        this.isVisible = options.isVisible ?? true;
        return this;
    }
    /**
     * Adds caption.
     * @param caption Caption.
     * @returns Self.
     */
    withCaption(caption) {
        ɵValidationUtilities.checkArgumentEmpty('caption', caption);
        this.caption = caption;
        return this;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.expression._getUsedColumns();
    }
}
