// @creatio/base/esm2022/lib/public/esq/data-model/expressions/arithmetic/arithmetic-expression.mjs
import { ɵExpressionType } from '../../enums';
import { ɵBaseExpression } from '../base-expression';
/**
 * @public
 * @description Arithmetic expression.
 * @extends {ɵBaseExpression}
 */
export class ɵArithmeticExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ArithmeticExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Expression type. Default 'ArithmeticOperation'.
         */
        this.expressionType = ɵExpressionType.ArithmeticOperation;
        this.arithmeticOperation = config?.arithmeticOperation;
        this.leftArithmeticOperand = config?.leftArithmeticOperand.clone();
        this.rightArithmeticOperand = config?.rightArithmeticOperand.clone();
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const leftArithmeticOperand = expressionParser?.fromJson(dto['leftArithmeticOperand'], filterParser);
        const rightArithmeticOperand = expressionParser?.fromJson(dto['rightArithmeticOperand'], filterParser);
        const config = {
            leftArithmeticOperand,
            rightArithmeticOperand,
            arithmeticOperation: dto['arithmeticOperation'],
            isBlock: dto['isBlock'],
        };
        return new ɵArithmeticExpression(config);
    }
    /**
     * @public
     * @description Clones instance.
     */
    clone() {
        return new ɵArithmeticExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [...this.leftArithmeticOperand._getUsedColumns(), ...this.rightArithmeticOperand._getUsedColumns()];
    }
}
