// @creatio/base/esm2022/lib/public/exceptions/validation-utilities.mjs
import { ɵArgumentEmptyException } from './argument-empty-exception';
import { ɵArgumentOutOfRangeException } from './argument-out-of-range-exception';
/**
 * @public
 * @description Utilities class with various validation methods.
 */
export class ɵValidationUtilities {
    // region Methods: Public
    /**
     * @public
     * @description Checks if argument is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkArgumentEmpty(argumentName, argumentValue) {
        if (argumentValue === undefined || argumentValue === null) {
            throw new ɵArgumentEmptyException(argumentName);
        }
    }
    /**
     * @public
     * @description Checks if array is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkArrayEmpty(argumentName, argumentValue) {
        ɵValidationUtilities.checkArgumentEmpty(argumentName, argumentValue);
        if (argumentValue.length === 0) {
            throw new ɵArgumentEmptyException(argumentName);
        }
    }
    /**
     * @public
     * @description Checks if string is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkStringIsNotEmpty(argumentName, argumentValue) {
        this.checkArgumentEmpty(argumentName, argumentValue);
        if (argumentValue === '') {
            throw new ɵArgumentEmptyException(argumentName);
        }
        return argumentValue;
    }
    /**
     * @public
     * @description Checks if argument in range.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     * @param range - Range.
     */
    static checkArgumentOutOfRange(argumentName, argumentValue, range) {
        if (range.begin && argumentValue < range.begin) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
        if (range.end && argumentValue > range.end) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
    }
    static checkEnumOutOfRange(argumentName, argumentValue, range) {
        if (!range || !Object.values(range).includes(argumentValue)) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
    }
}
