// @creatio-devkit/common/fesm2022/creatio-devkit-common.mjs
import { ɵAggregationType as _AggregationType, ɵAggregationFunction as _AggregationFunction, ɵAggregationEvalType as _AggregationEvalType, ɵDatePartType as _DatePartType, ɵFunctionType as _FunctionType, ɵOrderDirection as _OrderDirection, ɵgenerateGuid as _generateGuid, ɵNextHandlerAlreadySpecifiedException as _NextHandlerAlreadySpecifiedException, ɵBaseRequest as _BaseRequest, ɵSchemaViewMaskRequest as _SchemaViewMaskRequest, ɵSchemaViewMaskRequestAction as _SchemaViewMaskRequestAction } from '@creatio/base';
export { ɵAggregationEvalType as AggregationEvalType, ɵAggregationFunction as AggregationFunction, ɵAggregationFunctionColumn as AggregationFunctionColumn, ɵAggregationFunctionExpression as AggregationFunctionExpression, ɵAggregationSubQueryColumn as AggregationSubQueryColumn, ɵAggregationSubQueryExpression as AggregationSubQueryExpression, ɵAggregationType as AggregationType, ɵArgumentEmptyException as ArgumentEmptyException, ɵArgumentFunctionExpression as ArgumentFunctionExpression, ɵArgumentOutOfRangeException as ArgumentOutOfRangeException, ɵArithmeticExpression as ArithmeticExpression, ɵArithmeticOperation as ArithmeticOperation, ɵArithmeticQueryColumn as ArithmeticQueryColumn, ɵBaseExpression as BaseExpression, ɵBaseExpressionParser as BaseExpressionParser, ɵBaseFilter as BaseFilter, ɵBaseFilterParser as BaseFilterParser, ɵBaseFilterableQuery as BaseFilterableQuery, ɵBaseQuery as BaseQuery, ɵBaseQueryColumn as BaseQueryColumn, ɵBaseRequest as BaseRequest, ɵBetweenFilter as BetweenFilter, ɵColumnExpression as ColumnExpression, ɵCompareFilter as CompareFilter, ɵComparisonType as ComparisonType, ɵDataSchemaAttributeType as DataSchemaAttributeType, ɵDataSchemaAttributeUsageType as DataSchemaAttributeUsageType, ɵDataSchemaType as DataSchemaType, ɵDataSourceScope as DataSourceScope, ɵDataValueType as DataValueType, ɵDatePartFunctionColumn as DatePartFunctionColumn, ɵDatePartFunctionExpression as DatePartFunctionExpression, ɵDatePartType as DatePartType, ɵDefaultValueSource as DefaultValueSource, ɵDeleteQuery as DeleteQuery, ɵEMPTY_GUID as EMPTY_GUID, ɵEntityQueryColumn as EntityQueryColumn, ɵEntitySchemaQuery as EntitySchemaQuery, ɵExistsFilter as ExistsFilter, ɵExpressionMock as ExpressionMock, ɵExpressionParser as ExpressionParser, ɵExpressionResolver as ExpressionResolver, ɵExpressionType as ExpressionType, ɵFilterGroup as FilterGroup, ɵFilterMock as FilterMock, ɵFilterParser as FilterParser, ɵFilterResolver as FilterResolver, ɵFilterType as FilterType, ɵFunctionExpression as FunctionExpression, ɵFunctionType as FunctionType, ɵInFilter as InFilter, ɵInsertQuery as InsertQuery, ɵIsNullFilter as IsNullFilter, ɵItemNotFoundException as ItemNotFoundException, ɵLengthFunctionColumn as LengthFunctionColumn, ɵLengthFunctionExpression as LengthFunctionExpression, ɵLogicalOperatorType as LogicalOperatorType, ɵMacrosFunctionColumn as MacrosFunctionColumn, ɵMacrosFunctionExpression as MacrosFunctionExpression, ɵMessageChannelType as MessageChannelType, ɵModelInPageAction as ModelInPageAction, ɵModelParameterType as ModelParameterType, ɵNextHandlerAlreadySpecifiedException as NextHandlerAlreadySpecifiedException, ɵOrderDirection as OrderDirection, ɵParameterExpression as ParameterExpression, ɵParameterQueryColumn as ParameterQueryColumn, ɵPlatformAPIs as PlatformAPIs, ɵQueryMacrosType as QueryMacrosType, ɵQueryOperationType as QueryOperationType, ɵSchemaViewMaskRequest as SchemaViewMaskRequest, ɵSchemaViewMaskRequestAction as SchemaViewMaskRequestAction, ɵSelectLocalizationQuery as SelectLocalizationQuery, ɵSingleFilter as SingleFilter, ɵSubQueryColumn as SubQueryColumn, ɵSubQueryExpression as SubQueryExpression, ɵSubscription as Subscription, ɵUpdateQuery as UpdateQuery, ɵValidationUtilities as ValidationUtilities, ɵWindowFunctionColumn as WindowFunctionColumn, ɵWindowFunctionExpression as WindowFunctionExpression, ɵencodeDate as encodeDate, ɵgenerateGuid as generateGuid, ɵisEmptyGuid as isEmptyGuid, ɵisGuid as isGuid, ɵtoLocalISOString as toLocalISOString } from '@creatio/base';
import { get, set, upperFirst, camelCase, isEqual, cloneDeep, isEmpty } from 'lodash';
import 'reflect-metadata';
import { __decorate, __metadata } from 'tslib';

// noinspection NonAsciiCharacters, JSNonASCIINames

/**
 * @internal
 */
function toAggregationType(aggregationFunction) {
    switch (aggregationFunction) {
        case _AggregationFunction.Count:
            return _AggregationType.Count;
        case _AggregationFunction.Sum:
            return _AggregationType.Sum;
        case _AggregationFunction.Avg:
            return _AggregationType.Avg;
        case _AggregationFunction.Min:
            return _AggregationType.Min;
        case _AggregationFunction.Max:
            return _AggregationType.Max;
        case _AggregationFunction.TopOne:
            return _AggregationType.TopOne;
        default:
            return _AggregationType.None;
    }
}
/**
 * @internal
 */
function toAggregationFunction(aggregationType) {
    switch (aggregationType) {
        case _AggregationType.Count:
            return _AggregationFunction.Count;
        case _AggregationType.Sum:
            return _AggregationFunction.Sum;
        case _AggregationType.Avg:
            return _AggregationFunction.Avg;
        case _AggregationType.Min:
            return _AggregationFunction.Min;
        case _AggregationType.Max:
            return _AggregationFunction.Max;
        case _AggregationType.TopOne:
            return _AggregationFunction.TopOne;
        default:
            return _AggregationFunction.None;
    }
}
/**
 * @internal
 */
function toAggregationEvalType(aggregationEval) {
    switch (aggregationEval) {
        case 'all':
            return _AggregationEvalType.All;
        case 'distinct':
            return _AggregationEvalType.Distinct;
        default:
            return _AggregationEvalType.None;
    }
}
/**
 * @internal
 */
function toAggregationEval(aggregationEvalType) {
    switch (aggregationEvalType) {
        case _AggregationEvalType.All:
            return 'all';
        case _AggregationEvalType.Distinct:
            return 'distinct';
        default:
            return 'none';
    }
}

/**
 * @internal
 */
function toDatePartType(aggregationFunction) {
    switch (aggregationFunction) {
        case 'minute':
            return _DatePartType.HourMinute;
        case 'hour':
            return _DatePartType.Hour;
        case 'day':
            return _DatePartType.Day;
        case 'weekday':
            return _DatePartType.WeekDay;
        case 'week':
            return _DatePartType.Week;
        case 'month':
            return _DatePartType.Month;
        case 'year':
            return _DatePartType.Year;
        default:
            return _DatePartType.None;
    }
}
/**
 * @internal
 */
function toDatePart(aggregationType) {
    switch (aggregationType) {
        case _DatePartType.HourMinute:
            return 'minute';
        case _DatePartType.Hour:
            return 'hour';
        case _DatePartType.Day:
            return 'day';
        case _DatePartType.WeekDay:
            return 'weekday';
        case _DatePartType.Week:
            return 'week';
        case _DatePartType.Month:
            return 'month';
        case _DatePartType.Year:
            return 'year';
        default:
            return 'none';
    }
}

/**
 * @internal
 */
function toStringFunctionType(functionType) {
    switch (functionType) {
        case _FunctionType.Aggregation:
            return 'aggregation';
        case _FunctionType.DatePart:
            return 'datePart';
        default:
            return 'none';
    }
}
/**
 * @internal
 */
function toNumericFunctionType(functionType) {
    switch (functionType) {
        case 'aggregation':
            return _FunctionType.Aggregation;
        case 'datePart':
            return _FunctionType.DatePart;
        default:
            return _FunctionType.None;
    }
}

/**
 * @internal
 */
function toOrderDirection(sortDirection) {
    switch (sortDirection) {
        case 'asc':
            return _OrderDirection.Asc;
        case 'desc':
            return _OrderDirection.Desc;
        default:
            return _OrderDirection.None;
    }
}
/**
 * @internal
 */
function toSortDirection(orderDirection) {
    switch (orderDirection) {
        case _OrderDirection.Asc:
            return 'asc';
        case _OrderDirection.Desc:
            return 'desc';
        default:
            return 'none';
    }
}

/**
 * @internal
 */
const crtInputPrefix = 'crtInput:';
/**
 * @internal
 */
const crtOutputPrefix = 'crtOutput:';
/**
 * @internal
 */
const crtValidationInputPrefix = 'crtValidationInput:';
/**
 * @internal
 */
const crtInjectPrefix = 'crtInject:';

/**
 * @internal
 */
const VIEW_ELEMENT_DEFAULT_HEIGHT = '39px';

/**
 * @internal
 */
class LazyModulesManager {
    constructor() {
        this._refs = new Set();
        this.lazyLoadingDisabled = false;
    }
    static { this._globalInstancePath = 'creatio._lazyModulesManager'; }
    static get instance() {
        if (!this._instance) {
            let globalInstance = get(window, this._globalInstancePath);
            if (!globalInstance) {
                globalInstance = new this();
                set(window, this._globalInstancePath, globalInstance);
            }
            this._instance = globalInstance;
        }
        return this._instance;
    }
    add(ref) {
        this._refs.add(ref);
    }
    size() {
        return this._refs.size;
    }
    loadAll() {
        return Promise.all(Array.from(this._refs).map((ref) => ref.loadAndBootstrap()));
    }
    clear() {
        this._refs.clear();
    }
}

/**
 * @internal
 * @hostPublicApi
 */
var ViewElementGroupType;
(function (ViewElementGroupType) {
    ViewElementGroupType["Charts"] = "Charts";
    ViewElementGroupType["Components"] = "Components";
    ViewElementGroupType["LayoutElements"] = "LayoutElements";
    ViewElementGroupType["Inputs"] = "Inputs";
    ViewElementGroupType["CustomElements"] = "CustomElements";
})(ViewElementGroupType || (ViewElementGroupType = {}));

/**
 * @internal
 * @hostPublicApi
 */
var ViewElementDataGroupType;
(function (ViewElementDataGroupType) {
    ViewElementDataGroupType["Data"] = "Data";
})(ViewElementDataGroupType || (ViewElementDataGroupType = {}));
/**
 * @internal
 * @hostPublicApi
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const DesignerToolGroupType = { ...ViewElementDataGroupType, ...ViewElementGroupType };

/**
 * Handler source types in order of call priority.
 * @internal
 * @hostPublicApi
 */
var HandlerSourceType;
(function (HandlerSourceType) {
    HandlerSourceType[HandlerSourceType["Schema"] = 0] = "Schema";
    HandlerSourceType[HandlerSourceType["Remote"] = 1] = "Remote";
    HandlerSourceType[HandlerSourceType["Host"] = 2] = "Host";
})(HandlerSourceType || (HandlerSourceType = {}));

const global$4 = window;
/**
 * @public
 * @description For internal use only.
 * @hostPublicApi
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
class ɵBaseRegistry {
    constructor() {
        /**
         * @hostPubicApi available through window.creatio.registry.<registryKey>.store
         */
        this.store = {
            typeInfos: new Map(),
        };
    }
    static { this._prefix = 'creatio.registry.'; }
    static get store() {
        if (!this._instance) {
            let globalInstance = get(global$4, this._prefix + this.registryKey);
            if (!globalInstance) {
                globalInstance = new this();
                set(global$4, this._prefix + this.registryKey, globalInstance);
            }
            this._instance = globalInstance;
        }
        return this._instance.store;
    }
}

/**
 * @internal
 */
class PkgTopology extends ɵBaseRegistry {
    static { this.registryKey = 'pkgTopology'; }
    static get items() {
        return this.store.typeInfos;
    }
    static get(pkgName) {
        const order = this.items.get(pkgName);
        if (order === undefined) {
            throw new Error(`Package with name ${pkgName} not found in topology`);
        }
        return order;
    }
}

/**
 * @internal
 */
class HandlersOrderComparer {
    constructor(_request) {
        this._request = _request;
        const scopesOrderEntries = [...(this._request.scopes ?? [])].reverse().map((scope, index) => [scope, index]);
        this._scopesOrder = Object.fromEntries(scopesOrderEntries);
    }
    _getOrderByScopes(handler) {
        if (!handler.scopes) {
            return Number.MAX_VALUE;
        }
        const scopesOrder = handler.scopes.filter((s) => s in this._scopesOrder).map((s) => this._scopesOrder[s]);
        return Math.min(...scopesOrder);
    }
    _getOrderByPkgTopology(handler) {
        if (handler.source.type === HandlerSourceType.Host) {
            return Number.MAX_VALUE;
        }
        return PkgTopology.get(handler.source.packageName);
    }
    compare(a, b) {
        const aScopeOrder = this._getOrderByScopes(a);
        const bScopeOrder = this._getOrderByScopes(b);
        let diff = aScopeOrder - bScopeOrder;
        if (diff) {
            return diff;
        }
        const aPkgTopologyOrder = this._getOrderByPkgTopology(a);
        const bPkgTopologyOrder = this._getOrderByPkgTopology(b);
        diff = aPkgTopologyOrder - bPkgTopologyOrder;
        if (diff) {
            return diff;
        }
        diff = a.source.type - b.source.type;
        return diff;
    }
}

const moduleMetadataKey = 'module';
const viewElementMetadataKey = 'viewElement';
const interfaceDesignerItemMetadataKey = 'interfaceDesignerItem';
const requestHandlerMetadataKey = 'requestHandler';
const designTimeViewElementMetadataKey = 'designTimeViewElement';
const validatorMetadataKey = 'validator';
const converterMetadataKey = 'converter';
/**
 * @internal
 */
function setModuleMetadata(metadata, type) {
    Reflect.defineMetadata(moduleMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getModuleMetadata(type) {
    const metadata = Reflect.getMetadata(moduleMetadataKey, type);
    if (!metadata) {
        throw new Error(`Module metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function setViewElementMetadata(metadata, type) {
    Reflect.defineMetadata(viewElementMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getViewElementMetadata(type) {
    const metadata = Reflect.getMetadata(viewElementMetadataKey, type);
    if (!metadata) {
        throw new Error(`View element metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function setValidatorMetadata(metadata, type) {
    Reflect.defineMetadata(validatorMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getValidatorMetadata(type) {
    const metadata = Reflect.getOwnMetadata(validatorMetadataKey, type);
    if (!metadata) {
        throw new Error(`Validator metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function setConverterMetadata(metadata, type) {
    Reflect.defineMetadata(converterMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getConverterMetadata(type) {
    const metadata = Reflect.getOwnMetadata(converterMetadataKey, type);
    if (!metadata) {
        throw new Error(`Converter metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function setInterfaceDesignerItemMetadata(metadata, type) {
    Reflect.defineMetadata(interfaceDesignerItemMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getInterfaceDesignerItemMetadata(type) {
    const metadata = Reflect.getOwnMetadata(interfaceDesignerItemMetadataKey, type);
    if (!metadata) {
        throw new Error(`Interface designer item metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function setDesignTimeViewElementMetadata(metadata, type) {
    Reflect.defineMetadata(designTimeViewElementMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getDesignTimeViewElementMetadata(type) {
    const metadata = Reflect.getOwnMetadata(designTimeViewElementMetadataKey, type);
    if (!metadata) {
        throw new Error(`Design time view element metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function hasInterfaceDesignerItemMetadata(type) {
    return Reflect.hasOwnMetadata(interfaceDesignerItemMetadataKey, type);
}
/**
 * @internal
 */
function setRequestHandlerMetadata(metadata, type) {
    Reflect.defineMetadata(requestHandlerMetadataKey, metadata, type);
}
/**
 * @internal
 */
function getRequestHandlerMetadata(type) {
    const metadata = Reflect.getMetadata(requestHandlerMetadataKey, type);
    if (!metadata) {
        throw new Error(`Request handler metadata was not found on '${type.name}' class`);
    }
    return metadata;
}
/**
 * @internal
 */
function getHandlerChainServiceMetadata(typeRef) {
    const requestType = Reflect.getMetadata('requestType', typeRef);
    if (!requestType) {
        throw new Error(`requestType was not found on '${typeRef.name}' class`);
    }
    return { requestType, typeRef };
}
/**
 * @internal
 */
function getCrtInputsMetadata(type) {
    const keys = Reflect.getMetadataKeys(type);
    const crtInputKeys = keys.filter((x) => x.startsWith(crtInputPrefix));
    return Object.fromEntries(crtInputKeys.map((key) => {
        const data = Reflect.getMetadata(key, type);
        return [key.replace(crtInputPrefix, ''), data];
    }));
}
/**
 * @internal
 */
function getCrtOutputsMetadata(type) {
    const keys = Reflect.getMetadataKeys(type);
    const crtOutputKeys = keys.filter((x) => x.startsWith(crtOutputPrefix));
    return Object.fromEntries(crtOutputKeys.map((key) => {
        const data = Reflect.getMetadata(key, type);
        return [key.replace(crtOutputPrefix, ''), data];
    }));
}
/**
 * @internal
 */
function getCrtValidationInputsMetadata(type) {
    const keys = Reflect.getMetadataKeys(type);
    const crtValidationInputKeys = keys.filter((x) => x.startsWith(crtValidationInputPrefix));
    return Object.fromEntries(crtValidationInputKeys.map((key) => {
        const data = Reflect.getMetadata(key, type);
        return [key.replace(crtValidationInputPrefix, ''), data];
    }));
}
/**
 * @internal
 */
function getCrtInjectMetadata(type) {
    const keys = Reflect.getOwnMetadataKeys(type);
    const crtInjectKeys = keys.filter((x) => x.startsWith(crtInjectPrefix));
    return Object.fromEntries(crtInjectKeys.map((key) => {
        const index = Number(key.replace(crtInjectPrefix, ''));
        const data = Reflect.getMetadata(key, type);
        return [index, data];
    }));
}
/**
 * @internal
 */
function getParentType(type) {
    const parentProto = type.prototype ? Object.getPrototypeOf(type.prototype) : null;
    return parentProto ? parentProto.constructor : null;
}
/**
 * @internal
 */
function getTypeDependencies(type) {
    const paramTypes = Reflect.getOwnMetadata('design:paramtypes', type);
    if (!paramTypes) {
        if (type.length > 0) {
            throw new Error(`Dependency Injection error. Cannot read dependencies of class '${type.name}'. ` +
                `Please make sure class '${type.name}' has decorator and ` +
                "'emitDecoratorMetadata' option is enabled in your 'tsconfig.json' file.");
        }
        const parentType = getParentType(type);
        return parentType && parentType !== Object ? getTypeDependencies(parentType) : [];
    }
    const crtInjectMetadata = getCrtInjectMetadata(type);
    Object.entries(crtInjectMetadata).forEach(([index, token]) => (paramTypes[index] = token));
    return paramTypes;
}

function raiseLazyModulesDependencyError(module) {
    // TODO RND-60883
    console.error('Not supported: lazy modules should not be dependent on other lazy modules. ' +
        'Some functions may not work correctly', module);
}
/**
 * @internal
 * Retrieves the deep module dependencies for the given types.
 *
 * @param types An array of types.
 * @returns An array of unique module dependencies.
 */
function getDeepModuleDependencies(types) {
    if (!Array.isArray(types)) {
        return [];
    }
    if (types.length === 0) {
        return [];
    }
    const uniqueModules = new Set();
    const stack = [...types];
    while (stack.length > 0) {
        const module = stack.pop();
        if (!module || uniqueModules.has(module)) {
            continue;
        }
        if (typeof module !== 'function') {
            raiseLazyModulesDependencyError(module);
            continue;
        }
        uniqueModules.add(module);
        const nestedModules = getModuleMetadata(module)?.includes || [];
        stack.push(...nestedModules);
    }
    return [...uniqueModules];
}

/**
 * @internal
 */
function matchScopes(handlerScopes, requestScopes) {
    if (!handlerScopes?.length) {
        return true;
    }
    return handlerScopes.some((s) => requestScopes?.includes(s));
}

const global$3 = window;
global$3.creatio ??= {};
global$3.creatio._bootstrappedCrtModules ??= new Map();
/**
 * @internal
 */
const bootstrappedCrtModules = global$3.creatio._bootstrappedCrtModules;

/**
 * @internal
 */
function iterateRootModules(callback) {
    bootstrappedCrtModules.forEach((info, type) => {
        if (info.root) {
            callback(info, type);
        }
    });
}
/**
 * @internal
 */
function callOnContextCreateHooks(context) {
    iterateRootModules((info) => info.options?.onContextCreate?.(context));
}
/**
 * @internal
 */
function callOnContextUpdateHooks(context) {
    iterateRootModules((info) => info.options?.onContextUpdate?.(context));
}
/**
 * @internal
 */
function callOnContextDestroyHooks(context) {
    iterateRootModules((info) => info.options?.onContextDestroy?.(context));
}

/**
 * @internal
 * @hostPublicApi available through window.handlerChain
 */
class InternalHandlerChain {
    constructor() {
        this._handlers = [];
        this._requestListeners = new Map();
        this._handlersCache = new Map();
        this._handlersScopesByContextId = new Map();
        /**
         * @hostPublicApi
         */
        this.handlerChain$ = {
            /**
             * @hostPublicApi
             */
            process: async (request) => {
                let handler;
                if (this._hasCacheHandler(request)) {
                    handler = this._getCacheHandler(request);
                }
                else {
                    const handlers = this._getMatchedHandlers(request);
                    if (!LazyModulesManager.instance.lazyLoadingDisabled) {
                        await this._loadLazyHandlers(request, handlers);
                    }
                    handler = this._buildHandlersChain(request, handlers);
                }
                this._requestListeners.forEach((listener) => listener?.(request));
                return handler?.handle(request);
            },
            /**
             * @hostPublicApi
             */
            register: (info) => {
                this._handlersCache.delete(info.requestType);
                this._handlers.push(info);
                return () => {
                    const index = this._handlers.indexOf(info);
                    if (index > -1) {
                        this._handlers.splice(index, 1);
                        this._handlersCache.delete(info.requestType);
                    }
                    else {
                        console.warn(`${info?.requestType} not found to unsubscribe`);
                    }
                };
            },
            /**
             * @hostPublicApi
             */
            update: (updatedInfo) => {
                const config = this._handlers.find((h) => h.type === updatedInfo.type);
                if (config) {
                    config.createHandler = updatedInfo.createHandler;
                }
            },
            /**
             * @hostPublicApi
             */
            subscribe: (callback, context) => {
                if (!callback) {
                    return () => void 0;
                }
                const listenerId = _generateGuid();
                const listenerFn = (request) => {
                    if (!context || request.$context === context) {
                        callback(request);
                    }
                };
                this._requestListeners.set(listenerId, listenerFn);
                return () => this._requestListeners.delete(listenerId);
            },
        };
    }
    /**
     * @hostPublicApi
     */
    clearContextedHandlersCache(contextId) {
        const configs = this._handlersScopesByContextId.get(contextId) ?? [];
        return configs.every(({ requestType, scope }) => this._handlersCache.get(requestType)?.delete(scope));
    }
    clear() {
        this._handlers.splice(0);
        this._requestListeners.clear();
        this._handlersCache.clear();
        this._handlersScopesByContextId.clear();
    }
    _getCacheHandler(request) {
        const map = this._handlersCache.get(request.type);
        const key = this._getCacheScopesKey(request);
        return map.get(key);
    }
    _hasCacheHandler(request) {
        const map = this._handlersCache.get(request.type);
        const key = this._getCacheScopesKey(request);
        return map && map.has(key);
    }
    _getCacheScopesKey(request) {
        const injectionContextKey = this._getInjectionContext(request?.$context)?.id ?? '';
        return (request.scopes || ['']).join('_').concat(injectionContextKey);
    }
    _getInjectionContext(viewModel) {
        return (viewModel?.['_injectionContext'] ||
            (viewModel?.['parent'] && this._getInjectionContext(viewModel['parent'])));
    }
    _setContextedScopes(request, scope) {
        const injectionContext = this._getInjectionContext(request?.$context)?.id;
        const configs = this._handlersScopesByContextId.has(injectionContext)
            ? this._handlersScopesByContextId.get(injectionContext)
            : [];
        configs.push({ requestType: request.type, scope });
        this._handlersScopesByContextId.set(injectionContext, configs);
    }
    _setCacheHandler(request, handler) {
        const map = this._handlersCache.has(request.type)
            ? this._handlersCache.get(request.type)
            : new Map();
        const key = this._getCacheScopesKey(request);
        map.set(key, handler);
        if (this._getInjectionContext(request?.$context)?.id) {
            this._setContextedScopes(request, key);
        }
        this._handlersCache.set(request.type, map);
    }
    _isRequestHandler(handler) {
        return (typeof handler.handle === 'function' &&
            typeof handler.setNext === 'function');
    }
    _checkHandlerInstanceOf(handler) {
        if (!this._isRequestHandler(handler)) {
            throw new Error('Handler should implement "BaseRequestHandler"');
        }
    }
    async _loadLazyHandlers(request, configs) {
        const lazyModules = new Set();
        configs.forEach((config) => {
            if (config.lazy && !config.createHandler) {
                lazyModules.add(config.lazyModuleRef);
            }
        });
        const lazyModuleRefs = [...lazyModules];
        await Promise.all(lazyModuleRefs.map((m) => m.loadAndBootstrap()));
        const injectionContext = this._getInjectionContext(request?.$context);
        if (injectionContext && lazyModules.size) {
            const deepModuleDependencies = await getDeepModuleDependencies(lazyModuleRefs.map((ref) => ref.typeRef));
            for (const type of deepModuleDependencies) {
                injectionContext.addLazyModule(type);
            }
            callOnContextUpdateHooks(injectionContext);
        }
    }
    _getMatchedHandlers(request) {
        return this._handlers.filter((h) => {
            if (h.requestType !== request.type) {
                return false;
            }
            return matchScopes(h.scopes, request.scopes);
        });
    }
    _buildHandlersChain(request, handlers) {
        const comparer = new HandlersOrderComparer(request);
        handlers.sort((a, b) => comparer.compare(a, b));
        const injectionContext = this._getInjectionContext(request?.$context);
        const handlerInstances = handlers.map((handlerConfig) => handlerConfig.createHandler.call(handlerConfig, injectionContext));
        handlerInstances.forEach((handler) => this._checkHandlerInstanceOf(handler));
        this._joinHandlersInChain(handlerInstances);
        const result = handlerInstances[0];
        this._setCacheHandler(request, result);
        return result;
    }
    _joinHandlersInChain(handlers) {
        for (let index = 0; index < handlers.length - 1; index++) {
            try {
                handlers[index].setNext(handlers[index + 1]);
            }
            catch (error) {
                if (error instanceof _NextHandlerAlreadySpecifiedException) {
                    throw new Error('Handler factory should return new instance of ' +
                        'BaseRequestHandler class or its subclass ' +
                        '(next handler should not be specified)');
                }
                throw error;
            }
        }
    }
}

const global$2 = window;
/**
 * @public
 * @class HandlerChainService
 * @classdesc Service to design handlers in chain.
 */
class HandlerChainService {
    get _chain() {
        return global$2.handlerChain;
    }
    set _chain(value) {
        global$2.handlerChain = value;
    }
    static get instance() {
        return HandlerChainService._instance || (HandlerChainService._instance = new HandlerChainService());
    }
    constructor() {
        if (!this._chain) {
            this._createGlobalHandlerChain();
        }
    }
    _createGlobalHandlerChain() {
        this._chain = new InternalHandlerChain();
    }
    /**
     * @internal
     */
    clearContextedHandlersCache(contextId) {
        return this._chain.clearContextedHandlersCache(contextId);
    }
    /**
     * @internal
     */
    register(handlerConfig) {
        return this._chain.handlerChain$.register(handlerConfig);
    }
    /**
     * @internal
     */
    update(handlerConfig) {
        return this._chain.handlerChain$.update(handlerConfig);
    }
    /**
     * @description Method to process passed request
     * @param request - {@link ɵBaseRequest}
     * @return `Promise<unknown>`
     */
    process(request) {
        return this._chain.handlerChain$.process(request);
    }
    /**
     * @internal
     * Subscribes to requests.
     * @param callback A handler function for each request triggered.
     * @param context (optional) A context in which the requests were triggered.
     * The `callback` will be called only for requests with specified context.
     * If not  specified, will be called for all requests.
     * @returns A function to cancel the subscription.
     */
    subscribe(callback, context) {
        return this._chain.handlerChain$.subscribe(callback, context);
    }
    /**
     * @internal
     */
    clear() {
        this._chain.clear();
    }
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function ImplementUsingHandlerChain(config) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function (target, method, descriptor) {
        const methodImplementation = function (...args) {
            return HandlerChainService.instance.process({
                type: config?.requestType ? config.requestType : this.requestType,
                method: config?.methodName ? config.methodName : method,
                args,
            });
        };
        if (descriptor) {
            descriptor.value = methodImplementation;
        }
        else {
            target[method] = methodImplementation;
        }
    };
}

/**
 * @internal
 */
function checkTypeNameFormat(type, element) {
    const typeParts = type.split('.');
    if (typeParts.length < 2) {
        throw new Error(`Error when register ${type}. Type property should have format "VendorPrefix.${element}"`);
    }
}
/**
 * @public
 * @description For internal use only.
 */
function ɵcheckTypeNameFormat(type, element) {
    return checkTypeNameFormat(type, element);
}
/**
 * @internal
 */
function checkTypeNamePascalCase(type) {
    const [vendor, typeName] = type.split('.');
    const pascalTypeName = toPascalCase(typeName);
    if (!vendor || pascalTypeName !== typeName) {
        throw new Error(`Error when register ${type}. Type property should have Pascal case, e.g. ${vendor + '.' + pascalTypeName}. ` + 'The first letter of each word in a compound word should be capitalized');
    }
}
/**
 * @internal
 */
function checkCrtTypeNamePascalCase(type) {
    const [vendor, typeName] = type.split('.');
    const pascalTypeName = toPascalCase(typeName);
    if (vendor.toLowerCase() === 'crt' && pascalTypeName !== typeName) {
        throw new Error(`Error when register ${type}. Type property should have Pascal case, e.g. ${vendor + '.' + pascalTypeName}. ` + 'The first letter of each word in a compound word should be capitalized');
    }
}
/**
 * @internal
 */
function toPascalCase(source) {
    return upperFirst(camelCase(source));
}
/**
 * @internal
 */
function checkIfTypeRegistered(type, item) {
    if (!item) {
        throw new Error(`Error when resolve ${type}. Type was not registered.`);
    }
}
/**
 * @internal
 */
function checkOnlyOneRegistration(type, item, map) {
    if (map.has(type) && !isEqual(map.get(type), item)) {
        throw new Error(`Error when register ${type}. Element with this type has already exist`);
    }
}

/**
 * @internal
 * @hostPublicApi available through window.creatio.requestHandlerRegistry
 */
class InternalRequestHandlerRegistryType {
    static { this._addItemListeners = []; }
    static { this._removeItemListeners = []; }
    static { this._requestHandlersMap = new Map(); }
    /**
     * @publicHostApi
     */
    static get requestHandlersMap() {
        return this._requestHandlersMap;
    }
    /**
     * @publicHostApi
     */
    static onAddRequestHandler(handler) {
        this._addItemListeners.push(handler);
        return () => {
            this._addItemListeners.splice(this._addItemListeners.indexOf(handler), 1);
        };
    }
    /**
     * @publicHostApi
     */
    static onRemoveRequestHandler(handler) {
        this._removeItemListeners.push(handler);
        return () => {
            this._removeItemListeners.splice(this._removeItemListeners.indexOf(handler), 1);
        };
    }
    /**
     * @publicHostApi
     */
    static addRequestHandler(handlerType, config) {
        const existingHandler = this._requestHandlersMap.get(handlerType);
        if (existingHandler && (!existingHandler.instanceConfig?.lazy || existingHandler.instanceType)) {
            return;
        }
        const mergedConfig = {
            instanceType: config.instanceType,
            instanceConfig: {
                ...(existingHandler?.instanceConfig ?? {}),
                ...config.instanceConfig,
            },
        };
        this._requestHandlersMap.set(handlerType, mergedConfig);
        for (const listener of this._addItemListeners) {
            listener(handlerType, mergedConfig);
        }
    }
    /**
     * @publicHostApi
     */
    static removeRequestHandler(handlerType) {
        const config = this._requestHandlersMap.get(handlerType);
        if (!config) {
            return;
        }
        this._requestHandlersMap.delete(handlerType);
        for (const listener of this._removeItemListeners) {
            listener(handlerType, config);
        }
    }
    static clear() {
        for (const type of this._requestHandlersMap.keys()) {
            this.removeRequestHandler(type);
        }
    }
}
const global$1 = window;
global$1.creatio ??= {};
if (!global$1.creatio.requestHandlerRegistry) {
    Object.defineProperty(global$1.creatio, 'requestHandlerRegistry', {
        get: () => InternalRequestHandlerRegistryType,
        configurable: false,
    });
}
/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const InternalRequestHandlerRegistry = global$1.creatio.requestHandlerRegistry;
/**
 * @internal
 */
class RequestHandlerRegistryService {
    static { this._internalRegistry = global$1.creatio.requestHandlerRegistry; }
    /**
     * Register request handler class.
     * @param config Request handler registration config.
     * @param typeRef Request handler type reference (constructor).
     */
    static register(config, typeRef) {
        const requestType = config.requestType;
        const requestHandlerType = config.type;
        if (!requestType) {
            throw new Error(`Error when register ${requestHandlerType}. Request type not specified.`);
        }
        const requestHandlerMapValue = {
            instanceType: typeRef,
            instanceConfig: config,
        };
        RequestHandlerRegistryService._internalRegistry.addRequestHandler(requestHandlerType, requestHandlerMapValue);
    }
}

/**
 * @internal
 */
class RequestRegistryService {
    static { this._requestsMap = new Map(); }
    static { this._requestsRegistrationInfos = new Map(); }
    static get requestsRegistrationInfos() {
        return this._requestsRegistrationInfos;
    }
    /**
     * Registers a specified request class.
     * @param config Request registration config.
     * @param type Request type.
     */
    static register(config, type) {
        const requestType = config.type;
        // TODO RND-21723
        RequestRegistryService._requestsMap.set(requestType, type);
        RequestRegistryService._requestsRegistrationInfos.set(requestType, config);
    }
    /**
     * Creates request instance by specified registered request type.
     * @param type Request type.
     * @returns The instance of registered request.
     */
    static get(type) {
        const requestInstanceConstructor = RequestRegistryService._requestsMap.get(type);
        checkIfTypeRegistered(type, requestInstanceConstructor);
        return requestInstanceConstructor && new requestInstanceConstructor();
    }
}

/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ConverterRegistry
 */
class ConverterRegistry extends ɵBaseRegistry {
    static { this.registryKey = 'ConverterRegistry'; }
    static has(type) {
        return this.store.typeInfos.has(type);
    }
    static find(type) {
        return this.store.typeInfos.get(type);
    }
    static get(type) {
        const registryItem = this.find(type);
        if (!registryItem) {
            throw new Error(`'${type}' converter doesn't exist`);
        }
        return registryItem;
    }
    static register(item) {
        checkTypeNameFormat(item.type, 'ConverterTypeName');
        checkCrtTypeNamePascalCase(item.type);
        // TODO RND-21723
        const existingItem = this.find(item.type);
        this.store.typeInfos.set(item.type, {
            ...(existingItem ?? {}),
            ...item,
        });
    }
    static clear() {
        this.store.typeInfos.clear();
    }
}

/**
 * @internal
 */
const viewElementRegistryKey = 'ViewElementRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
class ViewElementRegistry extends ɵBaseRegistry {
    static { this.registryKey = viewElementRegistryKey; }
    static validate(config) {
        checkTypeNameFormat(config.type, 'ViewElementTypeName');
        const existingItem = ViewElementRegistry.viewElements.get(config.type);
        if (existingItem?.lazy && !('selector' in existingItem) && !existingItem.typeRef) {
            return;
        }
        checkOnlyOneRegistration(config.type, config, ViewElementRegistry.viewElements);
    }
    static get viewElements() {
        return this.store.typeInfos;
    }
    static register(config) {
        const existingConfig = this.viewElements.get(config.type);
        this.viewElements.set(config.type, {
            ...(existingConfig ?? {}),
            ...config,
        });
    }
    static getFormControlRelatesAttribute(type) {
        const config = this.store.typeInfos.get(type);
        return (config && 'formControlConfig' in config && config.formControlConfig?.relatesTo) || null;
    }
    static getViewElementInputs(type) {
        const config = this.store.typeInfos.get(type);
        return config?.inputs ?? {};
    }
    static getViewElementOutputs(type) {
        const config = this.store.typeInfos.get(type);
        return config?.outputs ?? {};
    }
    static getViewElementPropertiesDefaultValues(type) {
        const config = this.store.typeInfos.get(type);
        const propertiesDefaultValues = Object.entries(config?.inputs ?? {})
            .filter(([_, inputConfig]) => Object.prototype.hasOwnProperty.call(inputConfig, 'defaultValue'))
            .map(([propertyName, propertyConfig]) => ({
            name: propertyName,
            defaultValue: propertyConfig.defaultValue,
        }));
        return [
            ...propertiesDefaultValues,
            {
                name: 'visible',
                defaultValue: true,
            },
        ];
    }
    static getViewElementCompatiblePlatformAPIsOptions(type, platformAPI) {
        const config = this.store.typeInfos.get(type);
        return (config?.compatibleAPIs?.[platformAPI] ?? {});
    }
    static isViewElementAvailableByCompatibleAPIs(type, ...availableAPIs) {
        const config = this.store.typeInfos.get(type);
        return Object.entries(config?.compatibleAPIs ?? {})
            .filter(([key, _]) => availableAPIs.includes(key))
            .some(([_, value]) => {
            if (typeof value === 'boolean') {
                return value;
            }
            return (value || {}).enable;
        });
    }
}

/**
 * @internal
 */
const interfaceDesignerItemRegistryKey = 'InterfaceDesignerItemRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.InterfaceDesignerItemRegistry
 */
class InterfaceDesignerItemRegistry extends ɵBaseRegistry {
    static { this.registryKey = interfaceDesignerItemRegistryKey; }
    static validate(config) {
        checkTypeNameFormat(config.type, 'InterfaceDesignerItemTypeName');
        checkCrtTypeNamePascalCase(config.type);
        checkOnlyOneRegistration(config.type, config, this.interfaceDesignerItemRegistrationInfos);
    }
    static register(config) {
        this.interfaceDesignerItemRegistrationInfos.set(config.type, config);
    }
    static get interfaceDesignerItemRegistrationInfos() {
        return this.store.typeInfos;
    }
    static get(type) {
        return this.interfaceDesignerItemRegistrationInfos.get(type);
    }
}

const global = window;
/**
 * @internal
 */
class BootstrapNotifier {
    static get instance() {
        if (!global.__crtModuleBootstrapNotifier) {
            global.__crtModuleBootstrapNotifier = new BootstrapNotifier();
        }
        return global.__crtModuleBootstrapNotifier;
    }
    constructor() {
        this._subscribers = new Set();
        // private constructor prevent instantiation via 'new' keyword
    }
    notify(moduleType) {
        for (const subscriber of this._subscribers.values()) {
            subscriber(moduleType);
        }
    }
    subscribe(subscriber) {
        this._subscribers.add(subscriber);
        return {
            unsubscribe: () => this._subscribers.delete(subscriber),
        };
    }
}

/**
 * @internal
 */
class HandlerChainServicesRegistry extends ɵBaseRegistry {
    static { this.registryKey = 'HandlerChainServicesRegistry'; }
    static validate({ requestType }) {
        checkTypeNameFormat(requestType, this.registryKey);
        checkCrtTypeNamePascalCase(requestType);
        this.checkUniqueRequestType(requestType);
    }
    static checkUniqueRequestType(requestType) {
        if (this.services.get(requestType)) {
            throw new Error(`${requestType} already registered`);
        }
    }
    static register(config) {
        this.validate(config);
        this.services.set(config.requestType, config);
    }
    static get services() {
        return this.store.typeInfos;
    }
}

/**
 * @internal
 */
const designTimeViewElementRegistryKey = 'DesignTimeViewElementRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.DesignTimeViewElementRegistry
 */
class DesignTimeViewElementRegistry extends ɵBaseRegistry {
    static { this.registryKey = designTimeViewElementRegistryKey; }
    static validate(config) {
        checkTypeNameFormat(config.type, 'DesignTimeViewElementRegistryItemTypeName');
        checkCrtTypeNamePascalCase(config.type);
        checkOnlyOneRegistration(config.type, config, this.infos);
    }
    static register(config) {
        this.infos.set(config.type, config);
    }
    static getByOverriddenType(type) {
        const item = Array.from(this.infos).find(([_, info]) => info.overrideType === type);
        if (item) {
            return item[1];
        }
        return null;
    }
    static get infos() {
        return this.store.typeInfos;
    }
}

/**
 * @internal
 */
class ValidatorMapValue {
}
/**
 * @internal
 */
const validatorRegistryKey = 'ValidatorRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
class ValidatorRegistry extends ɵBaseRegistry {
    static { this.registryKey = validatorRegistryKey; }
    static _hasWithAnotherTypeRef(configType, typeRef) {
        const existingItem = ValidatorRegistry.items.get(configType);
        if (!existingItem) {
            return false;
        }
        if (existingItem.lazy && !existingItem.typeRef) {
            return false;
        }
        return existingItem.typeRef !== typeRef;
    }
    static validate(config) {
        checkTypeNameFormat(config.type, 'ValidatorTypeName');
        if (config.typeRef && ValidatorRegistry._hasWithAnotherTypeRef(config.type, config.typeRef)) {
            throw new Error(`Error when register ${config.type}. Element with this type has already exist`);
        }
    }
    static get items() {
        return this.store.typeInfos;
    }
    static register(config) {
        const existingConfig = this.items.get(config.type);
        this.items.set(config.type, {
            ...(existingConfig ?? {}),
            ...config,
        });
    }
}

/**
 * @internal
 * @dontChange
 */
class RemoteEntriesRegistry extends ɵBaseRegistry {
    /**
     * @dontChange
     */
    static { this.registryKey = 'remotes'; }
    /**
     * @dontChange
     */
    static get items() {
        return this.store.typeInfos;
    }
}

/**
 * @internal
 */
const registryKey = 'ChatMessageTypeRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ChatMessageTypeRegistry
 */
class ChatMessageTypeRegistry extends ɵBaseRegistry {
    static { this.registryKey = registryKey; }
    static get messageTypes() {
        return this.store.typeInfos;
    }
    static register(messageType, selector) {
        this.messageTypes.set(messageType, selector);
    }
    static getSelector(messageType) {
        return this.store.typeInfos.get(messageType);
    }
}

/**
 * @public
 * @description Function allows to register user-defined component as view element in Creatio platform.
 * @param config - registration config. {@link ViewElementRegistrationConfig}
 * @example
 * sdk.registerViewElement({
 *      type: 'usr.Timeline',
 *      selector: 'usr-timeline',
 *      inputs: {
 *          primaryColumnValue: {},
 *          cardSchemaName: {},
 *          entitySchemaName: {}
 *      }
 * });
 */
function registerViewElement(config) {
    ViewElementRegistry.validate(config);
    ViewElementRegistry.register(config);
}

/**
 * @internal
 * @description Function allows to register angular component as view element in Creatio platform.
 */
function registerAngularViewElement(config, typeRef) {
    ViewElementRegistry.validate(config);
    ViewElementRegistry.register({
        ...config,
        typeRef,
    });
}

/**
 * @internal
 */
class LazyModuleRef {
    get typeRef() {
        return this._typeRef;
    }
    constructor(_moduleDefinition, _bootstrapFn, 
    /** Name of source build ('host' or remote name) */
    sourceName) {
        this._moduleDefinition = _moduleDefinition;
        this._bootstrapFn = _bootstrapFn;
        this.sourceName = sourceName;
    }
    loadAndBootstrap() {
        if (!this._modulePromise) {
            this._modulePromise = this._moduleDefinition.load().then((type) => {
                this._typeRef = type;
                this._bootstrapFn(type);
                return type;
            });
        }
        return this._modulePromise;
    }
}

/**
 * @internal
 */
function objectDeepKeys(obj) {
    if (obj === null || typeof obj !== 'object') {
        return [];
    }
    return Object.keys(obj)
        .filter((key) => obj[key] instanceof Object)
        .map((key) => objectDeepKeys(obj[key]).map((k) => [key, ...k]))
        .reduce((x, y) => [...x, ...y], Object.keys(obj).map((x) => [x]));
}

/**
 * @internal
 */
function isComponentPropertyBindable(viewConfigType, componentProperty) {
    const componentConfig = ViewElementRegistry.viewElements.get(viewConfigType);
    if (!componentConfig?.inputs) {
        return false;
    }
    return !!componentConfig?.inputs[componentProperty]?.propertyBindable;
}

/**
 * @internal
 */
function instantiate(type, resolveDependency) {
    const deps = getTypeDependencies(type);
    if (deps.length > 0) {
        if (!resolveDependency) {
            throw new Error(`Cannot resolve dependencies for ${type.name}. ` +
                "Please provide 'resolveDependency' bootstrap option");
        }
        const depInstances = deps.map(resolveDependency);
        return new type(...depInstances);
    }
    return new type();
}
/**
 * @public
 * @description For internal use only.
 */
function ɵinstantiate(type, resolveDependency) {
    return instantiate(type, resolveDependency);
}

/**
 * @public
 */
class CrtInputRegistrationConfig {
}

/**
 * @public
 * @class DataRequest
 * @description Definition of data source request.
 * @extends {ɵBaseRequest} {@link ɵBaseRequest}
 * @property {string} dataSourceName - name of assigned data source.
 */
class DataRequest extends _BaseRequest {
}

/**
 * @public
 * @description Request payload
 * @property {string} request - request name.
 * @property {unknown} params - request parameters.
 */

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtRequest(config) {
    return function (type) {
        checkTypeNameFormat(config.type, 'RequestTypeName');
        checkCrtTypeNamePascalCase(config.type);
        RequestRegistryService.register(config, type);
    };
}

/**
 * @hostPublicApi
 * TODO: RND-55882
 */
let HandleViewModelAttributeChangeRequest = class HandleViewModelAttributeChangeRequest extends _BaseRequest {
    constructor(payload) {
        super();
        this.type = 'crt.HandleViewModelAttributeChangeRequest';
        this.attributeName = payload.attributeName;
        this.value = payload.value;
        this.oldValue = payload.oldValue;
        // eslint-disable-next-line sonar/deprecation
        this.silent = Boolean(payload.silent);
        this.preventAttributeChangeRequest = Boolean(payload.preventAttributeChangeRequest);
        this.preventStateChange = Boolean(payload.preventStateChange);
        this.preventRunBusinessRules = Boolean(payload.preventRunBusinessRules);
    }
};
HandleViewModelAttributeChangeRequest = __decorate([
    CrtRequest({
        type: 'crt.HandleViewModelAttributeChangeRequest',
    }),
    __metadata("design:paramtypes", [Object])
], HandleViewModelAttributeChangeRequest);

/**
 * @internal
 */
var ViewElementReuseStrategy;
(function (ViewElementReuseStrategy) {
    /**
     * Reuse early created view element when it state changed
     */
    ViewElementReuseStrategy[ViewElementReuseStrategy["Reuse"] = 0] = "Reuse";
    /**
     * Rerender(recreate) view element when it state changed
     */
    ViewElementReuseStrategy[ViewElementReuseStrategy["Rerender"] = 1] = "Rerender";
})(ViewElementReuseStrategy || (ViewElementReuseStrategy = {}));

/**
 * @public
 * @description Render strategy for view element.
 */
var ViewElementRenderStrategy;
(function (ViewElementRenderStrategy) {
    /**
     * @description Default render strategy.
     * The element is rendered immediately.
     */
    ViewElementRenderStrategy["Default"] = "default";
    /**
     * @description On viewport render strategy.
     * The element is rendered only when it becomes visible in the viewport.
     */
    ViewElementRenderStrategy["OnViewport"] = "on-viewport";
})(ViewElementRenderStrategy || (ViewElementRenderStrategy = {}));

/**
 * @public
 * @description Abstract Base Validator class for designing custom validators.
 * @example
 * @CrtValidator({
 *    type: 'crt.Required',
 * })
 * export class RequiredValidator extends BaseValidator {
 *    protected override async = false;
 *    public validate(controlState: CrtControlState, params?: ValidatorParametersValues): CrtValidationErrors | null {
 *      Implementation here...
 *    }
 * }
 */
class BaseValidator {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseValidator'; }
    /**
     * @description getter for async/sync indicator
     */
    get isAsync() {
        return this.async;
    }
    /**
     * @description Function used for validation
     */
    get validateFn() {
        return this.validate.bind(this);
    }
    /**
     * @internal
     */
    static [Symbol.hasInstance](instance) {
        return instance?.constructor?.['_instanceOfKey'] === BaseValidator._instanceOfKey;
    }
}

/**
 * @public
 */
class LocalizableValue {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_LocalizableValue'; }
    constructor(key) {
        this.key = key;
    }
    /**
     * @internal
     */
    static [Symbol.hasInstance](instance) {
        return instance?.constructor?.['_instanceOfKey'] === LocalizableValue._instanceOfKey;
    }
}

/**
 * @internal
 */
function applyLocalizeMetadataFn(obj, localizeMetadata) {
    const localizableEntries = objectDeepKeys(obj)
        .map((path) => ({ path, value: get(obj, path) }))
        .filter(({ value }) => value instanceof LocalizableValue);
    if (localizableEntries.length === 0) {
        return obj;
    }
    const localized = cloneDeep(obj);
    if (!localizeMetadata) {
        console.error("Unable to localize metadata. 'localizeMetadata' bootstrap option was not provided.");
    }
    localizableEntries.forEach(({ path, value }) => {
        const localizedValue = localizeMetadata?.(value.key) ?? value.key;
        set(localized, path, localizedValue);
    });
    return localized;
}

/**
 * @internal
 */
function checkRemoteName(remoteName) {
    if (!RemoteEntriesRegistry.items.has(remoteName)) {
        throw new Error(`Remote entry with name '${remoteName}' does not exist`);
    }
}
/**
 * @internal
 */
function resistantForEach(items, callback) {
    items.forEach((item) => {
        try {
            callback(item);
        }
        catch (error) {
            console.error(error);
        }
    });
}
/**
 * @internal
 */
function registerViewElements(types) {
    resistantForEach(types, (type) => {
        const registrationConfig = getViewElementMetadata(type);
        ViewElementRegistry.validate(registrationConfig);
        ViewElementRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyViewElements(configs, lazyModuleRef) {
    resistantForEach(configs, (config) => {
        const registryItem = { ...config, lazy: true, lazyModuleRef };
        ViewElementRegistry.validate(registryItem);
        ViewElementRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerInterfaceDesignerItems(types, localizeFn) {
    resistantForEach(types, (type) => {
        if (!hasInterfaceDesignerItemMetadata(type)) {
            return;
        }
        let config = getInterfaceDesignerItemMetadata(type);
        config = applyLocalizeMetadataFn(config, localizeFn);
        const viewElementMetadata = getViewElementMetadata(type);
        config.type = viewElementMetadata.type;
        InterfaceDesignerItemRegistry.validate(config);
        InterfaceDesignerItemRegistry.register(config);
    });
}
/**
 * @internal
 */
function registerDesignTimeViewElements(types) {
    resistantForEach(types, (type) => {
        const registrationConfig = getDesignTimeViewElementMetadata(type);
        DesignTimeViewElementRegistry.validate(registrationConfig);
        DesignTimeViewElementRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerValidators(types, options) {
    resistantForEach(types, (type) => {
        const metadata = getValidatorMetadata(type);
        const registrationConfig = {
            ...metadata,
            instantiate: function (injectionContext) {
                const resolveDependency = options?.resolveDependency
                    ? (token) => options.resolveDependency(token, injectionContext)
                    : null;
                return instantiate(type, resolveDependency);
            },
        };
        ValidatorRegistry.validate(registrationConfig);
        ValidatorRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyValidators(configs, lazyModuleRef) {
    resistantForEach(configs, (config) => {
        const registryItem = { ...config, lazy: true, lazyModuleRef };
        ValidatorRegistry.validate(registryItem);
        ValidatorRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerConverters(types, options) {
    resistantForEach(types, (type) => {
        const metadata = getConverterMetadata(type);
        const registrationConfig = {
            ...metadata,
            instantiate: function (injectionContext) {
                const resolveDependency = options?.resolveDependency
                    ? (token) => options.resolveDependency(token, injectionContext)
                    : null;
                return instantiate(type, resolveDependency);
            },
        };
        ConverterRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function registerLazyConverters(descriptors, lazyModuleRef) {
    resistantForEach(descriptors, (descriptor) => {
        const registryItem = { ...descriptor, lazy: true, lazyModuleRef };
        ConverterRegistry.register(registryItem);
    });
}
/**
 * @internal
 */
function registerRequestHandlers(types, options, remoteName) {
    if (options.moduleType === 'remote' && !remoteName) {
        console.error("To use request handlers, 'remoteName' parameter should be specified.");
        return;
    }
    resistantForEach(types, (type) => {
        const metadata = getRequestHandlerMetadata(type);
        let registrationConfig;
        const createHandler = (injectionContext) => {
            const resolveDependency = options?.resolveDependency
                ? (token) => options.resolveDependency(token, injectionContext)
                : null;
            return instantiate(type, resolveDependency);
        };
        if (options.moduleType === 'host') {
            registrationConfig = {
                ...metadata,
                source: { type: HandlerSourceType.Host },
                createHandler,
            };
        }
        else {
            const remote = RemoteEntriesRegistry.items.get(remoteName);
            registrationConfig = {
                ...metadata,
                source: {
                    type: HandlerSourceType.Remote,
                    packageName: remote.packageName,
                },
                createHandler,
            };
        }
        RequestHandlerRegistryService.register(registrationConfig, type);
    });
}
/**
 * @internal
 */
function registerLazyRequestHandlers(descriptors, options, remoteName, lazyModuleRef) {
    resistantForEach(descriptors, (descriptor) => {
        const registrationConfig = {
            ...descriptor,
            lazy: true,
            lazyModuleRef,
        };
        if (options.moduleType === 'host') {
            registrationConfig.source = { type: HandlerSourceType.Host };
        }
        else {
            const remote = RemoteEntriesRegistry.items.get(remoteName);
            registrationConfig.source = {
                type: HandlerSourceType.Remote,
                packageName: remote.packageName,
            };
        }
        RequestHandlerRegistryService.register(registrationConfig);
    });
}
/**
 * @internal
 */
function bootstrapLazyCrtModule(name, descriptor, options, lazyModuleRef) {
    try {
        if (descriptor.viewElements) {
            registerLazyViewElements(descriptor.viewElements, lazyModuleRef);
        }
        if (descriptor.requestHandlers) {
            registerLazyRequestHandlers(descriptor.requestHandlers, options, name, lazyModuleRef);
        }
        if (descriptor.validators) {
            registerLazyValidators(descriptor.validators, lazyModuleRef);
        }
        if (descriptor.converters) {
            registerLazyConverters(descriptor.converters, lazyModuleRef);
        }
        descriptor.includes?.forEach((m) => {
            bootstrapLazyCrtModule(name, m, options, lazyModuleRef);
        });
    }
    catch (error) {
        console.error(error);
    }
}
/**
 * @internal
 */
function bootstrapDefaultCrtModule(name, type, options, root) {
    if (bootstrappedCrtModules.has(type)) {
        return;
    }
    const source = options.moduleType === 'host' ? 'host' : name;
    bootstrappedCrtModules.set(type, { source, root, options });
    try {
        const moduleDefinition = getModuleMetadata(type);
        if (moduleDefinition.viewElements) {
            registerViewElements(moduleDefinition.viewElements);
            registerInterfaceDesignerItems(moduleDefinition.viewElements, options?.localizeMetadata);
        }
        if (moduleDefinition.designTimeViewElements) {
            registerDesignTimeViewElements(moduleDefinition.designTimeViewElements);
        }
        if (moduleDefinition.requestHandlers) {
            registerRequestHandlers(moduleDefinition.requestHandlers, options, name);
        }
        if (moduleDefinition.validators) {
            registerValidators(moduleDefinition.validators, options);
        }
        if (moduleDefinition.converters) {
            registerConverters(moduleDefinition.converters, options);
        }
        moduleDefinition.includes?.forEach((includedModuleMetadata) => {
            if (typeof includedModuleMetadata === 'function') {
                bootstrapDefaultCrtModule(name, includedModuleMetadata, options, false);
                bootstrapPredefinedModules(name, includedModuleMetadata, options);
            }
            else if (typeof includedModuleMetadata === 'object') {
                if (options.moduleType === 'remote' && !name) {
                    console.error("To use lazy loading, 'remoteName' parameter should be specified.");
                    return;
                }
                const ref = new LazyModuleRef(includedModuleMetadata, (x) => bootstrapDefaultCrtModule(name, x, options, false), source);
                bootstrapLazyCrtModule(name, includedModuleMetadata, options, ref);
                LazyModulesManager.instance.add(ref);
            }
        });
    }
    catch (error) {
        console.error(error);
    }
}
/**
 * @internal
 */
function bootstrapPredefinedModules(name, type, options) {
    const modulePrefix = 'module:';
    const metadataKeys = Reflect.getOwnMetadataKeys(type);
    for (const key of metadataKeys) {
        if (typeof key === 'string' && key.startsWith(modulePrefix)) {
            const moduleType = key.substring(modulePrefix.length);
            const predefinedBootstrapper = Reflect.getOwnMetadata('moduleBootstrapper:' + moduleType, type);
            try {
                predefinedBootstrapper(name, type, options);
            }
            catch (error) {
                console.error(error);
            }
        }
    }
}
/**
 * @internal
 */
function bootstrapCrtModule(...args) {
    const name = typeof args[0] === 'string' ? args.shift() : undefined;
    const type = args[0];
    const options = args[1] ?? {};
    options.moduleType ??= 'remote';
    if (name) {
        if (options.moduleType === 'remote') {
            checkRemoteName(name);
        }
        else {
            throw new Error('Host module cannot have name');
        }
    }
    bootstrapDefaultCrtModule(name, type, options, true);
    bootstrapPredefinedModules(name, type, options);
    if (options.moduleType === 'remote') {
        BootstrapNotifier.instance.notify(type);
    }
}

/**
 * @public
 * @description Use this function in decorators to indicate that a value should be localized.
 * Note that the "localizeMetadata" bootstrap option should be provided,
 * otherwise the value cannot be localized.
 *
 * @example
 *
 * ### Using 'localize' in decorator
 * ```ts
 * @CrtInterfaceDesignerItem({
 *    toolbarConfig: {
 *      caption: localize('MyElement.Caption')
 * 	  }
 * })
 * class MyElement {}
 * ```
 * ### Passing 'localizeMetadata' bootstrap option
 * ```ts
 * bootstrapCrtModule(MyModule, {
 *    localizeMetadata: (key: string) => thirdPartyTranslateService.translate(key)
 * });
 * ```
 */
function localize(key) {
    return new LocalizableValue(key);
}

/**
 * @internal
 */
function registerHandlerChainServices(types) {
    types?.forEach((type) => {
        const registrationConfig = getHandlerChainServiceMetadata(type);
        HandlerChainServicesRegistry.register(registrationConfig);
    });
}
/**
 * @internal
 */
function _bootstrapCrtModuleHandlerChainServices(type) {
    const moduleDefinition = getModuleMetadata(type);
    if (moduleDefinition.handlerChainServices) {
        registerHandlerChainServices(moduleDefinition.handlerChainServices);
    }
    moduleDefinition.includes?.forEach((t) => {
        if (typeof t === 'function') {
            _bootstrapCrtModuleHandlerChainServices(t);
        }
    });
}
/**
 * @internal
 */
function bootstrapCrtModuleHandlerChainServices(type) {
    _bootstrapCrtModuleHandlerChainServices(type);
}

/**
 * @internal
 * Clears global registries, resets state of global services.
 * Can be used in unit tests to clear global registries after bootstrapping module.
 * @example
 * beforeEach(() => {
 *   bootstrapCrtModule('test', TestModule);
 * });
 *
 * afterEach(() => {
 *   resetGlobals();
 * });
 */
function resetGlobals() {
    bootstrappedCrtModules.clear();
    ViewElementRegistry.viewElements.clear();
    InternalRequestHandlerRegistry.clear();
    HandlerChainService.instance.clear();
    ValidatorRegistry.items.clear();
    ConverterRegistry.clear();
    RemoteEntriesRegistry.items.clear();
    PkgTopology.items.clear();
    LazyModulesManager.instance.clear();
    LazyModulesManager.instance.lazyLoadingDisabled = false;
    DesignTimeViewElementRegistry.infos.clear();
}

/**
 * @internal
 * @description Defines render configuration for view element.
 */
function defineRenderConfig(config) {
    if (isEmpty(config.renderStrategy) && 'placeholderSize' in config) {
        config.renderStrategy = ViewElementRenderStrategy.OnViewport;
    }
    if (config.renderStrategy === ViewElementRenderStrategy.OnViewport && !config.placeholderSize) {
        config.placeholderSize = { height: VIEW_ELEMENT_DEFAULT_HEIGHT };
    }
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtDesignTimeViewElement(config) {
    return function (type) {
        const inputs = getCrtInputsMetadata(type);
        const registrationConfig = {
            ...config,
            typeRef: type,
            inputs,
        };
        defineRenderConfig(registrationConfig);
        setDesignTimeViewElementMetadata(registrationConfig, type);
    };
}

/**
 * @internal
 * @hostPublicApi
 */
class InterfaceDesignerItemRegistrationConfig {
}

/**
 * @internal
 * @class MethodInvokeRequest
 * @description Definition of invoke method request.
 * @extends {ɵBaseRequest} {@link ɵBaseRequest}
 * @property {string} method - method name.
 * @property {unknown[]} args - arguments passing to the method.
 */
class MethodInvokeRequest extends _BaseRequest {
}

/**
 * @internal
 */
var RemoteEntryType;
(function (RemoteEntryType) {
    RemoteEntryType["Script"] = "script";
    RemoteEntryType["Module"] = "module";
})(RemoteEntryType || (RemoteEntryType = {}));

/**
 * @public
 * @description Creatio platform interacts with input's of user-defined component through CrtInput decorator.
 * @param config {@link CrtInputRegistrationConfig}
 *
 * @example
 * export class AppComponent {
 *    @CrtInput()
 *    public value!: string;
 *  }
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtInput(config) {
    return function (target, propertyKey) {
        if (!config) {
            config = {
                propertyBindable: false,
            };
        }
        Reflect.defineMetadata(crtInputPrefix + propertyKey.toString(), config, target.constructor);
    };
}

/**
 * @public
 * @description User-defined components could notify Creatio platform about changes through CrtOutput decorator.
 * @example
 * export class AppComponent {
 *     @Input()
 *     @CrtOutput()
 *     public valueChange;
 *  }
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtOutput() {
    return function (target, propertyKey) {
        Reflect.defineMetadata(crtOutputPrefix + propertyKey.toString(), {}, target.constructor);
    };
}

/**
 * @public
 * @description User-defined components receive validationInfo changes through CrtValidationInput decorator.
 * Make sure you have defined input for every validationInfo input.
 * To properly declare validationInfo field:
 * 1. Declare input field inside inputs section. For example `value`
 * 2. In validationInputs section declare same field which ends with ValidationInfo.
 * For previous example `valueValidationInfo`
 *
 * @example
 * export class AppComponent {
 *      @CrtInput()
        public value;
 *
 *     	@CrtValidationInput()
 *      public valueValidationInfo;
 *  }
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtValidationInput() {
    return function (target, propertyKey) {
        Reflect.defineMetadata(crtValidationInputPrefix + propertyKey.toString(), {}, target.constructor);
    };
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtViewElement(config) {
    return function (type) {
        checkTypeNamePascalCase(config.type);
        const inputs = getCrtInputsMetadata(type);
        const outputs = getCrtOutputsMetadata(type);
        const validationInputs = getCrtValidationInputsMetadata(type);
        const baseRegistrationConfig = {
            ...config,
            inputs,
            outputs,
            validationInputs,
        };
        const registrationConfig = 'selector' in config
            ? { ...baseRegistrationConfig, selector: config.selector }
            : { ...baseRegistrationConfig, typeRef: type };
        registrationConfig.reuseStrategy = config?.reuseStrategy ?? ViewElementReuseStrategy.Rerender;
        defineRenderConfig(registrationConfig);
        setViewElementMetadata(registrationConfig, type);
    };
}

/**
 * @internal
 */
function CrtModule(moduleDefinition) {
    return (type) => setModuleMetadata(moduleDefinition, type);
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtInterfaceDesignerItem(config) {
    function _applyTypeCaption() {
        if (!config.typeCaption && config.toolbarConfig) {
            const toolbarConfig = Array.isArray(config.toolbarConfig)
                ? config.toolbarConfig[0]
                : config.toolbarConfig;
            config.typeCaption = toolbarConfig.caption;
        }
    }
    return function (type) {
        config.typeRef = type;
        config.viewElementGroupType ??= ViewElementGroupType.CustomElements;
        _applyTypeCaption();
        defineRenderConfig(config);
        setInterfaceDesignerItemMetadata(config, type);
    };
}

/**
 * @public
 * @description
 * Constructor parameter decorator.
 * Mark parameter as dependency that should be obtained by token.
 * @param token - token by which the dependency instance will be obtained
 * @example
 * class MyHandler {
 * 	constructor(@CrtInject(USER_NAME_TOKEN) userName: string) {}
 * }
 */
function CrtInject(token) {
    return function (target, paramKey, paramIndex) {
        Reflect.defineMetadata(crtInjectPrefix + paramIndex, token, target);
    };
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtValidator(config) {
    if (!config) {
        return (type) => {
            // the decorator won't do anything by itself,
            // but will be used by the compiler to generate class metadata,
            // which is required for DI
        };
    }
    return function (type) {
        checkTypeNameFormat(config.type, 'ValidatorTypeName');
        const registrationConfig = {
            ...config,
            typeRef: type,
        };
        setValidatorMetadata(registrationConfig, type);
    };
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtRequestHandler(config) {
    if (!config) {
        return (type) => {
            // the decorator won't do anything by itself,
            // but will be used by the compiler to generate class metadata,
            // which is required for DI
        };
    }
    return function (type) {
        checkTypeNameFormat(config.type, 'RequestHandlerTypeName');
        checkCrtTypeNamePascalCase(config.type);
        setRequestHandlerMetadata(config, type);
    };
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtConverter(config) {
    if (!config) {
        return (type) => { };
    }
    return function (typeRef) {
        const type = typeof config === 'string' ? config : config.type;
        checkTypeNameFormat(type, 'converterTypeName');
        const registrationConfig = {
            type: type,
            typeRef: typeRef,
        };
        setConverterMetadata(registrationConfig, typeRef);
    };
}

/**
 * @public
 * @description CrtChatMessageType decorator allows user to register components as a message type in chat.
 * @param config - registration config. {@link ChatMessageTypeRegistrationConfig}.
 *
 * @example
 * @CrtChatMessageType({
 *      selector: 'usr-video-message',
 *      messageType: 'video',
 *  })
 *  export class AppComponent { }
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function CrtChatMessageType(config) {
    return function (type) {
        ChatMessageTypeRegistry.register(config.messageType, config.selector);
    };
}

/**
 * @internal
 */
var DataSourceLoadType;
(function (DataSourceLoadType) {
    DataSourceLoadType["LoadNext"] = "loadNext";
    DataSourceLoadType["Load"] = "load";
    DataSourceLoadType["Reload"] = "reload";
})(DataSourceLoadType || (DataSourceLoadType = {}));

/**
 * @internal
 */
var ViewElementProperty;
(function (ViewElementProperty) {
    ViewElementProperty["Readonly"] = "READONLY_PROPERTY";
    ViewElementProperty["Disabled"] = "DISABLED_PROPERTY";
})(ViewElementProperty || (ViewElementProperty = {}));

/**
 * @public
 * @abstract
 * @classdesc Class for designing request handlers.
 */
class BaseRequestHandler {
    constructor() {
        this.handlerChain = HandlerChainService.instance;
    }
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseRequestHandler'; }
    get next() {
        return this._next;
    }
    /**
     * @internal
     */
    static [Symbol.hasInstance](instance) {
        return instance?.constructor?.['_instanceOfKey'] === BaseRequestHandler._instanceOfKey;
    }
    /**
     * @description Method to set request handler to handler chain.
     * @param next - link for next request handler - {@link ɵRequestHandler}
     */
    setNext(next) {
        if (this._next) {
            throw new _NextHandlerAlreadySpecifiedException();
        }
        this._next = next;
    }
}

class BaseHandlerChainService {
}

/**
 * @public
 * @classdesc Service to working with Creatio platform features using handler chain
 * @extends {BaseHandlerChainService}
 */
class FeatureService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.FeatureService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], FeatureService.prototype, "getFeatureState", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], FeatureService.prototype, "getFeaturesState", void 0);

/**
 * @public
 * @classdesc Service to working rights using handler chain
 * @extends {BaseHandlerChainService}
 */
class RightsService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.RightsService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], RightsService.prototype, "getCanExecuteOperation", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], RightsService.prototype, "getCanExecuteOperations", void 0);

/**
 * @public
 * @class SysSettingsService
 * @classdesc Service to working with system settings using handler chain
 * @extends {BaseHandlerChainService}
 */
// TODO RND-45286 (implement BaseSysSettingsService)
class SysSettingsService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.SysSettingsService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], SysSettingsService.prototype, "getByCodes", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], SysSettingsService.prototype, "getByCode", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], SysSettingsService.prototype, "update", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], SysSettingsService.prototype, "updateMany", void 0);

/**
 * @public
 * @class SysSetting
 * @classdesc Class contains info related system setting
 * @property {ɵGuid} id - id of system setting.
 * @property {string} name - name of system setting.
 * @property {string} code - code of system setting.
 * @property {isCacheable} boolean - indicator whether system setting is cacheable
 * @property {any} value - value of system setting
 * @property {string} displayValue - display value of system setting
 * @property {ɵDataValueType} ɵDataValueType - system setting {@link ɵDataValueType}
 */
class SysSetting {
}
/**
 * @public
 * @class SysSettingValue
 * @classdesc Shape to describe system setting value
 * @property {string} code - code of system setting
 * @property {any} value - value of system setting
 */
class SysSettingValue {
}

/**
 * @public
 * @classdesc
 * @class SysValuesService  Service to working with system values using handler chain
 * @extends {BaseHandlerChainService}
 */
class SysValuesService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.SysValuesService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], SysValuesService.prototype, "loadSysValues", void 0);

/**
 * @public
 * @classdesc Http service to perform Http requests
 * @extends {BaseHandlerChainService}
 */
class HttpClientService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.HttpClientService';
    }
    /**
     * @description HTTP GET request.
     * @return `Promise<HttpResponse> | void`  {(Promise<HttpResponse> | void)}
     */
    get() {
        // the method will be implemented dynamically by the decorator
    }
    /**
     * @description HTTP POST request.
     * @return `Promise<HttpResponse> | void`  {(Promise<HttpResponse> | void)}
     */
    post() {
        // the method will be implemented dynamically by the decorator
    }
    /**
     * @description HTTP DELETE request.
     * @return `Promise<HttpResponse> | void`  {(Promise<HttpResponse> | void)}
     */
    delete() {
        // the method will be implemented dynamically by the decorator
    }
    /**
     * @description HTTP PUT request.
     * @return `Promise<HttpResponse> | void``  {(Promise<HttpResponse> | void)}
     */
    put() {
        // the method will be implemented dynamically by the decorator
    }
    /**
     * @description HTTP PATCH request.
     * @return `Promise<HttpResponse> | void`  {(Promise<HttpResponse> | void)}
     */
    patch() {
        // the method will be implemented dynamically by the decorator
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], HttpClientService.prototype, "get", null);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], HttpClientService.prototype, "post", null);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], HttpClientService.prototype, "delete", null);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], HttpClientService.prototype, "put", null);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], HttpClientService.prototype, "patch", null);

/**
 * @public
 * @classdesc Represents a service to working with license restrictions.
 * @extends {BaseHandlerChainService}
 */
class LicenseService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.LicenseService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], LicenseService.prototype, "getLicenseOperationStatuses", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], LicenseService.prototype, "getExplicitLicenseOperationStatuses", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], LicenseService.prototype, "loadLicenseOperationStatuses", void 0);

/**
 * @public
 * @classdesc Service to working with mask using handler chain
 * @extends {BaseHandlerChainService}
 */
class MaskService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.MaskService';
    }
    /**
     * @internal
     * @param taskName
     * @param context
     */
    async showMask(taskName, context) {
        const request = new _SchemaViewMaskRequest(_SchemaViewMaskRequestAction.AddTask, taskName, context);
        await HandlerChainService.instance.process(request);
    }
    /**
     * @internal
     * @param taskName
     * @param context
     */
    async hideMask(taskName, context) {
        const request = new _SchemaViewMaskRequest(_SchemaViewMaskRequestAction.RemoveTask, taskName, context);
        await HandlerChainService.instance.process(request);
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], MaskService.prototype, "showBodyMask", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], MaskService.prototype, "hideBodyMask", void 0);

/**
 * @internal
 */
const _from = Symbol('CreateModelFrom');
/**
 * @public
 */
class Model extends BaseHandlerChainService {
    constructor(crtModel) {
        super();
        /**
         * @internal
         */
        this.requestType = 'crt.ModelGenerator';
        this._crtModel = crtModel;
    }
    static get instance() {
        if (!this._instance) {
            this._instance = new Model();
        }
        return this._instance;
    }
    /**
     * @internal
     * Gets the model name.
     */
    get name() {
        return this._crtModel?.name;
    }
    /**
     * @internal
     */
    static [_from](crtModel) {
        return new Model(crtModel);
    }
    static async create(dataSchema) {
        const crtModel = await this.instance._create({
            type: 'crt.EntityDataSource',
            config: {
                entitySchemaName: dataSchema,
            },
        }, '');
        return new Model(crtModel);
    }
    getSchema() {
        return this._crtModel._getAsyncResult(this._crtModel.getSchema());
    }
    load(config) {
        if (config?.loadOptions && !config?.options) {
            config.options = config.loadOptions;
        }
        return this._crtModel._getAsyncResult(this._crtModel.load(config));
    }
    insert(dto) {
        return this._crtModel._getAsyncResult(this._crtModel.insert(dto));
    }
    update(dto, parameters) {
        return this._crtModel._getAsyncResult(this._crtModel.update(dto, parameters));
    }
    copy(primaryColumnValue, data = {}) {
        return this._crtModel._getAsyncResult(this._crtModel.copy(primaryColumnValue, data));
    }
    create(data = {}) {
        return this._crtModel._getAsyncResult(this._crtModel.create(data));
    }
    delete(parameters) {
        return this._crtModel._getAsyncResult(this._crtModel.delete(parameters));
    }
    canSave(params) {
        return this._crtModel._getAsyncResult(this._crtModel.canSave(params));
    }
    canDelete(params) {
        return this._crtModel._getAsyncResult(this._crtModel.canDelete(params));
    }
}
__decorate([
    ImplementUsingHandlerChain({
        methodName: 'generateOne',
    }),
    __metadata("design:type", Function)
], Model.prototype, "_create", void 0);

/**
 * @public
 * @classdesc Service to handling a way of tasks execution in Creatio plaform
 * @extends {BaseHandlerChainService}
 */
class CrtZoneService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.ZoneService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], CrtZoneService.prototype, "runOutside", void 0);

/**
 * @public
 * @classdesc Service to working with dialog using handler chain.
 * @extends {BaseHandlerChainService}
 */
class DialogService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.DialogService';
    }
}
__decorate([
    ImplementUsingHandlerChain({
        methodName: 'openDialog',
    }),
    __metadata("design:type", Function)
], DialogService.prototype, "open", void 0);

/**
 * @public
 * @classdesc Service for working with messages sent or received through a server channel.
 * @extends {BaseHandlerChainService}
 * @implements {BaseMessageChannelService}
 */
class MessageChannelService extends BaseHandlerChainService {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.requestType = 'crt.MessageChannelService';
    }
}
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], MessageChannelService.prototype, "subscribe", void 0);
__decorate([
    ImplementUsingHandlerChain(),
    __metadata("design:type", Function)
], MessageChannelService.prototype, "sendMessage", void 0);

/**
 * Generated bundle index. Do not edit.
 */

export { BaseRequestHandler, BaseValidator, BootstrapNotifier, ChatMessageTypeRegistry, ConverterRegistry, CrtChatMessageType, CrtConverter, CrtDesignTimeViewElement, CrtInject, CrtInput, CrtInputRegistrationConfig, CrtInterfaceDesignerItem, CrtModule, CrtOutput, CrtRequest, CrtRequestHandler, CrtValidationInput, CrtValidator, CrtViewElement, CrtZoneService, DataRequest, DataSourceLoadType, DesignTimeViewElementRegistry, DesignerToolGroupType, DialogService, FeatureService, HandleViewModelAttributeChangeRequest, HandlerChainService, HandlerChainServicesRegistry, HandlerSourceType, HandlersOrderComparer, HttpClientService, ImplementUsingHandlerChain, InterfaceDesignerItemRegistrationConfig, InterfaceDesignerItemRegistry, InternalRequestHandlerRegistry, InternalRequestHandlerRegistryType, LazyModuleRef, LazyModulesManager, LicenseService, LocalizableValue, MaskService, MessageChannelService, MethodInvokeRequest, Model, PkgTopology, RemoteEntriesRegistry, RemoteEntryType, RequestHandlerRegistryService, RequestRegistryService, RightsService, SysSetting, SysSettingValue, SysSettingsService, SysValuesService, VIEW_ELEMENT_DEFAULT_HEIGHT, ValidatorMapValue, ValidatorRegistry, ViewElementDataGroupType, ViewElementGroupType, ViewElementProperty, ViewElementRegistry, ViewElementRenderStrategy, ViewElementReuseStrategy, _from, applyLocalizeMetadataFn, bootstrapCrtModule, bootstrapCrtModuleHandlerChainServices, bootstrappedCrtModules, callOnContextCreateHooks, callOnContextDestroyHooks, callOnContextUpdateHooks, checkCrtTypeNamePascalCase, checkIfTypeRegistered, checkOnlyOneRegistration, checkTypeNameFormat, checkTypeNamePascalCase, crtInjectPrefix, crtInputPrefix, crtOutputPrefix, crtValidationInputPrefix, defineRenderConfig, designTimeViewElementRegistryKey, getConverterMetadata, getCrtInputsMetadata, getCrtOutputsMetadata, getCrtValidationInputsMetadata, getDeepModuleDependencies, getDesignTimeViewElementMetadata, getHandlerChainServiceMetadata, getInterfaceDesignerItemMetadata, getModuleMetadata, getRequestHandlerMetadata, getTypeDependencies, getValidatorMetadata, getViewElementMetadata, hasInterfaceDesignerItemMetadata, instantiate, interfaceDesignerItemRegistryKey, isComponentPropertyBindable, localize, matchScopes, objectDeepKeys, registerAngularViewElement, registerViewElement, registryKey, resetGlobals, setConverterMetadata, setDesignTimeViewElementMetadata, setInterfaceDesignerItemMetadata, setModuleMetadata, setRequestHandlerMetadata, setValidatorMetadata, setViewElementMetadata, toAggregationEval, toAggregationEvalType, toAggregationFunction, toAggregationType, toDatePart, toDatePartType, toNumericFunctionType, toOrderDirection, toPascalCase, toSortDirection, toStringFunctionType, validatorRegistryKey, viewElementRegistryKey, ɵBaseRegistry, ɵcheckTypeNameFormat, ɵinstantiate };
