// @creatio-devkit/common/esm2022/lib/internal/services/request-handler-registry/request-handler-registry.service.mjs
/**
 * @internal
 * @hostPublicApi available through window.creatio.requestHandlerRegistry
 */
export class InternalRequestHandlerRegistryType {
    static { this._addItemListeners = []; }
    static { this._removeItemListeners = []; }
    static { this._requestHandlersMap = new Map(); }
    /**
     * @publicHostApi
     */
    static get requestHandlersMap() {
        return this._requestHandlersMap;
    }
    /**
     * @publicHostApi
     */
    static onAddRequestHandler(handler) {
        this._addItemListeners.push(handler);
        return () => {
            this._addItemListeners.splice(this._addItemListeners.indexOf(handler), 1);
        };
    }
    /**
     * @publicHostApi
     */
    static onRemoveRequestHandler(handler) {
        this._removeItemListeners.push(handler);
        return () => {
            this._removeItemListeners.splice(this._removeItemListeners.indexOf(handler), 1);
        };
    }
    /**
     * @publicHostApi
     */
    static addRequestHandler(handlerType, config) {
        const existingHandler = this._requestHandlersMap.get(handlerType);
        if (existingHandler && (!existingHandler.instanceConfig?.lazy || existingHandler.instanceType)) {
            return;
        }
        const mergedConfig = {
            instanceType: config.instanceType,
            instanceConfig: {
                ...(existingHandler?.instanceConfig ?? {}),
                ...config.instanceConfig,
            },
        };
        this._requestHandlersMap.set(handlerType, mergedConfig);
        for (const listener of this._addItemListeners) {
            listener(handlerType, mergedConfig);
        }
    }
    /**
     * @publicHostApi
     */
    static removeRequestHandler(handlerType) {
        const config = this._requestHandlersMap.get(handlerType);
        if (!config) {
            return;
        }
        this._requestHandlersMap.delete(handlerType);
        for (const listener of this._removeItemListeners) {
            listener(handlerType, config);
        }
    }
    static clear() {
        for (const type of this._requestHandlersMap.keys()) {
            this.removeRequestHandler(type);
        }
    }
}
const global = window;
global.creatio ??= {};
if (!global.creatio.requestHandlerRegistry) {
    Object.defineProperty(global.creatio, 'requestHandlerRegistry', {
        get: () => InternalRequestHandlerRegistryType,
        configurable: false,
    });
}
/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export const InternalRequestHandlerRegistry = global.creatio.requestHandlerRegistry;
/**
 * @internal
 */
export class RequestHandlerRegistryService {
    static { this._internalRegistry = global.creatio.requestHandlerRegistry; }
    /**
     * Register request handler class.
     * @param config Request handler registration config.
     * @param typeRef Request handler type reference (constructor).
     */
    static register(config, typeRef) {
        const requestType = config.requestType;
        const requestHandlerType = config.type;
        if (!requestType) {
            throw new Error(`Error when register ${requestHandlerType}. Request type not specified.`);
        }
        const requestHandlerMapValue = {
            instanceType: typeRef,
            instanceConfig: config,
        };
        RequestHandlerRegistryService._internalRegistry.addRequestHandler(requestHandlerType, requestHandlerMapValue);
    }
}
