// @creatio-devkit/common/esm2022/lib/internal/services/validator-registry/validator-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkTypeNameFormat } from '../element-registration-checker';
/**
 * @internal
 */
export class ValidatorMapValue {
}
/**
 * @internal
 */
export const validatorRegistryKey = 'ValidatorRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
export class ValidatorRegistry extends ɵBaseRegistry {
    static { this.registryKey = validatorRegistryKey; }
    static _hasWithAnotherTypeRef(configType, typeRef) {
        const existingItem = ValidatorRegistry.items.get(configType);
        if (!existingItem) {
            return false;
        }
        if (existingItem.lazy && !existingItem.typeRef) {
            return false;
        }
        return existingItem.typeRef !== typeRef;
    }
    static validate(config) {
        checkTypeNameFormat(config.type, 'ValidatorTypeName');
        if (config.typeRef && ValidatorRegistry._hasWithAnotherTypeRef(config.type, config.typeRef)) {
            throw new Error(`Error when register ${config.type}. Element with this type has already exist`);
        }
    }
    static get items() {
        return this.store.typeInfos;
    }
    static register(config) {
        const existingConfig = this.items.get(config.type);
        this.items.set(config.type, {
            ...(existingConfig ?? {}),
            ...config,
        });
    }
}
