// @creatio/base/esm2022/lib/public/esq/data-model/expressions/columns/column-expression.mjs
import { ɵExpressionType } from '../../enums';
import { ɵBaseExpression } from '../base-expression';
/**
 * @public
 * @description A class that defines an expression of type column EntitySchema.
 * @extends {ɵBaseExpression}
 */
export class ɵColumnExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ColumnExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Type of expression.
         */
        this.expressionType = ɵExpressionType.SchemaColumn;
        this.columnPath = config?.columnPath;
    }
    /**
     * Parses expression from json.
     * @public
     * @param dto - Serialized object.
     */
    static fromJson(dto) {
        const expression = new ɵColumnExpression({ columnPath: dto['columnPath'] });
        return expression;
    }
    /**
     * @description Clones instance.
     */
    clone() {
        return new ɵColumnExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            {
                path: this.columnPath,
            },
        ];
    }
}
