// @creatio-devkit/common/esm2022/lib/internal/services/view-element-registry/view-element-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkOnlyOneRegistration, checkTypeNameFormat } from '../element-registration-checker';
/**
 * @internal
 */
export const viewElementRegistryKey = 'ViewElementRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
export class ViewElementRegistry extends ɵBaseRegistry {
    static { this.registryKey = viewElementRegistryKey; }
    static validate(config) {
        checkTypeNameFormat(config.type, 'ViewElementTypeName');
        const existingItem = ViewElementRegistry.viewElements.get(config.type);
        if (existingItem?.lazy && !('selector' in existingItem) && !existingItem.typeRef) {
            return;
        }
        checkOnlyOneRegistration(config.type, config, ViewElementRegistry.viewElements);
    }
    static get viewElements() {
        return this.store.typeInfos;
    }
    static register(config) {
        const existingConfig = this.viewElements.get(config.type);
        this.viewElements.set(config.type, {
            ...(existingConfig ?? {}),
            ...config,
        });
    }
    static getFormControlRelatesAttribute(type) {
        const config = this.store.typeInfos.get(type);
        return (config && 'formControlConfig' in config && config.formControlConfig?.relatesTo) || null;
    }
    static getViewElementInputs(type) {
        const config = this.store.typeInfos.get(type);
        return config?.inputs ?? {};
    }
    static getViewElementOutputs(type) {
        const config = this.store.typeInfos.get(type);
        return config?.outputs ?? {};
    }
    static getViewElementPropertiesDefaultValues(type) {
        const config = this.store.typeInfos.get(type);
        const propertiesDefaultValues = Object.entries(config?.inputs ?? {})
            .filter(([_, inputConfig]) => Object.prototype.hasOwnProperty.call(inputConfig, 'defaultValue'))
            .map(([propertyName, propertyConfig]) => ({
            name: propertyName,
            defaultValue: propertyConfig.defaultValue,
        }));
        return [
            ...propertiesDefaultValues,
            {
                name: 'visible',
                defaultValue: true,
            },
        ];
    }
    static getViewElementCompatiblePlatformAPIsOptions(type, platformAPI) {
        const config = this.store.typeInfos.get(type);
        return (config?.compatibleAPIs?.[platformAPI] ?? {});
    }
    static isViewElementAvailableByCompatibleAPIs(type, ...availableAPIs) {
        const config = this.store.typeInfos.get(type);
        return Object.entries(config?.compatibleAPIs ?? {})
            .filter(([key, _]) => availableAPIs.includes(key))
            .some(([_, value]) => {
            if (typeof value === 'boolean') {
                return value;
            }
            return (value || {}).enable;
        });
    }
}
