// @creatio/base/esm2022/lib/public/esq/data-model/expressions/base-expression.mjs
import { classToPlain } from 'class-transformer';
/**
 * @public
 * @abstract
 * @description The base class of the expression.
 * @extends {ɵSerializable}
 */
export class ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [this._instanceOfKey];
    }
    /**
     * @public
     * @description Method for checking if an instance is of the current class type.
     * @param instance - The instance to check.
     * @returns `true` if the instance is of the current class type, `false` otherwise.
     */
    static [Symbol.hasInstance](instance) {
        const instanceOfKeys = instance?.constructor?.['_instanceOfKeys'] ?? [];
        return instanceOfKeys.includes(this._instanceOfKey);
    }
    constructor(config) {
        /**
         * @description True if expression has brackets.
         */
        this.isBlock = false;
        this.isBlock = config?.isBlock;
    }
    /**
     * @public
     * @description Parses expression from json.
     */
    static fromJson(dto, filterParser, expressionParser) {
        throw new Error(`fromJson not supported on expression type ${this.constructor.name}`);
    }
    /**
     * @description Converts instance to json object.
     * @param item - The instance to serialize.
     */
    serializeItem(item) {
        return typeof item.toJson === 'function'
            ? item.toJson()
            : classToPlain(item, {
                exposeUnsetFields: false,
                excludePrefixes: ['_'],
            });
    }
    /**
     * @public
     * @description Converts instance to json object.
     */
    toJson() {
        return classToPlain(this, {
            exposeUnsetFields: false,
            excludePrefixes: ['_'],
        });
    }
}
