// @creatio-devkit/common/esm2022/lib/public/models/request/handle-view-model-attribute-change.request.mjs
import { __decorate, __metadata } from "tslib";
import { BaseRequest } from '../../../base-exports';
import { CrtRequest } from '../../../internal/decorators/request';
/**
 * @hostPublicApi
 * TODO: RND-55882
 */
let HandleViewModelAttributeChangeRequest = class HandleViewModelAttributeChangeRequest extends BaseRequest {
    constructor(payload) {
        super();
        this.type = 'crt.HandleViewModelAttributeChangeRequest';
        this.attributeName = payload.attributeName;
        this.value = payload.value;
        this.oldValue = payload.oldValue;
        // eslint-disable-next-line sonar/deprecation
        this.silent = Boolean(payload.silent);
        this.preventAttributeChangeRequest = Boolean(payload.preventAttributeChangeRequest);
        this.preventStateChange = Boolean(payload.preventStateChange);
        this.preventRunBusinessRules = Boolean(payload.preventRunBusinessRules);
    }
};
HandleViewModelAttributeChangeRequest = __decorate([
    CrtRequest({
        type: 'crt.HandleViewModelAttributeChangeRequest',
    }),
    __metadata("design:paramtypes", [Object])
], HandleViewModelAttributeChangeRequest);
export { HandleViewModelAttributeChangeRequest };
