// @creatio/base/esm2022/lib/public/enums/aggregation-function.enum.mjs
/**
 * @public
 * The type of the aggregation function (can be set to None).
 */
export var ɵAggregationFunction;
(function (ɵAggregationFunction) {
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationFunction["None"] = "";
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationFunction["Count"] = "Count";
    /**
     * @description Sum of values of all elements.
     */
    ɵAggregationFunction["Sum"] = "Sum";
    /**
     * @description Average value for all elements.
     */
    ɵAggregationFunction["Avg"] = "Avg";
    /**
     * @description The minimum value among all elements.
     */
    ɵAggregationFunction["Min"] = "Min";
    /**
     * @description The maximum value among all elements.
     */
    ɵAggregationFunction["Max"] = "Max";
    /**
     * @description Top one record from collection.
     */
    ɵAggregationFunction["TopOne"] = "TopOne";
})(ɵAggregationFunction || (ɵAggregationFunction = {}));
