// @creatio/base/esm2022/lib/public/enums/aggregation-type.enum.mjs
/**
 * @public
 * The type of the aggregation (can be set to None).
 */
export var ɵAggregationType;
(function (ɵAggregationType) {
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationType[ɵAggregationType["None"] = 0] = "None";
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationType[ɵAggregationType["Count"] = 1] = "Count";
    /**
     * @description Sum of values of all elements.
     */
    ɵAggregationType[ɵAggregationType["Sum"] = 2] = "Sum";
    /**
     * @description Average value for all elements.
     */
    ɵAggregationType[ɵAggregationType["Avg"] = 3] = "Avg";
    /**
     * @description The minimum value among all elements.
     */
    ɵAggregationType[ɵAggregationType["Min"] = 4] = "Min";
    /**
     * @description The maximum value among all elements.
     */
    ɵAggregationType[ɵAggregationType["Max"] = 5] = "Max";
    /**
     * @description Top one record from collection.
     */
    ɵAggregationType[ɵAggregationType["TopOne"] = 6] = "TopOne";
})(ɵAggregationType || (ɵAggregationType = {}));
