// @creatio/base/esm2022/lib/public/enums/data-value-type.enum.mjs
/**
 * @public
 * Types of values.
 */
export var ɵDataValueType;
(function (ɵDataValueType) {
    /**
     * @description Global unique identifier.
     */
    ɵDataValueType[ɵDataValueType["Guid"] = 0] = "Guid";
    /**
     * @description String.
     */
    ɵDataValueType[ɵDataValueType["Text"] = 1] = "Text";
    /**
     * @description Integer.
     */
    ɵDataValueType[ɵDataValueType["Integer"] = 4] = "Integer";
    /**
     * @description Real number.
     */
    ɵDataValueType[ɵDataValueType["Float"] = 5] = "Float";
    /**
     * @description Currency.
     */
    ɵDataValueType[ɵDataValueType["Money"] = 6] = "Money";
    /**
     * @description Date and time.
     */
    ɵDataValueType[ɵDataValueType["DateTime"] = 7] = "DateTime";
    /**
     * @description Date.
     */
    ɵDataValueType[ɵDataValueType["Date"] = 8] = "Date";
    /**
     * @description Time.
     */
    ɵDataValueType[ɵDataValueType["Time"] = 9] = "Time";
    /**
     * @description Lookup.
     */
    ɵDataValueType[ɵDataValueType["Lookup"] = 10] = "Lookup";
    /**
     * @description Enumeration.
     */
    ɵDataValueType[ɵDataValueType["Enum"] = 11] = "Enum";
    /**
     * @description Boolean.
     */
    ɵDataValueType[ɵDataValueType["Boolean"] = 12] = "Boolean";
    /**
     * @description Binary data object.
     */
    ɵDataValueType[ɵDataValueType["Blob"] = 13] = "Blob";
    /**
     * @description The object of image binary data.
     */
    ɵDataValueType[ɵDataValueType["Image"] = 14] = "Image";
    /**
     * @description Custom object.
     */
    ɵDataValueType[ɵDataValueType["CUSTOM_OBJECT"] = 15] = "CUSTOM_OBJECT";
    /**
     * @description Image from a lookup.
     */
    ɵDataValueType[ɵDataValueType["IMAGELOOKUP"] = 16] = "IMAGELOOKUP";
    /**
     * @description Collection of items.
     */
    ɵDataValueType[ɵDataValueType["COLLECTION"] = 17] = "COLLECTION";
    /**
     * @description Color.
     */
    ɵDataValueType[ɵDataValueType["Color"] = 18] = "Color";
    /**
     * @description Localized string.
     */
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_STRING"] = 19] = "LOCALIZABLE_STRING";
    /**
     * @description Object (Entity).
     */
    ɵDataValueType[ɵDataValueType["ENTITY"] = 20] = "ENTITY";
    /**
     * @description Collection of objects (EntityCollection).
     */
    ɵDataValueType[ɵDataValueType["ENTITY_COLLECTION"] = 21] = "ENTITY_COLLECTION";
    /**
     * @description List of object column values (EntityColumnMappingCollection).
     */
    ɵDataValueType[ɵDataValueType["ENTITY_COLUMN_MAPPING_COLLECTION"] = 22] = "ENTITY_COLUMN_MAPPING_COLLECTION";
    /**
     * @description Hash string.
     */
    ɵDataValueType[ɵDataValueType["HASH_TEXT"] = 23] = "HASH_TEXT";
    /**
     * @description Encrypted string.
     */
    ɵDataValueType[ɵDataValueType["SECURE_TEXT"] = 24] = "SECURE_TEXT";
    /**
     * @description File.
     */
    ɵDataValueType[ɵDataValueType["FILE"] = 25] = "FILE";
    /**
     * @description Mapping.
     */
    ɵDataValueType[ɵDataValueType["MAPPING"] = 26] = "MAPPING";
    /**
     * @description Short length text.
     */
    ɵDataValueType[ɵDataValueType["SHORT_TEXT"] = 27] = "SHORT_TEXT";
    /**
     * @description Medium length text.
     */
    ɵDataValueType[ɵDataValueType["MEDIUM_TEXT"] = 28] = "MEDIUM_TEXT";
    /**
     * @description Unlimited length text.
     */
    ɵDataValueType[ɵDataValueType["MAXSIZE_TEXT"] = 29] = "MAXSIZE_TEXT";
    /**
     * @description Long length text.
     */
    ɵDataValueType[ɵDataValueType["LONG_TEXT"] = 30] = "LONG_TEXT";
    ɵDataValueType[ɵDataValueType["FLOAT1"] = 31] = "FLOAT1";
    ɵDataValueType[ɵDataValueType["FLOAT2"] = 32] = "FLOAT2";
    ɵDataValueType[ɵDataValueType["FLOAT3"] = 33] = "FLOAT3";
    ɵDataValueType[ɵDataValueType["FLOAT4"] = 34] = "FLOAT4";
    /**
     * @description Localizable parameter values list.
     */
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_PARAMETER_VALUES_LIST"] = 35] = "LOCALIZABLE_PARAMETER_VALUES_LIST";
    /**
     * @description Process not localizable text.
     */
    ɵDataValueType[ɵDataValueType["METADATA_TEXT"] = 36] = "METADATA_TEXT";
    /**
     * @description Stage indicator.
     */
    ɵDataValueType[ɵDataValueType["STAGE_INDICATOR"] = 37] = "STAGE_INDICATOR";
    /**
     * @description Serializable list of values.
     */
    ɵDataValueType[ɵDataValueType["OBJECT_LIST"] = 38] = "OBJECT_LIST";
    /**
     * @description Serializable list of composite values.
     */
    ɵDataValueType[ɵDataValueType["COMPOSITE_OBJECT_LIST"] = 39] = "COMPOSITE_OBJECT_LIST";
    ɵDataValueType[ɵDataValueType["FLOAT8"] = 40] = "FLOAT8";
    /**
     * @description File.
     */
    ɵDataValueType[ɵDataValueType["FILE_LOCATOR"] = 41] = "FILE_LOCATOR";
    /**
     * @description Phone.
     */
    ɵDataValueType[ɵDataValueType["PHONE_TEXT"] = 42] = "PHONE_TEXT";
    /**
     * @description Text with formating tags.
     */
    ɵDataValueType[ɵDataValueType["RICH_TEXT"] = 43] = "RICH_TEXT";
    /**
     * @description Web link.
     */
    ɵDataValueType[ɵDataValueType["WEB_TEXT"] = 44] = "WEB_TEXT";
    /**
     * @description Email.
     */
    ɵDataValueType[ɵDataValueType["EMAIL_TEXT"] = 45] = "EMAIL_TEXT";
    ɵDataValueType[ɵDataValueType["COMPOSITE_OBJECT"] = 46] = "COMPOSITE_OBJECT";
})(ɵDataValueType || (ɵDataValueType = {}));
