// @creatio/base/esm2022/lib/public/esq/data-model/expressions/sub-queries/sub-query-expression.mjs
import { ɵExpressionType as ExpressionType } from '../../enums';
import { ɵFilterGroup as FilterGroup } from '../../filters/filter-group';
import { ɵBaseExpression as BaseExpression } from '../base-expression';
/**
 * @public
 * @description The class that defines the sub query expression.
 * @extends {ɵBaseExpression}
 */
export class ɵSubQueryExpression extends BaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_SubQueryExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...BaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Type of expression.
         */
        this.expressionType = ExpressionType.SubQuery;
        this.columnPath = config?.columnPath;
        const subFilters = config?.subFilters;
        if (typeof subFilters?.clone === 'function') {
            this.subFilters = subFilters.clone();
        }
        else if (subFilters) {
            this.subFilters = new FilterGroup(subFilters.logicalOperation, subFilters.rootSchemaName);
            Object.entries(subFilters.items).forEach(([key, filter]) => {
                this.subFilters.add(filter, key);
            });
        }
        else {
            this.subFilters = subFilters;
        }
        this.subOrderDirection = config.subOrderDirection;
        this.subOrderColumn = config.subOrderColumn;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser) {
        const expression = new ɵSubQueryExpression({
            columnPath: dto['columnPath'],
            subFilters: filterParser.fromJson(dto['subFilters']),
            subOrderDirection: dto['subOrderDirection'],
            subOrderColumn: dto['subOrderColumn'],
        });
        return expression;
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵSubQueryExpression(this);
    }
    /** {@inheritDoc ɵBaseExpression.toJson} */
    toJson() {
        const result = super.toJson();
        if (this.subFilters) {
            result['subFilters'] = this.serializeItem(this.subFilters);
        }
        return result;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            {
                path: this.columnPath,
            },
            ...(this.subFilters?._getUsedColumns() ?? []),
        ];
    }
}
