// @creatio/base/fesm2022/creatio-base.mjs
import { classToPlain, Exclude, Expose } from 'class-transformer';
import { __decorate, __metadata } from 'tslib';
import { isFunction } from 'lodash';

function isHistoricalDate(date) {
    return date.getFullYear() < 1970;
}
function getISOStringWithOffset(date) {
    const timeZoneOffset = date.getTimezoneOffset() * 60000;
    return new Date(date.getTime() - timeZoneOffset).toISOString().slice(0, -1);
}
function getISOStringForPre1970Date(date) {
    const year = date.getFullYear();
    const month = padToTwoDigits(date.getMonth() + 1);
    const day = padToTwoDigits(date.getDate());
    const hours = padToTwoDigits(date.getHours());
    const minutes = padToTwoDigits(date.getMinutes());
    const seconds = padToTwoDigits(date.getSeconds());
    const milliseconds = padToThreeDigits(date.getMilliseconds());
    return `${year}-${month}-${day}T${hours}:${minutes}:${seconds}.${milliseconds}`;
}
function padToTwoDigits(value) {
    return String(value).padStart(2, '0');
}
function padToThreeDigits(value) {
    return String(value).padStart(3, '0');
}
/**
 * @public
 * Returns a string represent of date in ISO 8601 without time zone specification(doesn't convert it
 * using local timezone).
 *
 * Example:
 *
 *      const date = new Date(2018, 01, 20, 15, 10, 30);
 *      const result = toLocalISOString(date);
 *      //result: '2018-02-20T15:10:30'
 *
 * @param date Date value.
 */
function ɵtoLocalISOString(date) {
    if (!(date instanceof Date)) {
        throw new TypeError();
    }
    return isHistoricalDate(date) ? getISOStringForPre1970Date(date) : getISOStringWithOffset(date);
}
/**
 * @public
 * Returns encoded date in string value.
 * @param date Date value.
 */
function ɵencodeDate(date) {
    return `"${ɵtoLocalISOString(date)}"`;
}

/**
 * @public
 */
var ɵDataSourceScope;
(function (ɵDataSourceScope) {
    ɵDataSourceScope["ViewElement"] = "viewElement";
    ɵDataSourceScope["Page"] = "page";
})(ɵDataSourceScope || (ɵDataSourceScope = {}));

/**
 * @public
 */
var ɵDataSchemaAttributeType;
(function (ɵDataSchemaAttributeType) {
    ɵDataSchemaAttributeType["OwnAttribute"] = "Own";
    ɵDataSchemaAttributeType["ForwardReferenceAttribute"] = "ForwardReference";
    ɵDataSchemaAttributeType["BackReferenceAttribute"] = "BackReference";
    ɵDataSchemaAttributeType["AggregationAttribute"] = "Aggregation";
    ɵDataSchemaAttributeType["FunctionAttribute"] = "Function";
})(ɵDataSchemaAttributeType || (ɵDataSchemaAttributeType = {}));
/**
 * @public
 */
var ɵDataSchemaAttributeUsageType;
(function (ɵDataSchemaAttributeUsageType) {
    ɵDataSchemaAttributeUsageType[ɵDataSchemaAttributeUsageType["General"] = 0] = "General";
    ɵDataSchemaAttributeUsageType[ɵDataSchemaAttributeUsageType["Advanced"] = 1] = "Advanced";
    ɵDataSchemaAttributeUsageType[ɵDataSchemaAttributeUsageType["None"] = 2] = "None";
})(ɵDataSchemaAttributeUsageType || (ɵDataSchemaAttributeUsageType = {}));

/**
 * @public
 */
var ɵDefaultValueSource;
(function (ɵDefaultValueSource) {
    ɵDefaultValueSource[ɵDefaultValueSource["None"] = 0] = "None";
    ɵDefaultValueSource[ɵDefaultValueSource["Const"] = 1] = "Const";
    ɵDefaultValueSource[ɵDefaultValueSource["Settings"] = 2] = "Settings";
    ɵDefaultValueSource[ɵDefaultValueSource["SystemValue"] = 3] = "SystemValue";
    ɵDefaultValueSource[ɵDefaultValueSource["Sequence"] = 4] = "Sequence";
})(ɵDefaultValueSource || (ɵDefaultValueSource = {}));

/**
 * Lookup mode
 * @internal
 */
var LookupMode;
(function (LookupMode) {
    LookupMode["List"] = "List";
    LookupMode["SelectionWindow"] = "SelectionWindow";
})(LookupMode || (LookupMode = {}));

/**
 * Available platform APIs
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
var ɵPlatformAPIs;
(function (ɵPlatformAPIs) {
    /**
     * Selection metadata API.
     */
    ɵPlatformAPIs["Selection"] = "selection";
    /**
     * Filtration metadata API.
     */
    ɵPlatformAPIs["Filtration"] = "filtration";
})(ɵPlatformAPIs || (ɵPlatformAPIs = {}));

/**
 * @public
 * Types of values.
 */
var ɵDataValueType;
(function (ɵDataValueType) {
    /**
     * @description Global unique identifier.
     */
    ɵDataValueType[ɵDataValueType["Guid"] = 0] = "Guid";
    /**
     * @description String.
     */
    ɵDataValueType[ɵDataValueType["Text"] = 1] = "Text";
    /**
     * @description Integer.
     */
    ɵDataValueType[ɵDataValueType["Integer"] = 4] = "Integer";
    /**
     * @description Real number.
     */
    ɵDataValueType[ɵDataValueType["Float"] = 5] = "Float";
    /**
     * @description Currency.
     */
    ɵDataValueType[ɵDataValueType["Money"] = 6] = "Money";
    /**
     * @description Date and time.
     */
    ɵDataValueType[ɵDataValueType["DateTime"] = 7] = "DateTime";
    /**
     * @description Date.
     */
    ɵDataValueType[ɵDataValueType["Date"] = 8] = "Date";
    /**
     * @description Time.
     */
    ɵDataValueType[ɵDataValueType["Time"] = 9] = "Time";
    /**
     * @description Lookup.
     */
    ɵDataValueType[ɵDataValueType["Lookup"] = 10] = "Lookup";
    /**
     * @description Enumeration.
     */
    ɵDataValueType[ɵDataValueType["Enum"] = 11] = "Enum";
    /**
     * @description Boolean.
     */
    ɵDataValueType[ɵDataValueType["Boolean"] = 12] = "Boolean";
    /**
     * @description Binary data object.
     */
    ɵDataValueType[ɵDataValueType["Blob"] = 13] = "Blob";
    /**
     * @description The object of image binary data.
     */
    ɵDataValueType[ɵDataValueType["Image"] = 14] = "Image";
    /**
     * @description Custom object.
     */
    ɵDataValueType[ɵDataValueType["CUSTOM_OBJECT"] = 15] = "CUSTOM_OBJECT";
    /**
     * @description Image from a lookup.
     */
    ɵDataValueType[ɵDataValueType["IMAGELOOKUP"] = 16] = "IMAGELOOKUP";
    /**
     * @description Collection of items.
     */
    ɵDataValueType[ɵDataValueType["COLLECTION"] = 17] = "COLLECTION";
    /**
     * @description Color.
     */
    ɵDataValueType[ɵDataValueType["Color"] = 18] = "Color";
    /**
     * @description Localized string.
     */
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_STRING"] = 19] = "LOCALIZABLE_STRING";
    /**
     * @description Object (Entity).
     */
    ɵDataValueType[ɵDataValueType["ENTITY"] = 20] = "ENTITY";
    /**
     * @description Collection of objects (EntityCollection).
     */
    ɵDataValueType[ɵDataValueType["ENTITY_COLLECTION"] = 21] = "ENTITY_COLLECTION";
    /**
     * @description List of object column values (EntityColumnMappingCollection).
     */
    ɵDataValueType[ɵDataValueType["ENTITY_COLUMN_MAPPING_COLLECTION"] = 22] = "ENTITY_COLUMN_MAPPING_COLLECTION";
    /**
     * @description Hash string.
     */
    ɵDataValueType[ɵDataValueType["HASH_TEXT"] = 23] = "HASH_TEXT";
    /**
     * @description Encrypted string.
     */
    ɵDataValueType[ɵDataValueType["SECURE_TEXT"] = 24] = "SECURE_TEXT";
    /**
     * @description File.
     */
    ɵDataValueType[ɵDataValueType["FILE"] = 25] = "FILE";
    /**
     * @description Mapping.
     */
    ɵDataValueType[ɵDataValueType["MAPPING"] = 26] = "MAPPING";
    /**
     * @description Short length text.
     */
    ɵDataValueType[ɵDataValueType["SHORT_TEXT"] = 27] = "SHORT_TEXT";
    /**
     * @description Medium length text.
     */
    ɵDataValueType[ɵDataValueType["MEDIUM_TEXT"] = 28] = "MEDIUM_TEXT";
    /**
     * @description Unlimited length text.
     */
    ɵDataValueType[ɵDataValueType["MAXSIZE_TEXT"] = 29] = "MAXSIZE_TEXT";
    /**
     * @description Long length text.
     */
    ɵDataValueType[ɵDataValueType["LONG_TEXT"] = 30] = "LONG_TEXT";
    ɵDataValueType[ɵDataValueType["FLOAT1"] = 31] = "FLOAT1";
    ɵDataValueType[ɵDataValueType["FLOAT2"] = 32] = "FLOAT2";
    ɵDataValueType[ɵDataValueType["FLOAT3"] = 33] = "FLOAT3";
    ɵDataValueType[ɵDataValueType["FLOAT4"] = 34] = "FLOAT4";
    /**
     * @description Localizable parameter values list.
     */
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_PARAMETER_VALUES_LIST"] = 35] = "LOCALIZABLE_PARAMETER_VALUES_LIST";
    /**
     * @description Process not localizable text.
     */
    ɵDataValueType[ɵDataValueType["METADATA_TEXT"] = 36] = "METADATA_TEXT";
    /**
     * @description Stage indicator.
     */
    ɵDataValueType[ɵDataValueType["STAGE_INDICATOR"] = 37] = "STAGE_INDICATOR";
    /**
     * @description Serializable list of values.
     */
    ɵDataValueType[ɵDataValueType["OBJECT_LIST"] = 38] = "OBJECT_LIST";
    /**
     * @description Serializable list of composite values.
     */
    ɵDataValueType[ɵDataValueType["COMPOSITE_OBJECT_LIST"] = 39] = "COMPOSITE_OBJECT_LIST";
    ɵDataValueType[ɵDataValueType["FLOAT8"] = 40] = "FLOAT8";
    /**
     * @description File.
     */
    ɵDataValueType[ɵDataValueType["FILE_LOCATOR"] = 41] = "FILE_LOCATOR";
    /**
     * @description Phone.
     */
    ɵDataValueType[ɵDataValueType["PHONE_TEXT"] = 42] = "PHONE_TEXT";
    /**
     * @description Text with formating tags.
     */
    ɵDataValueType[ɵDataValueType["RICH_TEXT"] = 43] = "RICH_TEXT";
    /**
     * @description Web link.
     */
    ɵDataValueType[ɵDataValueType["WEB_TEXT"] = 44] = "WEB_TEXT";
    /**
     * @description Email.
     */
    ɵDataValueType[ɵDataValueType["EMAIL_TEXT"] = 45] = "EMAIL_TEXT";
    ɵDataValueType[ɵDataValueType["COMPOSITE_OBJECT"] = 46] = "COMPOSITE_OBJECT";
})(ɵDataValueType || (ɵDataValueType = {}));

/**
 * @public
 * The type of the aggregation (can be set to None).
 */
var ɵAggregationType;
(function (ɵAggregationType) {
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationType[ɵAggregationType["None"] = 0] = "None";
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationType[ɵAggregationType["Count"] = 1] = "Count";
    /**
     * @description Sum of values of all elements.
     */
    ɵAggregationType[ɵAggregationType["Sum"] = 2] = "Sum";
    /**
     * @description Average value for all elements.
     */
    ɵAggregationType[ɵAggregationType["Avg"] = 3] = "Avg";
    /**
     * @description The minimum value among all elements.
     */
    ɵAggregationType[ɵAggregationType["Min"] = 4] = "Min";
    /**
     * @description The maximum value among all elements.
     */
    ɵAggregationType[ɵAggregationType["Max"] = 5] = "Max";
    /**
     * @description Top one record from collection.
     */
    ɵAggregationType[ɵAggregationType["TopOne"] = 6] = "TopOne";
})(ɵAggregationType || (ɵAggregationType = {}));

/**
 * @public
 * The type of the aggregation function (can be set to None).
 */
var ɵAggregationFunction;
(function (ɵAggregationFunction) {
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationFunction["None"] = "";
    /**
     * @description Aggregation function type not defined.
     */
    ɵAggregationFunction["Count"] = "Count";
    /**
     * @description Sum of values of all elements.
     */
    ɵAggregationFunction["Sum"] = "Sum";
    /**
     * @description Average value for all elements.
     */
    ɵAggregationFunction["Avg"] = "Avg";
    /**
     * @description The minimum value among all elements.
     */
    ɵAggregationFunction["Min"] = "Min";
    /**
     * @description The maximum value among all elements.
     */
    ɵAggregationFunction["Max"] = "Max";
    /**
     * @description Top one record from collection.
     */
    ɵAggregationFunction["TopOne"] = "TopOne";
})(ɵAggregationFunction || (ɵAggregationFunction = {}));

/**
 * @public
 * @decription Type of DataSchema. In combination with the name is used for unique identification of {@link ɵDataSchema}.
 */
var ɵDataSchemaType;
(function (ɵDataSchemaType) {
    /**
     * @description Entity.
     */
    ɵDataSchemaType["Entity"] = "Entity";
    /**
     * @description Client unit.
     */
    ɵDataSchemaType["ClientUnit"] = "ClientUnit";
    /**
     * @description Intent.
     */
    ɵDataSchemaType["Intent"] = "Intent";
})(ɵDataSchemaType || (ɵDataSchemaType = {}));

/**
 * @public
 * @description Type of the action of the model.
 */
var ɵModelInPageAction;
(function (ɵModelInPageAction) {
    /**
     * @description Action for editing existed record.
     */
    ɵModelInPageAction["Edit"] = "edit";
    /**
     * @description Action for adding new record.
     */
    ɵModelInPageAction["Add"] = "add";
    /**
     * @description Action for copying the record.
     */
    ɵModelInPageAction["Copy"] = "copy";
})(ɵModelInPageAction || (ɵModelInPageAction = {}));

/**
 * @public
 * @description Type of message channel. Used to specify where the message can be received.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
var ɵMessageChannelType;
(function (ɵMessageChannelType) {
    /**
     * @public
     * @description Message will be sent to the same user who sent the message (client side).
     */
    ɵMessageChannelType["PTP"] = "PTPMsg";
    /**
     * @public
     * @description Message will be sent to all users (client side).
     */
    ɵMessageChannelType["BROADCAST"] = "BroadcastMsg";
    /**
     * @public
     * @description Message will be sent to the server.
     */
    ɵMessageChannelType["SERVER"] = "ServerMsg";
})(ɵMessageChannelType || (ɵMessageChannelType = {}));

/**
 * @public
 * @description SysAdminUnitInRole source type.
 */
var eAdminUnitRoleSources;
(function (eAdminUnitRoleSources) {
    /**
     * @description Empty value.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["None"] = 0] = "None";
    /**
     * @description Self.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["Self"] = 1] = "Self";
    /**
     * @description Explicit entry into role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["ExplicitEntry"] = 2] = "ExplicitEntry";
    /**
     * @description Delegated role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["Delegated"] = 4] = "Delegated";
    /**
     * @description Gets functional role from organizational role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["FuncRoleFromOrgRole"] = 8] = "FuncRoleFromOrgRole";
    /**
     * @description Gets role up hierarchy from role.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["UpHierarchy"] = 16] = "UpHierarchy";
    /**
     * @description Gets role as a manager.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["AsManager"] = 32] = "AsManager";
    /**
     * @description Gets all roles.
     */
    eAdminUnitRoleSources[eAdminUnitRoleSources["All"] = 63] = "All";
})(eAdminUnitRoleSources || (eAdminUnitRoleSources = {}));

/**
 * @public
 * @description Scope of the aggregation function.
 */
var ɵAggregationEvalType;
(function (ɵAggregationEvalType) {
    /**
     * @description The scope of the aggregating function is not defined.
     */
    ɵAggregationEvalType[ɵAggregationEvalType["None"] = 0] = "None";
    /**
     * @description Applies to all elements.
     */
    ɵAggregationEvalType[ɵAggregationEvalType["All"] = 1] = "All";
    /**
     * @description Applies to unique values.
     */
    ɵAggregationEvalType[ɵAggregationEvalType["Distinct"] = 2] = "Distinct";
})(ɵAggregationEvalType || (ɵAggregationEvalType = {}));

/**
 * @public
 * @description Arithmetic operation.
 */
var ɵArithmeticOperation;
(function (ɵArithmeticOperation) {
    /**
     * @description Addition.
     */
    ɵArithmeticOperation[ɵArithmeticOperation["Addition"] = 0] = "Addition";
    /**
     * @description Subtraction.
     */
    ɵArithmeticOperation[ɵArithmeticOperation["Subtraction"] = 1] = "Subtraction";
    /**
     * @description Multiplication.
     */
    ɵArithmeticOperation[ɵArithmeticOperation["Multiplication"] = 2] = "Multiplication";
    /**
     * @description Division.
     */
    ɵArithmeticOperation[ɵArithmeticOperation["Division"] = 3] = "Division";
})(ɵArithmeticOperation || (ɵArithmeticOperation = {}));

/**
 * @public
 * @description The type of comparison operation in the filter.
 */
/* eslint-disable @typescript-eslint/naming-convention */
var ɵComparisonType;
(function (ɵComparisonType) {
    /**
     * @public
     * @description Value is between two specified values.
     */
    ɵComparisonType[ɵComparisonType["Between"] = 0] = "Between";
    /**
     * @public
     * @description Value contains a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Contain"] = 11] = "Contain";
    /**
     * @public
     * @description Value ends with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["End_with"] = 13] = "End_with";
    /**
     * @public
     * @description Value is equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Equal"] = 3] = "Equal";
    /**
     * @public
     * @description Value exists (is not null or undefined).
     */
    ɵComparisonType[ɵComparisonType["Exists"] = 15] = "Exists";
    /**
     * @public
     * @description Value is greater than a specified value.
     */
    ɵComparisonType[ɵComparisonType["Greater"] = 7] = "Greater";
    /**
     * @public
     * @description Value is greater than or equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Greater_or_equal"] = 8] = "Greater_or_equal";
    /**
     * @public
     * @description Value is not null.
     */
    ɵComparisonType[ɵComparisonType["Is_not_null"] = 2] = "Is_not_null";
    /**
     * @public
     * @description Value is null.
     */
    ɵComparisonType[ɵComparisonType["Is_null"] = 1] = "Is_null";
    /**
     * @public
     * @description Value is less than a specified value.
     */
    ɵComparisonType[ɵComparisonType["Less"] = 5] = "Less";
    /**
     * @public
     * @description Value is less than or equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Less_or_equal"] = 6] = "Less_or_equal";
    /**
     * @public
     * @description Value does not contain a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_contain"] = 12] = "Not_contain";
    /**
     * @public
     * @description Value does not end with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_end_with"] = 14] = "Not_end_with";
    /**
     * @public
     * @description Value is not equal to a specified value.
     */
    ɵComparisonType[ɵComparisonType["Not_equal"] = 4] = "Not_equal";
    /**
     * @public
     * @description Value does not exist (is null or undefined).
     */
    ɵComparisonType[ɵComparisonType["Not_exists"] = 16] = "Not_exists";
    /**
     * @public
     * @description Value does not start with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Not_start_with"] = 10] = "Not_start_with";
    /**
     * @public
     * @description Value starts with a specified substring.
     */
    ɵComparisonType[ɵComparisonType["Start_with"] = 9] = "Start_with";
})(ɵComparisonType || (ɵComparisonType = {}));

/**
 * @public
 * @description Part of the date.
 */
var ɵDatePartType;
(function (ɵDatePartType) {
    /**
     * @description Empty value.
     */
    ɵDatePartType[ɵDatePartType["None"] = 0] = "None";
    /**
     * @description Day.
     */
    ɵDatePartType[ɵDatePartType["Day"] = 1] = "Day";
    /**
     * @description Week.
     */
    ɵDatePartType[ɵDatePartType["Week"] = 2] = "Week";
    /**
     * @description Month.
     */
    ɵDatePartType[ɵDatePartType["Month"] = 3] = "Month";
    /**
     * @description Year.
     */
    ɵDatePartType[ɵDatePartType["Year"] = 4] = "Year";
    /**
     * @description Day of the week.
     */
    ɵDatePartType[ɵDatePartType["WeekDay"] = 5] = "WeekDay";
    /**
     * @description Hour.
     */
    ɵDatePartType[ɵDatePartType["Hour"] = 6] = "Hour";
    /**
     * @description Minute.
     */
    ɵDatePartType[ɵDatePartType["HourMinute"] = 7] = "HourMinute";
})(ɵDatePartType || (ɵDatePartType = {}));

/**
 * @public
 * @description Type of expression.
 */
var ɵExpressionType;
(function (ɵExpressionType) {
    /**
     * @description Column.
     */
    ɵExpressionType[ɵExpressionType["SchemaColumn"] = 0] = "SchemaColumn";
    /**
     * @description Function.
     */
    ɵExpressionType[ɵExpressionType["Function"] = 1] = "Function";
    /**
     * @description Parameter.
     */
    ɵExpressionType[ɵExpressionType["Parameter"] = 2] = "Parameter";
    /**
     * @description Subselect query.
     */
    ɵExpressionType[ɵExpressionType["SubQuery"] = 3] = "SubQuery";
    /**
     * @description Arithmetic operation.
     */
    ɵExpressionType[ɵExpressionType["ArithmeticOperation"] = 4] = "ArithmeticOperation";
})(ɵExpressionType || (ɵExpressionType = {}));

/**
 * @public
 * @description The type of filter.
 */
var ɵFilterType;
(function (ɵFilterType) {
    /**
     * @public
     * @description No filter applied.
     */
    ɵFilterType[ɵFilterType["None"] = 0] = "None";
    /**
     * @public
     * @description Comparison filter (e.g., equal to, greater than).
     */
    ɵFilterType[ɵFilterType["Compare"] = 1] = "Compare";
    /**
     * @public
     * @description Checks if a value is null.
     */
    ɵFilterType[ɵFilterType["IsNull"] = 2] = "IsNull";
    /**
     * @public
     * @description Checks if a value is between two specified values.
     */
    ɵFilterType[ɵFilterType["Between"] = 3] = "Between";
    /**
     * @public
     * @description Checks if a value is in a specified set of values.
     */
    ɵFilterType[ɵFilterType["In"] = 4] = "In";
    /**
     * @public
     * @description Checks if a value exists.
     */
    ɵFilterType[ɵFilterType["Exists"] = 5] = "Exists";
    /**
     * @public
     * @description A group of filters combined with logical operators (AND, OR).
     */
    ɵFilterType[ɵFilterType["FilterGroup"] = 6] = "FilterGroup";
})(ɵFilterType || (ɵFilterType = {}));

/**
 * @public
 * @description The type of function in the filter.
 */
var ɵFunctionType;
(function (ɵFunctionType) {
    /**
     * @description Function type not defined.
     * */
    ɵFunctionType[ɵFunctionType["None"] = 0] = "None";
    /**
     * @description Substitution by macro.
     * */
    ɵFunctionType[ɵFunctionType["Macros"] = 1] = "Macros";
    /**
     * @description Aggregating function.
     * */
    ɵFunctionType[ɵFunctionType["Aggregation"] = 2] = "Aggregation";
    /**
     * @description Date fragment.
     * */
    ɵFunctionType[ɵFunctionType["DatePart"] = 3] = "DatePart";
    /**
     * @description The size of the value in bytes, used for binary data.
     * */
    ɵFunctionType[ɵFunctionType["Length"] = 4] = "Length";
    /**
     * @description Window query function
     * */
    ɵFunctionType[ɵFunctionType["Window"] = 5] = "Window";
    /**
     * @description Date add.
     * */
    ɵFunctionType[ɵFunctionType["DateAdd"] = 6] = "DateAdd";
    /**
     * @description Date diff.
     * */
    ɵFunctionType[ɵFunctionType["DateDiff"] = 7] = "DateDiff";
})(ɵFunctionType || (ɵFunctionType = {}));

/**
 * @public
 * @description The type of logical operation.
 */
var ɵLogicalOperatorType;
(function (ɵLogicalOperatorType) {
    /**
     * @description Logical And.
     */
    ɵLogicalOperatorType[ɵLogicalOperatorType["And"] = 0] = "And";
    /**
     * @description Logical Or.
     */
    ɵLogicalOperatorType[ɵLogicalOperatorType["Or"] = 1] = "Or";
})(ɵLogicalOperatorType || (ɵLogicalOperatorType = {}));

/**
 * @public
 * @description Sort order (can be set to None).
 */
var ɵOrderDirection;
(function (ɵOrderDirection) {
    /**
     * @description Order not defined.
     * */
    ɵOrderDirection[ɵOrderDirection["None"] = 0] = "None";
    /**
     * @description Ascending.
     * */
    ɵOrderDirection[ɵOrderDirection["Asc"] = 1] = "Asc";
    /**
     * @description Descending.
     * */
    ɵOrderDirection[ɵOrderDirection["Desc"] = 2] = "Desc";
})(ɵOrderDirection || (ɵOrderDirection = {}));

/**
 * @public
 * @description The type of macros.
 */
var ɵQueryMacrosType;
(function (ɵQueryMacrosType) {
    /**
     * @description Macro type not defined.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["None"] = 0] = "None";
    /**
     * @description Current user.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUser"] = 1] = "CurrentUser";
    /**
     * @description Current user contact.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUserContact"] = 2] = "CurrentUserContact";
    /**
     * @description Yesterday.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Yesterday"] = 3] = "Yesterday";
    /**
     * @description Today.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Today"] = 4] = "Today";
    /**
     * @description Tomorrow.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["Tomorrow"] = 5] = "Tomorrow";
    /**
     * @description Previous week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousWeek"] = 6] = "PreviousWeek";
    /**
     * @description This week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentWeek"] = 7] = "CurrentWeek";
    /**
     * @description Next week.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextWeek"] = 8] = "NextWeek";
    /**
     * @description Previous month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousMonth"] = 9] = "PreviousMonth";
    /**
     * @description Current month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentMonth"] = 10] = "CurrentMonth";
    /**
     * @description Next month.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextMonth"] = 11] = "NextMonth";
    /**
     * @description Previous quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousQuarter"] = 12] = "PreviousQuarter";
    /**
     * @description Current quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentQuarter"] = 13] = "CurrentQuarter";
    /**
     * @description Next quarter.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextQuarter"] = 14] = "NextQuarter";
    /**
     * @description Previous half of the year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHalfYear"] = 15] = "PreviousHalfYear";
    /**
     * @description Current half-year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHalfYear"] = 16] = "CurrentHalfYear";
    /**
     * @description Next half-year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHalfYear"] = 17] = "NextHalfYear";
    /**
     * @description Last year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousYear"] = 18] = "PreviousYear";
    /**
     * @description This year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentYear"] = 19] = "CurrentYear";
    /**
     * @description Previous hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHour"] = 20] = "PreviousHour";
    /**
     * @description Current hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHour"] = 21] = "CurrentHour";
    /**
     * @description Next hour.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHour"] = 22] = "NextHour";
    /**
     * @description Next year.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextYear"] = 23] = "NextYear";
    /**
     * @description Next N Days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDays"] = 24] = "NextNDays";
    /**
     * @description Previous N days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDays"] = 25] = "PreviousNDays";
    /**
     * @description Next N hours.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNHours"] = 26] = "NextNHours";
    /**
     * @description Previous N hours.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNHours"] = 27] = "PreviousNHours";
    /**
     * @description Primary column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColumn"] = 34] = "PrimaryColumn";
    /**
     * @description Primary displayed column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryDisplayColumn"] = 35] = "PrimaryDisplayColumn";
    /**
     * @description Primary image display column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryImageColumn"] = 36] = "PrimaryImageColumn";
    /**
     * @description Anniversary today.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearToday"] = 37] = "DayOfYearToday";
    /**
     * @description Anniversary on the date computed as today plus days offset.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearTodayPlusDaysOffset"] = 38] = "DayOfYearTodayPlusDaysOffset";
    /**
     * @description Anniversary on the next several days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDaysOfYear"] = 39] = "NextNDaysOfYear";
    /**
     * @description Anniversary on the previous several days.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDaysOfYear"] = 40] = "PreviousNDaysOfYear";
    /**
     * @description Primary color display column.
     */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColorColumn"] = 41] = "PrimaryColorColumn";
})(ɵQueryMacrosType || (ɵQueryMacrosType = {}));

/**
 * @public
 * @abstract
 * @description The base class of the expression.
 * @extends {ɵSerializable}
 */
class ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [this._instanceOfKey];
    }
    /**
     * @public
     * @description Method for checking if an instance is of the current class type.
     * @param instance - The instance to check.
     * @returns `true` if the instance is of the current class type, `false` otherwise.
     */
    static [Symbol.hasInstance](instance) {
        const instanceOfKeys = instance?.constructor?.['_instanceOfKeys'] ?? [];
        return instanceOfKeys.includes(this._instanceOfKey);
    }
    constructor(config) {
        /**
         * @description True if expression has brackets.
         */
        this.isBlock = false;
        this.isBlock = config?.isBlock;
    }
    /**
     * @public
     * @description Parses expression from json.
     */
    static fromJson(dto, filterParser, expressionParser) {
        throw new Error(`fromJson not supported on expression type ${this.constructor.name}`);
    }
    /**
     * @description Converts instance to json object.
     * @param item - The instance to serialize.
     */
    serializeItem(item) {
        return typeof item.toJson === 'function'
            ? item.toJson()
            : classToPlain(item, {
                exposeUnsetFields: false,
                excludePrefixes: ['_'],
            });
    }
    /**
     * @public
     * @description Converts instance to json object.
     */
    toJson() {
        return classToPlain(this, {
            exposeUnsetFields: false,
            excludePrefixes: ['_'],
        });
    }
}

/**
 * @public
 * @description Arithmetic expression.
 * @extends {ɵBaseExpression}
 */
class ɵArithmeticExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ArithmeticExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Expression type. Default 'ArithmeticOperation'.
         */
        this.expressionType = ɵExpressionType.ArithmeticOperation;
        this.arithmeticOperation = config?.arithmeticOperation;
        this.leftArithmeticOperand = config?.leftArithmeticOperand.clone();
        this.rightArithmeticOperand = config?.rightArithmeticOperand.clone();
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const leftArithmeticOperand = expressionParser?.fromJson(dto['leftArithmeticOperand'], filterParser);
        const rightArithmeticOperand = expressionParser?.fromJson(dto['rightArithmeticOperand'], filterParser);
        const config = {
            leftArithmeticOperand,
            rightArithmeticOperand,
            arithmeticOperation: dto['arithmeticOperation'],
            isBlock: dto['isBlock'],
        };
        return new ɵArithmeticExpression(config);
    }
    /**
     * @public
     * @description Clones instance.
     */
    clone() {
        return new ɵArithmeticExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [...this.leftArithmeticOperand._getUsedColumns(), ...this.rightArithmeticOperand._getUsedColumns()];
    }
}

/**
 * @public
 * @description A class that defines an expression of type column EntitySchema.
 * @extends {ɵBaseExpression}
 */
class ɵColumnExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ColumnExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Type of expression.
         */
        this.expressionType = ɵExpressionType.SchemaColumn;
        this.columnPath = config?.columnPath;
    }
    /**
     * Parses expression from json.
     * @public
     * @param dto - Serialized object.
     */
    static fromJson(dto) {
        const expression = new ɵColumnExpression({ columnPath: dto['columnPath'] });
        return expression;
    }
    /**
     * @description Clones instance.
     */
    clone() {
        return new ɵColumnExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            {
                path: this.columnPath,
            },
        ];
    }
}

/**
 * @public
 * @description Expression parser interface.
 */
class ɵBaseExpressionParser {
    /**
     * @public
     * @description Parses filter from json.
     * @param dto - Serialized object.
     * @param filterParser - Type of parser.
     * @return InstanceType of expression.
     */
    static fromJson(dto, filterParser) {
        throw new Error('not implemented');
    }
}

/**
 * @public
 * @abstract
 * @description A class that describes a functional expression.
 * @extends {ɵFunctionExpression}
 */
class ɵFunctionExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_FunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Expression type. Default 'Function'.
         */
        this.expressionType = ɵExpressionType.Function;
    }
}

/**
 * @public
 * @abstract
 * @description A class that describes a functional expression with argument.
 * @extends {ɵFunctionExpression}
 */
class ɵArgumentFunctionExpression extends ɵFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ArgumentFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        this.functionArgument = config?.functionArgument?.clone();
    }
    /** {@inheritDoc ɵBaseExpression.toJson} */
    toJson() {
        const result = super.toJson();
        if (this.functionArgument) {
            result['functionArgument'] = this.serializeItem(this.functionArgument);
        }
        return result;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.functionArgument?._getUsedColumns() ?? [];
    }
}

/**
 * @public
 * @description A class that describes expression with aggregate function.
 * @extends {ɵArgumentFunctionExpression}
 */
class ɵAggregationFunctionExpression extends ɵArgumentFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_AggregationFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵArgumentFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description The type of the functional expression passed for initialization.
         */
        this.functionType = ɵFunctionType.Aggregation;
        this.aggregationType = config?.aggregationType;
        this.aggregationEvalType = config?.aggregationEvalType;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = expressionParser?.fromJson(dto['functionArgument'], filterParser);
        const aggregationType = dto['aggregationType'];
        const aggregationEvalType = dto['aggregationEvalType'];
        return new ɵAggregationFunctionExpression({
            functionArgument,
            aggregationType,
            aggregationEvalType,
            isBlock: dto['isBlock'],
        });
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵAggregationFunctionExpression(this);
    }
}

/**
 * @public
 * @description A class that describes expression part of date function.
 * @extends {ɵArgumentFunctionExpression}
 */
class ɵDatePartFunctionExpression extends ɵArgumentFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_DatePartFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵArgumentFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Function type. Default 'DatePart'.
         */
        this.functionType = ɵFunctionType.DatePart;
        this.datePartType = config?.datePartType;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = expressionParser?.fromJson(dto['functionArgument'], filterParser);
        const datePartType = dto['datePartType'];
        return new ɵDatePartFunctionExpression({ functionArgument, datePartType, isBlock: dto['isBlock'] });
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵDatePartFunctionExpression(this);
    }
}

/**
 * @public
 * @description A class that describes expression with length function.
 * @extends {ɵArgumentFunctionExpression}
 */
class ɵLengthFunctionExpression extends ɵArgumentFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_LengthFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵArgumentFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Function type. Default 'Length'.
         */
        this.functionType = ɵFunctionType.Length;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = expressionParser?.fromJson(dto['functionArgument'], filterParser);
        return new ɵLengthFunctionExpression({ functionArgument, isBlock: dto['isBlock'] });
    }
    /**
     * @description Method to create cloned instance.
     */
    clone() {
        return new ɵLengthFunctionExpression(this);
    }
}

/**
 * @public
 * @description A class that describes expression with macros function.
 * @extends {ɵFunctionExpression}
 */
class ɵMacrosFunctionExpression extends ɵFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_MacrosFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Function type. Default 'Macros'.
         */
        this.functionType = ɵFunctionType.Macros;
        this.macrosType = config?.macrosType;
        this.functionArgument = config?.functionArgument;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = filterParser && expressionParser && dto['functionArgument']
            ? expressionParser.fromJson(dto['functionArgument'], filterParser)
            : undefined;
        return new ɵMacrosFunctionExpression({ functionArgument, macrosType: dto['macrosType'] });
    }
    /**
     * @description method to create cloned instance.
     */
    clone() {
        return new ɵMacrosFunctionExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.functionArgument?._getUsedColumns() ?? [];
    }
}

/**
 * @public
 * @description A class that describes expression with window function.
 * @extends {ɵArgumentFunctionExpression}
 */
class ɵWindowFunctionExpression extends ɵArgumentFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_WindowFunctionExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵArgumentFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Function type.
         */
        this.functionType = ɵFunctionType.Window;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = expressionParser?.fromJson(dto['functionArgument'], filterParser);
        return new ɵWindowFunctionExpression({ functionArgument, isBlock: dto['isBlock'] });
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵWindowFunctionExpression(this);
    }
}

/**
 * @public
 * @description The class that defines the parametric expression.
 * @extends {ɵBaseExpression}
 */
class ɵParameterExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ParameterExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Type of expression.
         */
        this.expressionType = ɵExpressionType.Parameter;
        const value = config?.value;
        if (value instanceof Date) {
            const cloneDate = new Date();
            cloneDate.setTime(value.getTime());
            this.value = cloneDate;
        }
        else if (!value) {
            this.value = value;
        }
        else {
            this.value = typeof value === 'object' ? { ...value } : value;
        }
        this.dataValueType = config?.dataValueType ?? ɵDataValueType.Text;
        this.parameter = {
            dataValueType: this.dataValueType,
            ...this._getParameterValueMetadata(),
        };
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto) {
        const expression = new ɵParameterExpression(dto['parameter']);
        return expression;
    }
    _convertStringToBlobArray(str) {
        const out = [];
        let position = 0;
        for (let i = 0; i < str.length; i++) {
            const code = str.charCodeAt(i);
            if (code < 128) {
                out[position++] = String.fromCharCode(code);
            }
            else if (code < 2048) {
                /* eslint-disable no-bitwise */
                out[position++] = String.fromCharCode((code >> 6) | 192);
                out[position++] = String.fromCharCode((code & 63) | 128);
            }
            else {
                out[position++] = String.fromCharCode((code >> 12) | 224);
                out[position++] = String.fromCharCode(((code >> 6) & 63) | 128);
                out[position++] = String.fromCharCode((code & 63) | 128);
                /* eslint-enable no-bitwise */
            }
        }
        return out;
    }
    _getParameterValueMetadata() {
        if (this.dataValueType === ɵDataValueType.Blob) {
            return {
                arrayValue: !this.value ? [] : this._convertStringToBlobArray(this.value),
            };
        }
        else if (this.value instanceof Date) {
            return {
                value: ɵencodeDate(this.value),
            };
        }
        else {
            return {
                value: this.value,
            };
        }
    }
    /**
     * @description Returns cloned instance.
     */
    clone() {
        return new ɵParameterExpression(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [];
    }
}
__decorate([
    Exclude(),
    __metadata("design:type", Object)
], ɵParameterExpression.prototype, "value", void 0);
__decorate([
    Exclude(),
    __metadata("design:type", Number)
], ɵParameterExpression.prototype, "dataValueType", void 0);

/**
 * @public
 * @abstract
 * @description The base class of the filter.
 * @extends {ɵSerializable}
 */
class ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BaseFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [this._instanceOfKey];
    }
    /**
     * @public
     * @description Method for checking if an instance is of the current class type.
     * @param instance - The instance to check.
     * @returns `true` if the instance is of the current class type, `false` otherwise.
     */
    static [Symbol.hasInstance](instance) {
        const instanceOfKeys = instance?.constructor?.['_instanceOfKeys'] ?? [];
        return instanceOfKeys.includes(this._instanceOfKey);
    }
    constructor(filterType) {
        /**
         * @description Indicates that the filter is active and will be taken into account when building the query.
         */
        this.isEnabled = true;
        /**
         * @description Drop the time for the date / time type parameters.
         */
        this.trimDateTimeParameterToDate = false;
        this.filterType = filterType;
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        throw new Error(`fromJson not supported on filter type ${this.constructor.name}`);
    }
    serializeItem(item) {
        return typeof item.toJson === 'function'
            ? item.toJson()
            : classToPlain(item, {
                exposeUnsetFields: false,
                excludePrefixes: ['_'],
            });
    }
    /**
     * Converts instance to json object.
     * @public
     */
    toJson() {
        return classToPlain(this, {
            exposeUnsetFields: false,
            excludePrefixes: ['_'],
        });
    }
}

/**
 * @public
 * @description Basic filter class with one condition.
 * @extends {ɵBaseFilter}
 */
class ɵSingleFilter extends ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_SingleFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(filterType, comparisonType, leftExpression) {
        super(filterType);
        this.comparisonType = comparisonType;
        this.leftExpression = leftExpression?.clone();
    }
    /** {@inheritDoc ɵSerializable.toJson} */
    toJson() {
        return {
            ...super.toJson(),
            leftExpression: this.serializeItem(this.leftExpression),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.leftExpression._getUsedColumns();
    }
}

/**
 * @public
 * @description Filter class of the "Between" type.
 * @extends {ɵSingleFilter}
 */
class ɵBetweenFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_BetweenFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(leftExpression, rightLessExpression, rightGreaterExpression) {
        super(ɵFilterType.Between, ɵComparisonType.Between, leftExpression);
        this.rightLessExpression = rightLessExpression?.clone();
        this.rightGreaterExpression = rightGreaterExpression?.clone();
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const filter = new ɵBetweenFilter(expressionParser.fromJson(dto['leftExpression'], filterParser), expressionParser.fromJson(dto['rightLessExpression'], filterParser), expressionParser.fromJson(dto['rightGreaterExpression'], filterParser));
        return filter;
    }
    clone() {
        const filter = new ɵBetweenFilter(this.leftExpression, this.rightLessExpression, this.rightGreaterExpression);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
    toJson() {
        return {
            ...super.toJson(),
            rightLessExpression: this.serializeItem(this.rightLessExpression),
            rightGreaterExpression: this.serializeItem(this.rightGreaterExpression),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            ...super._getUsedColumns(),
            ...this.rightLessExpression._getUsedColumns(),
            ...this.rightGreaterExpression._getUsedColumns(),
        ];
    }
}

/**
 * @public
 * @description Comparison filter class.
 * @extends {ɵSingleFilter}
 */
class ɵCompareFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_CompareFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(comparisonType, leftExpression, rightExpression) {
        super(ɵFilterType.Compare, comparisonType, leftExpression);
        this.rightExpression = rightExpression?.clone();
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const filter = new ɵCompareFilter(dto['comparisonType'], expressionParser.fromJson(dto['leftExpression'], filterParser), expressionParser.fromJson(dto['rightExpression'], filterParser));
        return filter;
    }
    clone() {
        const filter = new ɵCompareFilter(this.comparisonType, this.leftExpression, this.rightExpression);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
    toJson() {
        return {
            ...super.toJson(),
            rightExpression: this.serializeItem(this.rightExpression),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [...super._getUsedColumns(), ...this.rightExpression._getUsedColumns()];
    }
}

/**
 * @public
 * @description The comparison filter class [Exists according to the specified condition].
 * @extends {ɵSingleFilter}
 */
class ɵExistsFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ExistsFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(leftExpression, subFilters, comparisonType = ɵComparisonType.Exists, isAggregative = true) {
        super(ɵFilterType.Exists, comparisonType, leftExpression);
        this.subFilters = subFilters?.clone();
        this.isAggregative = isAggregative;
    }
    /**
     * Parses filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const leftExpression = expressionParser.fromJson(dto['leftExpression'], filterParser);
        const subFiltersDto = dto['subFilters'];
        let subFilters;
        if (subFiltersDto) {
            subFilters = filterParser.fromJson(dto['subFilters']);
        }
        return new ɵExistsFilter(leftExpression, subFilters, dto['comparisonType'], dto['isAggregative']);
    }
    /**
     * @public
     * @description Clones instance.
     */
    clone() {
        const filter = new ɵExistsFilter(this.leftExpression, this.subFilters, this.comparisonType, this.isAggregative);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
    /** {@inheritDoc ɵSerializable.toJson} */
    toJson() {
        const result = super.toJson();
        if (this.subFilters) {
            result['subFilters'] = this.serializeItem(this.subFilters);
        }
        return result;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [...super._getUsedColumns(), ...(this.subFilters?._getUsedColumns() ?? [])];
    }
}

/**
 * @public
 * @description In-Filter Class
 * @extends {ɵSingleFilter}
 */
class ɵInFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_InFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(comparisonType, leftExpression, rightExpressions) {
        super(ɵFilterType.In, comparisonType, leftExpression);
        this.rightExpressions = rightExpressions.map((expression) => expression?.clone()) ?? [];
    }
    /**
     * @description Parses in filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const dtoRightExpressions = (dto['rightExpressions'] ?? []);
        return new ɵInFilter(dto['comparisonType'], expressionParser.fromJson(dto['leftExpression'], filterParser), dtoRightExpressions.map((expression) => expressionParser.fromJson(expression, filterParser)));
    }
    /** {@inheritDoc ɵBaseFilter.clone} */
    clone() {
        const filter = new ɵInFilter(this.comparisonType, this.leftExpression, this.rightExpressions);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
    /** {@inheritDoc ɵSerializable.toJson} */
    toJson() {
        return {
            ...super.toJson(),
            rightExpressions: this.rightExpressions.map((expression) => this.serializeItem(expression)),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        const rightExpressionsUsedColumns = this.rightExpressions
            .map((expression) => expression._getUsedColumns())
            .flat();
        return [...super._getUsedColumns(), ...rightExpressionsUsedColumns];
    }
}

/**
 * @public
 * @description IsNull filter Class
 * @extends {ɵSingleFilter}
 */
class ɵIsNullFilter extends ɵSingleFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_IsNullFilter'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSingleFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(comparisonType, leftExpression) {
        if (![ɵComparisonType.Is_null, ɵComparisonType.Is_not_null].includes(comparisonType)) {
            console.warn(`Invalid comparison type: ${ɵComparisonType[comparisonType]} in IsNullFilter.`);
        }
        super(ɵFilterType.IsNull, ɵIsNullFilter._getComparisonType(comparisonType), leftExpression);
        /**
         * @description Flag that indicates if expression is empty or not.
         */
        this.isNull = true;
        this.isNull = this.comparisonType === ɵComparisonType.Is_null;
    }
    static _getComparisonType(comparisonType) {
        return comparisonType === ɵComparisonType.Is_not_null ? comparisonType : ɵComparisonType.Is_null;
    }
    /**
     * @description Parses isNull filter from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        return new ɵIsNullFilter(ɵIsNullFilter._getComparisonType(dto['comparisonType']), expressionParser.fromJson(dto['leftExpression'], filterParser));
    }
    /** {@inheritDoc ɵBaseFilter.clone} */
    clone() {
        const filter = new ɵIsNullFilter(ɵIsNullFilter._getComparisonType(this.comparisonType), this.leftExpression);
        filter.isEnabled = this.isEnabled;
        filter.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filter;
    }
}

/**
 * @public
 * @description Empty Guid constant.
 */
const ɵEMPTY_GUID = '00000000-0000-0000-0000-000000000000';
/**
 * @description Regular expression for Guid validation.
 * @public
 */
const validator = new RegExp('^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$', 'i');
/**
 * @description Generates guid part.
 * @param length Part length.
 * @public
 */
function generateGuidPart(length) {
    let out = '';
    for (let i = 0; i < length; i++) {
        // eslint-disable-next-line no-bitwise
        out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    }
    return out;
}
/**
 * @description Generates new Guid.
 * @public
 */
function ɵgenerateGuid() {
    return [
        generateGuidPart(2),
        generateGuidPart(1),
        generateGuidPart(1),
        generateGuidPart(1),
        generateGuidPart(3),
    ].join('-');
}
/**
 * @description Checks if value is Guid.
 * @param value Value for check.
 * @public
 */
function ɵisGuid(value) {
    return Boolean(value) && validator.test(value);
}
/**
 * @description Checks if Guid is empty value or equal to EMPTY_GUID.
 * @param guid Guid for check.
 * @public
 */
function ɵisEmptyGuid(guid) {
    return !guid || guid === ɵEMPTY_GUID;
}

/**
 * @public
 * @description An exception thrown when next handler already specified.
 */
class ɵNextHandlerAlreadySpecifiedException extends Error {
    constructor() {
        super('Next handler already specified');
    }
}

/**
 * @public
 * @description The exception thrown when trying to pass an empty or null argument.
 * @extends {Error}
 */
class ɵArgumentEmptyException extends Error {
    // region Constructors: Public
    constructor(argumentName) {
        super(`Argument with name '${argumentName}' is empty.`);
    }
}

/**
 * @public
 * @description The exception thrown when trying to pass argument that not in range.
 * @extends {Error}
 */
class ɵArgumentOutOfRangeException extends Error {
    // region Constructors: Public
    constructor(argumentName) {
        super(`Argument with name '${argumentName}' is out of range.`);
    }
}

/**
 * @public
 * @description An exception thrown when trying to retrieve a collection item and it not exists.
 * @extends {Error}
 */
class ɵItemNotFoundException extends Error {
    // region Constructors: Public
    constructor(key) {
        super(`Item with key '${key}' does not exist.`);
    }
}

/**
 * @public
 * @description Utilities class with various validation methods.
 */
class ɵValidationUtilities {
    // region Methods: Public
    /**
     * @public
     * @description Checks if argument is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkArgumentEmpty(argumentName, argumentValue) {
        if (argumentValue === undefined || argumentValue === null) {
            throw new ɵArgumentEmptyException(argumentName);
        }
    }
    /**
     * @public
     * @description Checks if array is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkArrayEmpty(argumentName, argumentValue) {
        ɵValidationUtilities.checkArgumentEmpty(argumentName, argumentValue);
        if (argumentValue.length === 0) {
            throw new ɵArgumentEmptyException(argumentName);
        }
    }
    /**
     * @public
     * @description Checks if string is empty.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     */
    static checkStringIsNotEmpty(argumentName, argumentValue) {
        this.checkArgumentEmpty(argumentName, argumentValue);
        if (argumentValue === '') {
            throw new ɵArgumentEmptyException(argumentName);
        }
        return argumentValue;
    }
    /**
     * @public
     * @description Checks if argument in range.
     * @param argumentName - Argument name.
     * @param argumentValue - Value of argument.
     * @param range - Range.
     */
    static checkArgumentOutOfRange(argumentName, argumentValue, range) {
        if (range.begin && argumentValue < range.begin) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
        if (range.end && argumentValue > range.end) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
    }
    static checkEnumOutOfRange(argumentName, argumentValue, range) {
        if (!range || !Object.values(range).includes(argumentValue)) {
            throw new ɵArgumentOutOfRangeException(argumentName);
        }
    }
}

/**
 * @public
 * @description Group of filters. Groups of filters can be nested into each other,
 * i.e. The collection itself can be an element of another collection.
 * @extends {ɵBaseFilter}
 */
class ɵFilterGroup extends ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_FilterGroup'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseFilter._instanceOfKeys, this._instanceOfKey];
    }
    /**
     * @public
     * @description Return collecion of deserialized filters.
     */
    get items() {
        const result = {};
        this.filters.forEach((filter, key) => {
            result[key] = this.serializeItem(filter);
        });
        return result;
    }
    constructor(logicalOperation = ɵLogicalOperatorType.And, rootSchemaName) {
        super(ɵFilterType.FilterGroup);
        /**
         * @public
         * @description Collection of child filters.
         */
        this.filters = new Map();
        this.logicalOperation = logicalOperation;
        this.rootSchemaName = rootSchemaName;
    }
    /**
     * @public
     * @description Parses filter group from json.
     */
    static fromJson(dto, filterParser) {
        const logicalOperation = dto['logicalOperation'];
        const group = new ɵFilterGroup(logicalOperation || ɵLogicalOperatorType.And, dto['rootSchemaName']);
        Object.entries(dto['items']).forEach(([key, filterDTO]) => {
            group.add(filterParser.fromJson(filterDTO), key);
        });
        return group;
    }
    /** {@inheritDoc ɵSerializable.toJson} */
    toJson() {
        const result = {};
        result['filterType'] = this.filterType;
        result['isEnabled'] = this.isEnabled;
        result['logicalOperation'] = this.logicalOperation;
        result['trimDateTimeParameterToDate'] = this.trimDateTimeParameterToDate;
        result['items'] = this.items;
        if (this.rootSchemaName) {
            result['rootSchemaName'] = this.rootSchemaName;
        }
        return result;
    }
    _createParameterValueExpressions(values) {
        return values.map((value) => new ɵParameterExpression({
            value: value,
        }));
    }
    /**
     * @public
     * @description Returns filter item by index.
     * @param index Item index.
     * @returns Filter.
     */
    getItemByIndex(index) {
        return this.items[Object.keys(this.items)[index]];
    }
    /**
     * @public
     * @description Creates and adds compare filter by value.
     * @param comparisonType Comparison type.
     * @param columnPath Column path.
     * @param parameterValue Parameter value.
     * @param filterKey Filter key.
     */
    addSchemaColumnFilterWithParameter(comparisonType, columnPath, parameterValue, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const rightExpression = new ɵParameterExpression({ value: parameterValue });
        const filter = new ɵCompareFilter(comparisonType, leftExpression, rightExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds in filter by values.
     * @param comparisonType Comparison type.
     * @param columnPath Column path.
     * @param parameterValues Parameter values.
     * @param filterKey Filter key.
     */
    addSchemaColumnInFilterWithParameters(comparisonType, columnPath, parameterValues, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const expressions = this._createParameterValueExpressions(parameterValues);
        const filter = new ɵInFilter(comparisonType, leftExpression, expressions);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds in filter for primary column.
     * @param primaryColumnValues Primary column values.
     * @param filterKey Filter key.
     */
    addPrimarySchemaColumnInFilter(primaryColumnValues, filterKey) {
        const leftExpression = new ɵMacrosFunctionExpression({ macrosType: ɵQueryMacrosType.PrimaryColumn });
        const expressions = this._createParameterValueExpressions(primaryColumnValues);
        const filter = new ɵInFilter(ɵComparisonType.Equal, leftExpression, expressions);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds null filter.
     * @param columnPath Path of column.
     * @param filterKey Filter key.
     */
    addSchemaColumnIsNullFilter(columnPath, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵIsNullFilter(ɵComparisonType.Is_null, leftExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds not null filter.
     * @param columnPath Column path.
     * @param filterKey Filter key.
     */
    addSchemaColumnIsNotNullFilter(columnPath, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵIsNullFilter(ɵComparisonType.Is_not_null, leftExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds filter between less value and greater value.
     * @param columnPath Column path.
     * @param lessParamValue Less parameter value.
     * @param greaterParamValue Greater parameter value.
     * @param filterKey Filter key.
     */
    addSchemaColumnBetweenFilterWithParameters(columnPath, lessParamValue, greaterParamValue, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const rightLessExpression = new ɵParameterExpression({ value: lessParamValue });
        const rightGreaterExpression = new ɵParameterExpression({ value: greaterParamValue });
        const filter = new ɵBetweenFilter(leftExpression, rightLessExpression, rightGreaterExpression);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds exists filter.
     * @param columnPath Column path.
     * @param subFilters Filters for exists expression.
     * @param filterKey Filter key.
     */
    addExistsFilter(columnPath, subFilters, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵExistsFilter(leftExpression, subFilters);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Creates and adds not exists filter.
     * @param columnPath Column path.
     * @param subFilters Filters for exist expression.
     * @param filterKey Filter key.
     */
    addNotExistsFilter(columnPath, subFilters, filterKey) {
        const leftExpression = new ɵColumnExpression({ columnPath });
        const filter = new ɵExistsFilter(leftExpression, subFilters, ɵComparisonType.Not_exists);
        this.add(filter, filterKey);
    }
    /**
     * @public
     * @description Adds filter.
     * @param filter Instance of filter.
     * @param filterKey Filter key.
     */
    add(filter, filterKey) {
        ɵValidationUtilities.checkArgumentEmpty('filter', filter);
        const key = filterKey || ɵgenerateGuid();
        this.filters.set(key, filter);
    }
    /**
     * @public
     * @description Clones instance.
     */
    clone() {
        const filterGroup = new ɵFilterGroup(this.logicalOperation, this.rootSchemaName);
        this.filters.forEach((filter, filterKey) => {
            filterGroup.filters.set(filterKey, filter.clone());
        });
        filterGroup.isEnabled = this.isEnabled;
        filterGroup.trimDateTimeParameterToDate = this.trimDateTimeParameterToDate;
        return filterGroup;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        const result = [];
        this.filters.forEach((filter) => {
            let filterUsedColumns = filter._getUsedColumns ? filter._getUsedColumns() : [];
            filterUsedColumns = filterUsedColumns.map((usedColumn) => ({
                path: usedColumn.path,
                schemaName: usedColumn.schemaName ?? this.rootSchemaName,
            }));
            result.push(...filterUsedColumns);
        });
        return result;
    }
}
__decorate([
    Exclude(),
    __metadata("design:type", Object)
], ɵFilterGroup.prototype, "filters", void 0);
__decorate([
    Expose(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], ɵFilterGroup.prototype, "items", null);

/**
 * @public
 * @description The class that defines the sub query expression.
 * @extends {ɵBaseExpression}
 */
class ɵSubQueryExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_SubQueryExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        /**
         * @description Type of expression.
         */
        this.expressionType = ɵExpressionType.SubQuery;
        this.columnPath = config?.columnPath;
        const subFilters = config?.subFilters;
        if (typeof subFilters?.clone === 'function') {
            this.subFilters = subFilters.clone();
        }
        else if (subFilters) {
            this.subFilters = new ɵFilterGroup(subFilters.logicalOperation, subFilters.rootSchemaName);
            Object.entries(subFilters.items).forEach(([key, filter]) => {
                this.subFilters.add(filter, key);
            });
        }
        else {
            this.subFilters = subFilters;
        }
        this.subOrderDirection = config.subOrderDirection;
        this.subOrderColumn = config.subOrderColumn;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser) {
        const expression = new ɵSubQueryExpression({
            columnPath: dto['columnPath'],
            subFilters: filterParser.fromJson(dto['subFilters']),
            subOrderDirection: dto['subOrderDirection'],
            subOrderColumn: dto['subOrderColumn'],
        });
        return expression;
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵSubQueryExpression(this);
    }
    /** {@inheritDoc ɵBaseExpression.toJson} */
    toJson() {
        const result = super.toJson();
        if (this.subFilters) {
            result['subFilters'] = this.serializeItem(this.subFilters);
        }
        return result;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [
            {
                path: this.columnPath,
            },
            ...(this.subFilters?._getUsedColumns() ?? []),
        ];
    }
}

/**
 * @public
 * @description A class that describes expression with aggregation sub query.
 * @extends {ɵSubQueryExpression}
 */
class ɵAggregationSubQueryExpression extends ɵSubQueryExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_AggregationSubQueryExpression'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵSubQueryExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        this.aggregationType = config?.aggregationType;
        this.functionType = config?.functionType;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser) {
        const subQueryExpression = super.fromJson(dto, filterParser);
        const expression = new ɵAggregationSubQueryExpression({
            aggregationType: dto['aggregationType'],
            functionType: dto['functionType'],
            ...subQueryExpression,
        });
        return expression;
    }
    /** {@inheritDoc ɵBaseExpression.clone} */
    clone() {
        return new ɵAggregationSubQueryExpression(this);
    }
}

/**
 * @public
 * @description Function expression resolver.
 */
class FunctionExpressionResolver {
    static resolve(dto) {
        const expressionClass = {
            [ɵFunctionType.Macros]: ɵMacrosFunctionExpression,
            [ɵFunctionType.Length]: ɵLengthFunctionExpression,
            [ɵFunctionType.Window]: ɵWindowFunctionExpression,
            [ɵFunctionType.DatePart]: ɵDatePartFunctionExpression,
            [ɵFunctionType.DateAdd]: ɵDatePartFunctionExpression,
            [ɵFunctionType.DateDiff]: ɵDatePartFunctionExpression,
        };
        return expressionClass[dto['functionType']];
    }
}
/**
 * @description Expression resolver.
 * @public
 */
class ɵExpressionResolver {
    /**
     * @public
     * @description Resolves class of expression.
     * @param expressionType - Type of expression.
     * @param dto - Serialized object.
     * @return Class of expression.
     */
    static resolve(expressionType, dto) {
        const expressionClass = {
            [ɵExpressionType.SchemaColumn]: ɵColumnExpression,
            [ɵExpressionType.Parameter]: ɵParameterExpression,
            [ɵExpressionType.SubQuery]: ɵAggregationSubQueryExpression,
            [ɵExpressionType.ArithmeticOperation]: ɵArithmeticExpression,
            [ɵExpressionType.Function]: FunctionExpressionResolver.resolve(dto),
        };
        return expressionClass[expressionType];
    }
}

/**
 * @public
 * @description Expression parser.
 * @extends {ɵBaseExpressionParser}
 */
class ɵExpressionParser extends ɵBaseExpressionParser {
    /**
     * @public
     * @description Parses expression from json.
     */
    static fromJson(dto, filterParser) {
        const expressionType = dto['expressionType'];
        return ɵExpressionResolver.resolve(expressionType, dto).fromJson(dto, filterParser, ɵExpressionParser);
    }
}

/**
 * @internal
 */
const ɵEXPRESSION_TYPE_MOCK = 0;
class ɵExpressionMock extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_ExpressionMock'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config || {});
        this.expressionType = ɵEXPRESSION_TYPE_MOCK;
    }
    // eslint-disable-next-line @typescript-eslint/ban-types
    get plainObject() {
        return classToPlain(this, { exposeUnsetFields: false, excludePrefixes: ['_'] });
    }
    clone() {
        return new ɵExpressionMock(this);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [];
    }
}

class ɵFilterMock extends ɵBaseFilter {
    /**
     * @internal
     * @dontChange
     */
    static { this._instanceOfKey = 'devkit_FilterMock'; }
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseFilter._instanceOfKeys, this._instanceOfKey];
    }
    constructor(filterType = ɵFilterType.None) {
        super(filterType);
    }
    // eslint-disable-next-line @typescript-eslint/ban-types
    get plainObject() {
        return classToPlain(this, { exposeUnsetFields: false, excludePrefixes: ['_'] });
    }
    clone() {
        return new ɵFilterMock(this.filterType);
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return [];
    }
}

/**
 * @internal
 */
const ɵparserMock = {
    fromJson: (data) => ({
        ...data,
        clone: () => data,
    }),
    clone: () => this,
};

/**
 * @public
 * @description Filter parser interface.
 */
class ɵBaseFilterParser {
    /**
     * @description Creates filter from JSON.
     * @param dto - Serialized filter.
     * @return Instance of filter.
     */
    static fromJson(dto) {
        throw new Error('not implemented');
    }
}

/**
 * @public
 * @description Filter resolver.
 */
class ɵFilterResolver {
    /**
     * @public
     * @description Resolves class of filter.
     * @param filterType - Type of filter.
     * @return Class of filter.
     */
    static resolve(filterType) {
        const filterClass = {
            [ɵFilterType.Compare]: ɵCompareFilter,
            [ɵFilterType.FilterGroup]: ɵFilterGroup,
            [ɵFilterType.In]: ɵInFilter,
            [ɵFilterType.IsNull]: ɵIsNullFilter,
            [ɵFilterType.Exists]: ɵExistsFilter,
            [ɵFilterType.Between]: ɵBetweenFilter,
        };
        return filterClass[filterType];
    }
}

/**
 * @public
 * @description Filter parser.
 */
class ɵFilterParser extends ɵBaseFilterParser {
    /**
     * @description Creates filter from JSON.
     * @param dto - Serialized filter.
     */
    static fromJson(dto) {
        const filterType = dto['filterType'];
        const isEnabled = dto['isEnabled'];
        const trimDateTimeParameterToDate = dto['trimDateTimeParameterToDate'];
        const filter = ɵFilterResolver.resolve(filterType).fromJson(dto, ɵFilterParser, ɵExpressionParser);
        filter.isEnabled = isEnabled !== false;
        filter.trimDateTimeParameterToDate = trimDateTimeParameterToDate === true;
        return filter;
    }
}

/**
 * @internal
 */
const ɵDEFAULT_FILTER_PLAIN_OBJ = {
    isEnabled: true,
    trimDateTimeParameterToDate: false,
};
/**
 * @internal
 */
function ɵgetFilterPlainObj(filterPlainObj) {
    return {
        ...ɵDEFAULT_FILTER_PLAIN_OBJ,
        ...filterPlainObj,
    };
}

/**
 * @public
 * @abstract
 * @classdesc Class contains info about column for query.
 */
class ɵBaseQueryColumn {
    constructor(expression) {
        /**
         * @description Order direction.
         */
        this.orderDirection = ɵOrderDirection.None;
        /**
         * @description Column order position.
         */
        this.orderPosition = -1;
        /**
         * @description If true then column data will be present in the result set.
         */
        this.isVisible = true;
        this.expression = expression;
    }
    /**
     * Adds ordering options.
     * @param orderDirection Direction.
     * @param orderPosition Position.
     * @returns Self.
     */
    withOrdering(orderDirection, orderPosition) {
        ɵValidationUtilities.checkArgumentEmpty('orderDirection', orderDirection);
        ɵValidationUtilities.checkArgumentEmpty('orderPosition', orderPosition);
        ɵValidationUtilities.checkArgumentOutOfRange('orderPosition', orderPosition, { begin: -1 });
        this.orderDirection = orderDirection;
        this.orderPosition = orderPosition;
        return this;
    }
    /**
     * Adds options.
     * @param options Options.
     * @returns Self.
     */
    withOptions(options) {
        ɵValidationUtilities.checkArgumentEmpty('options', options);
        this.isVisible = options.isVisible ?? true;
        return this;
    }
    /**
     * Adds caption.
     * @param caption Caption.
     * @returns Self.
     */
    withCaption(caption) {
        ɵValidationUtilities.checkArgumentEmpty('caption', caption);
        this.caption = caption;
        return this;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        return this.expression._getUsedColumns();
    }
}

/**
 * @public
 * @classdesc Class contains info about column with arithmetic operation.
 * @extends {ɵBaseQueryColumn}
 */
class ɵArithmeticQueryColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(arithmeticOperation, leftArithmeticOperand, rightArithmeticOperand) {
        super(new ɵArithmeticExpression({ arithmeticOperation, leftArithmeticOperand, rightArithmeticOperand }));
    }
}

/**
 * @public
 * @classdesc Class contains info about common column from entity.
 * @extends {ɵBaseQueryColumn}
 */
class ɵEntityQueryColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(columnPath) {
        super(new ɵColumnExpression({ columnPath }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column that contains parameter.
 * @extends {ɵBaseQueryColumn}
 */
class ɵParameterQueryColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(value, dataValueType) {
        super(new ɵParameterExpression({ value, dataValueType }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with sub query and aggregation.
 * @extends {ɵBaseQueryColumn}
 */
class ɵAggregationSubQueryColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(columnPath, aggregationType, subFilters, subOrderDirection, subOrderColumn) {
        super(new ɵAggregationSubQueryExpression({
            aggregationType,
            columnPath,
            subFilters,
            subOrderDirection,
            subOrderColumn,
        }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with sub query.
 * @extends {ɵBaseQueryColumn}
 */
class ɵSubQueryColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(columnPath, subFilters) {
        super(new ɵSubQueryExpression({ columnPath, subFilters }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with aggregation function.
 * @extends {ɵBaseQueryColumn}
 */
class ɵAggregationFunctionColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(aggregationType, aggregationEvalType, functionArgument) {
        super(new ɵAggregationFunctionExpression({
            aggregationType,
            aggregationEvalType,
            functionArgument,
        }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with date part.
 * @extends {ɵBaseQueryColumn}
 */
class ɵDatePartFunctionColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(datePartType, functionArgument) {
        super(new ɵDatePartFunctionExpression({
            datePartType,
            functionArgument,
        }));
    }
    // endregion
    // region Properties: Public
    /**
     * @description Part of date.
     */
    get datePartType() {
        return this.expression?.datePartType;
    }
}

/**
 * @public
 * @classdesc Class contains info about column with length function.
 * @extends {ɵBaseQueryColumn}
 */
class ɵLengthFunctionColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(functionArgument) {
        super(new ɵLengthFunctionExpression({ functionArgument }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with macros function.
 * @extends {ɵBaseQueryColumn}
 */
class ɵMacrosFunctionColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(macrosType) {
        super(new ɵMacrosFunctionExpression({ macrosType }));
    }
}

/**
 * @public
 * @classdesc Class contains info about column with window function.
 * @extends {ɵBaseQueryColumn}
 */
class ɵWindowFunctionColumn extends ɵBaseQueryColumn {
    // region Constructors: Public
    constructor(functionArgument) {
        super(new ɵWindowFunctionExpression({ functionArgument }));
    }
}

/**
 * @internal
 */
const ɵDEFAULT_COLUMN_PLAIN_OBJ = {
    isVisible: true,
    orderPosition: -1,
    orderDirection: ɵOrderDirection.None,
};
/**
 * @internal
 */
function ɵgetColumnPlainObj(expressionPlainObj) {
    return {
        ...ɵDEFAULT_COLUMN_PLAIN_OBJ,
        expression: expressionPlainObj,
    };
}

/**
 * @public
 * @description The base class of the data request. Implements {@link ɵMetadataProvider}
 */
class ɵBaseQuery {
    // region Constructors: Protected
    constructor(operationType, rootSchemaName) {
        this.operationType = operationType;
        this.rootSchemaName = rootSchemaName;
        ɵValidationUtilities.checkArgumentEmpty('_operationType', operationType);
        ɵValidationUtilities.checkArgumentEmpty('_rootSchemaName', rootSchemaName);
    }
    // endregion
    // region Methods: Public
    /** {@inheritDoc ɵMetadataProvider.getMetadata} */
    getMetadata() {
        return classToPlain(this, { exposeUnsetFields: false, excludePrefixes: ['_'] });
    }
}

/**
 * @public
 * @description The base class of the filtered data request.
 * @extends {ɵBaseQuery}
 */
class ɵBaseFilterableQuery extends ɵBaseQuery {
    constructor(operationType, rootSchemaName) {
        super(operationType, rootSchemaName);
        /**
         * @description Collection of filters.
         */
        this.filters = new ɵFilterGroup();
    }
    /**
     * @public
     * @description Clears all filters.
     */
    clearFilters() {
        this.filters.filters.clear();
    }
    /** {@inheritDoc ɵMetadataProvider.getMetadata} */
    getMetadata() {
        return {
            ...super.getMetadata(),
            filters: this.filters.toJson(),
        };
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        const usedColumns = this.filters._getUsedColumns();
        usedColumns.forEach((usedColumn) => {
            if (!usedColumn.schemaName) {
                usedColumn.schemaName = this.rootSchemaName;
            }
        });
        return usedColumns;
    }
}
__decorate([
    Exclude(),
    __metadata("design:type", Object)
], ɵBaseFilterableQuery.prototype, "filters", void 0);

/**
 * @public
 * @description Type of data query
 */
var ɵQueryOperationType;
(function (ɵQueryOperationType) {
    /**
     * @description Select query.
     */
    ɵQueryOperationType[ɵQueryOperationType["Select"] = 0] = "Select";
    /**
     * @description Insert query.
     */
    ɵQueryOperationType[ɵQueryOperationType["Insert"] = 1] = "Insert";
    /**
     * @description Edit query.
     */
    ɵQueryOperationType[ɵQueryOperationType["Update"] = 2] = "Update";
    /**
     * @description Delete query.
     */
    ɵQueryOperationType[ɵQueryOperationType["Delete"] = 3] = "Delete";
    // Batch,
    // ListSelect,
    // FilteredSelect,
    // LookupSelect
})(ɵQueryOperationType || (ɵQueryOperationType = {}));

/**
 * @public
 * @description Request to delete data.
 * @extends {ɵBaseFilterableQuery}
 */
class ɵDeleteQuery extends ɵBaseFilterableQuery {
    // region Constructors: Public
    constructor(rootSchemaName) {
        super(ɵQueryOperationType.Delete, rootSchemaName);
    }
}

/**
 * @public
 * @description The query class for data fetching.
 * @extends {ɵBaseFilterableQuery}
 */
class ɵEntitySchemaQuery extends ɵBaseFilterableQuery {
    constructor(rootSchemaName) {
        super(ɵQueryOperationType.Select, rootSchemaName);
        this._columns = new Map();
        /**
         * @description Determines if merge duplicates in query result.
         */
        this.isDistinct = false;
        /**
         * @description Number of rows to select.
         */
        this.rowCount = -1;
        /**
         * @description Rows count to skip.
         */
        this.rowsOffset = -1;
        /**
         * @description The parameter that determines whether to use the localized data.
         */
        this.useLocalization = true;
        /**
         * @description Represent parameter which determines disabling data in filtering.
         */
        this.useRecordDeactivation = false;
        this._getIsPageable = () => this.rowCount > 0 && this.rowsOffset > -1;
    }
    _internalAddColumn(key, column) {
        if (!this._columns.has(key)) {
            this._columns.set(key, column);
        }
    }
    // endregion
    // region Methods: Public
    /**
     * @public
     * @description Returns column instance by column alias.
     * @param columnAlias Column alias.
     * @returns [BaseQueryColumn] Column instance.
     */
    getColumn(columnAlias) {
        if (!this._columns.has(columnAlias)) {
            throw new ɵItemNotFoundException(columnAlias);
        }
        return this.findColumn(columnAlias);
    }
    /**
     * @public
     * @description Returns column instance by column alias if not found returns null.
     */
    findColumn(columnAlias) {
        return this._columns.get(columnAlias);
    }
    /** {@inheritDoc ɵMetadataProvider.getMetadata} */
    getMetadata() {
        return {
            ...super.getMetadata(),
            columns: {
                items: classToPlain(this._columns, { exposeUnsetFields: false, excludePrefixes: ['_'] }),
            },
            allColumns: false,
            isDistinct: this.isDistinct,
            ignoreDisplayValues: false,
            rowCount: this.rowCount,
            rowsOffset: this.rowsOffset,
            adminUnitRoleSources: this.adminUnitRoleSources,
            isPageable: this._getIsPageable(),
            conditionalValues: this.conditionalValues ? this.conditionalValues : null,
            isHierarchical: false,
            hierarchicalMaxDepth: 0,
            hierarchicalColumnName: null,
            hierarchicalColumnValue: null,
            useLocalization: this.useLocalization,
            useRecordDeactivation: this.useRecordDeactivation,
        };
    }
    /**
     * @public
     * @description Add column.
     * @param column Column path.
     * @param columnAlias Column alias.
     * @returns [BaseQueryColumn] Column instance.
     */
    addColumn(column, columnAlias) {
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add column by column path.
     * @param columnPath Column path.
     * @param columnAlias Column alias.
     * @returns [EntityQueryColumn] Column instance.
     */
    addSchemaColumn(columnPath, columnAlias) {
        ɵValidationUtilities.checkArgumentEmpty('columnPath', columnPath);
        const column = new ɵEntityQueryColumn(columnPath);
        this._internalAddColumn(columnAlias || columnPath, column);
        return column;
    }
    /**
     * @public
     * @description Add aggregation sub query column.
     * @param columnPath Column path.
     * @param aggregationType Aggregation type.
     * @param subFilters Sub query filters.
     * @param columnAlias Column alias.
     * @returns [AggregationSubQueryColumn] Column instance.
     */
    addAggregationSubQueryColumn(columnPath, aggregationType, aggregationConfig, columnAlias) {
        const column = new ɵAggregationSubQueryColumn(columnPath, aggregationType, aggregationConfig.filter, aggregationConfig.sortByDirection, aggregationConfig.sortByColumn);
        this._internalAddColumn(columnAlias || columnPath, column);
        return column;
    }
    /**
     * @public
     * @description Add parameter column.
     * @param value Parameter value.
     * @param dataValueType Parameter data value type.
     * @param columnAlias Column alias.
     * @returns [ParameterQueryColumn] Column instance.
     */
    addParameterColumn(value, dataValueType, columnAlias) {
        const column = new ɵParameterQueryColumn(value, dataValueType);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add arithmetic column.
     * @param arithmeticOperation Arithmetic operation.
     * @param leftArithmeticOperand Left arithmetic operand expression.
     * @param rightArithmeticOperand Right arithmetic operand expression.
     * @param columnAlias Column alias.
     * @returns [ArithmeticQueryColumn] Column instance.
     */
    addArithmeticColumn(arithmeticOperation, leftArithmeticOperand, rightArithmeticOperand, columnAlias) {
        const column = new ɵArithmeticQueryColumn(arithmeticOperation, leftArithmeticOperand, rightArithmeticOperand);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add macros function column.
     * @param queryMacrosType Macros type.
     * @param columnAlias Column alias.
     * @returns [MacrosFunctionColumn] Column instance.
     */
    addMacrosFunctionColumn(queryMacrosType, columnAlias) {
        const column = new ɵMacrosFunctionColumn(queryMacrosType);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add date part function column.
     * @param columnPath Column path.
     * @param datePartType Date part type.
     * @param columnAlias Column alias.
     * @returns [DatePartFunctionColumn] Column instance.
     */
    addDatePartFunctionColumn(columnPath, datePartType, columnAlias) {
        const functionArgument = new ɵColumnExpression({ columnPath });
        const column = new ɵDatePartFunctionColumn(datePartType, functionArgument);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add aggregation function column.
     * @param columnPath Column path.
     * @param aggregationType Aggregation type.
     * @param columnAlias Column alias.
     * @param aggregationEvalType Aggregation eval type.
     * @returns [AggregationFunctionColumn] Column instance.
     */
    addAggregationFunctionColumn(columnPath, aggregationType, columnAlias, aggregationEvalType = ɵAggregationEvalType.None) {
        const functionArgument = new ɵColumnExpression({ columnPath });
        const column = new ɵAggregationFunctionColumn(aggregationType, aggregationEvalType, functionArgument);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add length function column.
     * @param columnPath Column path.
     * @param columnAlias Column alias.
     * @returns [LengthFunctionColumn] Column instance.
     */
    addLengthFunctionColumn(columnPath, columnAlias) {
        const functionArgument = new ɵColumnExpression({ columnPath });
        const column = new ɵLengthFunctionColumn(functionArgument);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @public
     * @description Add window function column.
     * @param columnPath Column path.
     * @param aggregationType Aggregation type.
     * @param columnAlias Column alias.
     * @param aggregationEvalType Aggregation eval type.
     * @returns [WindowFunctionColumn] Column instance.
     */
    addWindowFunctionColumn(columnPath, aggregationType, columnAlias, aggregationEvalType = ɵAggregationEvalType.None) {
        const functionArgument = new ɵAggregationFunctionExpression({
            aggregationType,
            aggregationEvalType,
            functionArgument: new ɵColumnExpression({ columnPath }),
        });
        const column = new ɵWindowFunctionColumn(functionArgument);
        this._internalAddColumn(columnAlias, column);
        return column;
    }
    /**
     * @internal
     */
    _getUsedColumns() {
        const usedColumns = [];
        this._columns.forEach((column) => {
            usedColumns.push(...column._getUsedColumns());
        });
        usedColumns.forEach((usedColumn) => {
            if (!usedColumn.schemaName) {
                usedColumn.schemaName = this.rootSchemaName;
            }
        });
        return [...usedColumns, ...super._getUsedColumns()];
    }
}

/**
 * @public
 * @description Request to insert data.
 */
class ɵInsertQuery extends ɵBaseQuery {
    // endregion
    // region Constructors: Public
    constructor(rootSchemaName) {
        super(ɵQueryOperationType.Insert, rootSchemaName);
        // region Fields: Private
        this._columnValues = new Map();
    }
    // endregion
    // region Properties: Public
    /**
     * @description A collection of column values for updating entities.
     */
    get columnValues() {
        return {
            items: classToPlain(this._columnValues),
        };
    }
    // endregion
    // region Methods: Public
    /**
     * @public
     * @description Adds the value parameter. An instance of the value-parameter is created and
     * is added to the column values collection.
     * @param columnName The alias of the column to set value in
     * @param columnValue The value of the parameter. The value must match the data type.
     * @param dataValueType The type of the given parameter value.
     */
    addColumn(columnName, columnValue, dataValueType) {
        ɵValidationUtilities.checkArgumentEmpty('columnName', columnName);
        this._columnValues.set(columnName, new ɵParameterExpression({ value: columnValue, dataValueType }));
    }
}
__decorate([
    Expose(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], ɵInsertQuery.prototype, "columnValues", null);

/**
 * @public
 * @description Request for data changes.
 * @extends {ɵBaseFilterableQuery}
 */
class ɵUpdateQuery extends ɵBaseFilterableQuery {
    // endregion
    // region Constructors: Public
    constructor(rootSchemaName) {
        super(ɵQueryOperationType.Update, rootSchemaName);
        // region Fields: Private
        this._columnValues = new Map();
        // endregion
        // region Fields: Public
        /**
         * @description Flag that indicates whether to force save data.
         */
        this.isForceUpdate = false;
    }
    // endregion
    // region Properties: Public
    /**
     * @description A collection of column values for updating entities.
     */
    get columnValues() {
        return {
            items: classToPlain(this._columnValues),
        };
    }
    // endregion
    // region Methods: Public
    /**
     * @public
     * @description Adds the value parameter. An instance of the value-parameter is created and
     * is added to the column values collection.
     * @param columnName The alias of the column to set value in
     * @param columnValue The value of the parameter. The value must match the data type.
     * @param dataValueType The type of the given parameter value.
     */
    addColumn(columnName, columnValue, dataValueType) {
        ɵValidationUtilities.checkArgumentEmpty('columnName', columnName);
        this._columnValues.set(columnName, new ɵParameterExpression({ value: columnValue, dataValueType }));
    }
}
__decorate([
    Expose(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], ɵUpdateQuery.prototype, "columnValues", null);

/**
 * @public
 * @description The query class for localized data fetching.
 * @extends {ɵEntitySchemaQuery}
 */
class ɵSelectLocalizationQuery extends ɵEntitySchemaQuery {
}

/**
 * @public
 * @description Base definition of request.
 * @template TEvent
 */
class ɵBaseRequest {
}

/* eslint-disable @typescript-eslint/no-explicit-any */

// noinspection NonAsciiCharacters, JSNonASCIINames

// noinspection NonAsciiCharacters, JSNonASCIINames
/**
 * @public
 * @classdesc Represents a disposable resource, similar to the execution of Observable. A
 * Subscription has one important method, `unsubscribe`, that just disposes
 * the resource held by the subscription.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
class ɵSubscription {
    /**
     * @public
     * @description A flag to indicate whether this Subscription has already been unsubscribed.
     */
    get closed() {
        return this._closed;
    }
    /**
     * @param _initialTeardown A function executed as part of the finalization.
     * process when {@link #unsubscribe} is called.
     */
    constructor(_initialTeardown) {
        this._initialTeardown = _initialTeardown;
        this._closed = false;
    }
    /**
     * @public
     * @description Disposes the resources held by the subscription.
     */
    unsubscribe() {
        if (!this._closed) {
            this._closed = true;
            if (isFunction(this._initialTeardown)) {
                this._initialTeardown();
            }
        }
    }
}

/**
 * @public
 * @description Enum for model parameter types.
 */
var ɵModelParameterType;
(function (ɵModelParameterType) {
    /**
     * @public
     * @description Parameter represents primary column value.
     */
    ɵModelParameterType["PrimaryColumnValue"] = "primaryColumnValue";
    /**
     * @public
     * @description Parameter represents column value.
     */
    ɵModelParameterType["ColumnValue"] = "columnValue";
    /**
     * @public
     * @description Parameter represents filter value.
     */
    ɵModelParameterType["Filter"] = "filter";
    /**
     * @public
     * @description Parameter represents primary display value filter.
     */
    ɵModelParameterType["PrimaryDisplayValueFilter"] = "primaryDisplayValueFilter";
    /**
     * @public
     * @description Parameter is empty.
     */
    ɵModelParameterType["Empty"] = "empty";
    /**
     * @public
     * @description Parameter is never used.
     */
    ɵModelParameterType["Never"] = "never";
})(ɵModelParameterType || (ɵModelParameterType = {}));

/**
 * @public
 * @description Actions of mask request.
 */
var ɵSchemaViewMaskRequestAction;
(function (ɵSchemaViewMaskRequestAction) {
    /**
     * @description To add task.
     */
    ɵSchemaViewMaskRequestAction["AddTask"] = "addTask";
    /**
     * @description To remove task.
     */
    ɵSchemaViewMaskRequestAction["RemoveTask"] = "removeTask";
})(ɵSchemaViewMaskRequestAction || (ɵSchemaViewMaskRequestAction = {}));
/**
 * @public
 * @description Request to show mask.
 */
class ɵSchemaViewMaskRequest extends ɵBaseRequest {
    constructor(action, taskName, $context) {
        super();
        this.action = action;
        this.taskName = taskName;
        this.$context = $context;
        /**
         * @description Request type
         */
        this.type = 'crt.SchemaViewMaskRequest';
    }
}

// noinspection NonAsciiCharacters, JSNonASCIINames

/**
 * Generated bundle index. Do not edit.
 */

export { LookupMode, eAdminUnitRoleSources, ɵAggregationEvalType, ɵAggregationFunction, ɵAggregationFunctionColumn, ɵAggregationFunctionExpression, ɵAggregationSubQueryColumn, ɵAggregationSubQueryExpression, ɵAggregationType, ɵArgumentEmptyException, ɵArgumentFunctionExpression, ɵArgumentOutOfRangeException, ɵArithmeticExpression, ɵArithmeticOperation, ɵArithmeticQueryColumn, ɵBaseExpression, ɵBaseExpressionParser, ɵBaseFilter, ɵBaseFilterParser, ɵBaseFilterableQuery, ɵBaseQuery, ɵBaseQueryColumn, ɵBaseRequest, ɵBetweenFilter, ɵColumnExpression, ɵCompareFilter, ɵComparisonType, ɵDEFAULT_COLUMN_PLAIN_OBJ, ɵDEFAULT_FILTER_PLAIN_OBJ, ɵDataSchemaAttributeType, ɵDataSchemaAttributeUsageType, ɵDataSchemaType, ɵDataSourceScope, ɵDataValueType, ɵDatePartFunctionColumn, ɵDatePartFunctionExpression, ɵDatePartType, ɵDefaultValueSource, ɵDeleteQuery, ɵEMPTY_GUID, ɵEXPRESSION_TYPE_MOCK, ɵEntityQueryColumn, ɵEntitySchemaQuery, ɵExistsFilter, ɵExpressionMock, ɵExpressionParser, ɵExpressionResolver, ɵExpressionType, ɵFilterGroup, ɵFilterMock, ɵFilterParser, ɵFilterResolver, ɵFilterType, ɵFunctionExpression, ɵFunctionType, ɵInFilter, ɵInsertQuery, ɵIsNullFilter, ɵItemNotFoundException, ɵLengthFunctionColumn, ɵLengthFunctionExpression, ɵLogicalOperatorType, ɵMacrosFunctionColumn, ɵMacrosFunctionExpression, ɵMessageChannelType, ɵModelInPageAction, ɵModelParameterType, ɵNextHandlerAlreadySpecifiedException, ɵOrderDirection, ɵParameterExpression, ɵParameterQueryColumn, ɵPlatformAPIs, ɵQueryMacrosType, ɵQueryOperationType, ɵSchemaViewMaskRequest, ɵSchemaViewMaskRequestAction, ɵSelectLocalizationQuery, ɵSingleFilter, ɵSubQueryColumn, ɵSubQueryExpression, ɵSubscription, ɵUpdateQuery, ɵValidationUtilities, ɵWindowFunctionColumn, ɵWindowFunctionExpression, ɵencodeDate, ɵgenerateGuid, ɵgetColumnPlainObj, ɵgetFilterPlainObj, ɵisEmptyGuid, ɵisGuid, ɵparserMock, ɵtoLocalISOString };
